// stream-chat-react-native-core/lib/commonjs/components/MessageList/utils/getGroupStyles.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getGroupStyles = exports.getGroupStyle = void 0;
var _utils = require("../../../utils/utils");
var getGroupStyle = exports.getGroupStyle = function getGroupStyle(_ref) {
  var _message$user, _previousMessage$user, _nextMessage$user;
  var message = _ref.message,
    previousMessage = _ref.previousMessage,
    nextMessage = _ref.nextMessage,
    maxTimeBetweenGroupedMessages = _ref.maxTimeBetweenGroupedMessages,
    nextMessageDateSeparatorDate = _ref.nextMessageDateSeparatorDate,
    dateSeparatorDate = _ref.dateSeparatorDate;
  var groupStyles = [];
  var isPrevMessageTypeDeleted = (previousMessage == null ? void 0 : previousMessage.type) === 'deleted';
  var isNextMessageTypeDeleted = (nextMessage == null ? void 0 : nextMessage.type) === 'deleted';
  var userId = (message == null || (_message$user = message.user) == null ? void 0 : _message$user.id) || null;
  var isTopMessage = !previousMessage || previousMessage.type === 'system' || previousMessage.type === 'error' || userId !== (previousMessage == null || (_previousMessage$user = previousMessage.user) == null ? void 0 : _previousMessage$user.id) || !!isPrevMessageTypeDeleted || dateSeparatorDate || (0, _utils.isEditedMessage)(previousMessage);
  var isBottomMessage = !nextMessage || nextMessage.type === 'system' || nextMessage.type === 'error' || userId !== (nextMessage == null || (_nextMessage$user = nextMessage.user) == null ? void 0 : _nextMessage$user.id) || !!isNextMessageTypeDeleted || nextMessageDateSeparatorDate || maxTimeBetweenGroupedMessages !== undefined && nextMessage.created_at.getTime() - message.created_at.getTime() > maxTimeBetweenGroupedMessages || (0, _utils.isEditedMessage)(message);
  if (isTopMessage) {
    groupStyles.push('top');
  }
  var isMessageTypeDeleted = message.type === 'deleted';
  if (isBottomMessage) {
    if (isTopMessage || isMessageTypeDeleted || message.type === 'error') {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    } else {
      groupStyles.push('bottom');
    }
  }
  if (!isTopMessage && !isBottomMessage) {
    if (isMessageTypeDeleted || message.type === 'error') {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    } else {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('middle');
    }
  }
  return groupStyles;
};
var getGroupStyles = exports.getGroupStyles = function getGroupStyles(params) {
  var dateSeparators = params.dateSeparators,
    maxTimeBetweenGroupedMessages = params.maxTimeBetweenGroupedMessages,
    messages = params.messages,
    noGroupByUser = params.noGroupByUser;
  if (noGroupByUser) {
    return {};
  }
  var messageGroupStyles = {};
  for (var i = 0; i < messages.length; i++) {
    var previousMessage = messages[i - 1];
    var message = messages[i];
    var nextMessage = messages[i + 1];
    if (message.id) {
      messageGroupStyles[message.id] = getGroupStyle({
        dateSeparatorDate: dateSeparators[message.id],
        maxTimeBetweenGroupedMessages: maxTimeBetweenGroupedMessages,
        message: message,
        nextMessage: nextMessage,
        previousMessage: previousMessage
      });
    }
  }
  return messageGroupStyles;
};
