// react-native-compressor/lib/module/utils/index.js
"use strict";

import { Compressor } from '../Main';
import { Platform } from 'react-native';
import { toNativeOptions } from './helpers';
const INCORRECT_INPUT_PATH = 'Incorrect input path. Please provide a valid one';
export const DEFAULT_COMPRESS_AUDIO_OPTIONS = {
  // bitrate: 96,
  quality: 'medium'
};
export const generateFilePath = extension => {
  return new Promise((resolve, reject) => {
    Compressor.generateFilePath(extension).then(result => resolve('file://' + result)).catch(error => reject(error));
  });
};
export const getRealPath = (path, type = 'video') => {
  return Compressor.getRealPath(path, type);
};
export const getVideoMetaData = path => {
  return Compressor.getVideoMetaData(path);
};
const unifyMetaData = exifResult => {
  const output = {};
  const isIos = Platform.OS === 'ios';
  output.ImageWidth = isIos ? exifResult?.PixelWidth : parseInt(exifResult.ImageWidth);
  output.ImageHeight = isIos ? exifResult?.PixelHeight : parseInt(exifResult.ImageLength);
  output.Orientation = isIos ? exifResult.Orientation : parseInt(exifResult.Orientation);
  output.size = exifResult.size;
  output.extension = exifResult.extension;
  output.exif = exifResult;
  return output;
};
export const getImageMetaData = async path => {
  const result = await Compressor.getImageMetaData(path);
  return unifyMetaData(result);
};
export const createVideoThumbnail = (fileUrl, options = {}) => {
  return Compressor.createVideoThumbnail(fileUrl, toNativeOptions(options));
};
export const clearCache = cacheDir => {
  return Compressor.clearCache(cacheDir);
};
const isValidUrl = url => /^(?:\w+:)?\/\/([^\s\.]+\.\S{2}|localhost[\:?\d]*)\S*$/.test(url);
const getFullFilename = path => {
  if (typeof path === 'string') {
    let _path = path;

    // In case of remote media, check if the url would be valid one
    if (path.includes('http') && !isValidUrl(path)) {
      return INCORRECT_INPUT_PATH;
    }

    // In case of url, check if it ends with "/" and do not consider it furthermore
    if (_path[_path.length - 1] === '/') _path = _path.substring(0, path.length - 1);
    const array = _path.split('/');
    return array.length > 1 ? array[array.length - 1] : INCORRECT_INPUT_PATH;
  }
  return INCORRECT_INPUT_PATH;
};
const isFileNameError = filename => {
  return filename === INCORRECT_INPUT_PATH;
};
const getFilename = path => {
  const fullFilename = getFullFilename(path);
  if (fullFilename && !isFileNameError(fullFilename)) {
    const array = fullFilename.split('.');
    return array.length > 1 ? array.slice(0, -1).join('') : array.join('');
  }
  return fullFilename;
};
const isRemoteMedia = path => {
  return typeof path === 'string' ? path?.split(':/')?.[0]?.includes('http') : null;
};
export const getDetails = (mediaFullPath, extesnion = 'mp3') => {
  return new Promise(async (resolve, reject) => {
    try {
      // Since we used "-v error", a work around is to call first this command before the following
      const result = {};
      if (result !== 0) {
        throw new Error('Failed to execute command');
      }

      // get the output result of the command
      // example of output {"programs": [], "streams": [{"width": 640,"height": 360}], "format": {"size": "15804433"}}
      let mediaInfo = await {};
      mediaInfo = JSON.parse(mediaInfo);

      // execute second command
      const mediaInformation = await {};

      // treat both results
      mediaInformation.filename = getFilename(mediaFullPath);
      mediaInformation.bitrate = mediaInformation.getMediaProperties().bit_rate;
      mediaInformation.extension = extesnion;
      mediaInformation.isRemoteMedia = isRemoteMedia(mediaFullPath);
      mediaInformation.size = Number(mediaInfo.format.size);
      resolve(mediaInformation);
    } catch (e) {
      reject(e);
    }
  });
};
export const getFileSize = async filePath => {
  return Compressor.getFileSize(filePath);
};
export * from './Downloader';
export * from './Uploader';
export * from './helpers';
