// react-native-compressor/lib/module/utils/Uploader.js
"use strict";

import { Platform } from 'react-native';
import { Compressor } from '../Main';
import { toNativeOptions, uuidv4 } from './helpers';
export let UploadType = /*#__PURE__*/function (UploadType) {
  UploadType[UploadType["BINARY_CONTENT"] = 0] = "BINARY_CONTENT";
  UploadType[UploadType["MULTIPART"] = 1] = "MULTIPART";
  return UploadType;
}({});
export let UploaderHttpMethod = /*#__PURE__*/function (UploaderHttpMethod) {
  UploaderHttpMethod["POST"] = "POST";
  UploaderHttpMethod["PUT"] = "PUT";
  UploaderHttpMethod["PATCH"] = "PATCH";
  return UploaderHttpMethod;
}({});
export const cancelUpload = (uuid = '', shouldCancelAll = false) => {
  return Compressor.cancelUpload(uuid, shouldCancelAll);
};
export const backgroundUpload = async (url, fileUrl, options, onProgress, abortSignal) => {
  const uuid = uuidv4();
  try {
    if (Platform.OS === 'android' && fileUrl.includes('file://')) {
      fileUrl = fileUrl.replace('file://', '');
    }
    if (options?.getCancellationId) {
      options?.getCancellationId(uuid);
    }
    abortSignal?.addEventListener('abort', () => cancelUpload(uuid));
    const result = await Compressor.upload(fileUrl, toNativeOptions({
      ...options,
      uuid,
      method: options.httpMethod,
      url
    }), onProgress);
    return result;
  } finally {
    abortSignal?.removeEventListener('abort', () => cancelUpload(uuid));
  }
};
