// stream-chat-react-native-core/lib/commonjs/components/MessageList/hooks/useMessageList.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMessageList = exports.shouldIncludeMessageInList = void 0;
var _react = require("react");
var _ChannelContext = require("../../../contexts/channelContext/ChannelContext");
var _ChatContext = require("../../../contexts/chatContext/ChatContext");
var _MessagesContext = require("../../../contexts/messagesContext/MessagesContext");
var _PaginatedMessageListContext = require("../../../contexts/paginatedMessageListContext/PaginatedMessageListContext");
var _ThreadContext = require("../../../contexts/threadContext/ThreadContext");
var _hooks = require("../../../hooks");
var _getDateSeparators = require("../utils/getDateSeparators");
var _getGroupStyles = require("../utils/getGroupStyles");
var shouldIncludeMessageInList = exports.shouldIncludeMessageInList = function shouldIncludeMessageInList(message, options) {
  var _message$user;
  var deletedMessagesVisibilityType = options.deletedMessagesVisibilityType,
    userId = options.userId;
  var isMessageTypeDeleted = message.type === 'deleted';
  var isSender = ((_message$user = message.user) == null ? void 0 : _message$user.id) === userId;
  if (!isMessageTypeDeleted) {
    return true;
  }
  switch (deletedMessagesVisibilityType) {
    case 'always':
      return true;
    case 'sender':
      return isSender;
    case 'receiver':
      return !isSender;
    case 'never':
    default:
      return false;
  }
};
var useMessageList = exports.useMessageList = function useMessageList(params) {
  var noGroupByUser = params.noGroupByUser,
    threadList = params.threadList,
    isLiveStreaming = params.isLiveStreaming,
    _params$isFlashList = params.isFlashList,
    isFlashList = _params$isFlashList === void 0 ? false : _params$isFlashList;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  var _useChannelContext = (0, _ChannelContext.useChannelContext)(),
    hideDateSeparators = _useChannelContext.hideDateSeparators,
    maxTimeBetweenGroupedMessages = _useChannelContext.maxTimeBetweenGroupedMessages;
  var _useMessagesContext = (0, _MessagesContext.useMessagesContext)(),
    deletedMessagesVisibilityType = _useMessagesContext.deletedMessagesVisibilityType,
    _useMessagesContext$g = _useMessagesContext.getMessagesGroupStyles,
    getMessagesGroupStyles = _useMessagesContext$g === void 0 ? _getGroupStyles.getGroupStyles : _useMessagesContext$g;
  var _usePaginatedMessageL = (0, _PaginatedMessageListContext.usePaginatedMessageListContext)(),
    messages = _usePaginatedMessageL.messages,
    viewabilityChangedCallback = _usePaginatedMessageL.viewabilityChangedCallback;
  var _useThreadContext = (0, _ThreadContext.useThreadContext)(),
    threadMessages = _useThreadContext.threadMessages;
  var messageList = threadList ? threadMessages : messages;
  var processedMessageList = (0, _react.useMemo)(function () {
    var newMessageList = [];
    for (var message of messageList) {
      if (!shouldIncludeMessageInList(message, {
        deletedMessagesVisibilityType: deletedMessagesVisibilityType,
        userId: client.userID
      })) {
        continue;
      }
      if (isFlashList) {
        newMessageList.push(message);
      } else {
        newMessageList.unshift(message);
      }
    }
    return newMessageList;
  }, [messageList, deletedMessagesVisibilityType, client.userID, isFlashList]);
  var dateSeparators = (0, _react.useMemo)(function () {
    return (0, _getDateSeparators.getDateSeparators)({
      hideDateSeparators: hideDateSeparators,
      messages: processedMessageList
    });
  }, [hideDateSeparators, processedMessageList]);
  var dateSeparatorsRef = (0, _react.useRef)(dateSeparators);
  dateSeparatorsRef.current = dateSeparators;
  var messageGroupStyles = (0, _react.useMemo)(function () {
    return getMessagesGroupStyles({
      dateSeparators: dateSeparatorsRef.current,
      hideDateSeparators: hideDateSeparators,
      maxTimeBetweenGroupedMessages: maxTimeBetweenGroupedMessages,
      messages: processedMessageList,
      noGroupByUser: noGroupByUser,
      userId: client.userID
    });
  }, [getMessagesGroupStyles, hideDateSeparators, maxTimeBetweenGroupedMessages, processedMessageList, noGroupByUser, client.userID]);
  var messageGroupStylesRef = (0, _react.useRef)(messageGroupStyles);
  messageGroupStylesRef.current = messageGroupStyles;
  var data = (0, _hooks.useRAFCoalescedValue)(processedMessageList, isLiveStreaming);
  return (0, _react.useMemo)(function () {
    return {
      dateSeparatorsRef: dateSeparatorsRef,
      messageGroupStylesRef: messageGroupStylesRef,
      processedMessageList: data,
      rawMessageList: messageList,
      viewabilityChangedCallback: viewabilityChangedCallback
    };
  }, [data, messageList, viewabilityChangedCallback]);
};
