// stream-chat-react-native-core/lib/module/store/mappers/mapMessageToStorable.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapMessageToStorable = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _mapDateTimeToStorable = require("./mapDateTimeToStorable");
var _excluded = ["attachments", "cid", "created_at", "deleted_at", "deleted_for_me", "id", "latest_reactions", "message_text_updated_at", "own_reactions", "poll", "poll_id", "reminder", "reaction_groups", "shared_location", "text", "type", "updated_at", "user"];
var mapMessageToStorable = exports.mapMessageToStorable = function mapMessageToStorable(message) {
  var attachments = message.attachments,
    cid = message.cid,
    created_at = message.created_at,
    deleted_at = message.deleted_at,
    deleted_for_me = message.deleted_for_me,
    id = message.id,
    latest_reactions = message.latest_reactions,
    message_text_updated_at = message.message_text_updated_at,
    own_reactions = message.own_reactions,
    poll = message.poll,
    poll_id = message.poll_id,
    reminder = message.reminder,
    reaction_groups = message.reaction_groups,
    shared_location = message.shared_location,
    text = message.text,
    type = message.type,
    updated_at = message.updated_at,
    user = message.user,
    extraData = (0, _objectWithoutProperties2.default)(message, _excluded);
  return {
    attachments: JSON.stringify(attachments),
    cid: cid || '',
    createdAt: (0, _mapDateTimeToStorable.mapDateTimeToStorable)(created_at),
    deletedAt: (0, _mapDateTimeToStorable.mapDateTimeToStorable)(deleted_at),
    deletedForMe: deleted_for_me,
    extraData: JSON.stringify(extraData),
    id: id,
    messageTextUpdatedAt: (0, _mapDateTimeToStorable.mapDateTimeToStorable)(message_text_updated_at),
    poll_id: poll_id || '',
    reactionGroups: JSON.stringify(reaction_groups),
    shared_location: JSON.stringify(shared_location),
    text: text,
    type: type,
    updatedAt: (0, _mapDateTimeToStorable.mapDateTimeToStorable)(updated_at),
    userId: user == null ? void 0 : user.id
  };
};
