// stream-chat-react-native-core/lib/module/store/mappers/mapStorableToMessage.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapStorableToMessage = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _mapStorableToPoll = require("./mapStorableToPoll");
var _mapStorableToReaction = require("./mapStorableToReaction");
var _mapStorableToReminder = require("./mapStorableToReminder");
var _mapStorableToUser = require("./mapStorableToUser");
var _excluded = ["createdAt", "deletedAt", "deletedForMe", "extraData", "messageTextUpdatedAt", "poll_id", "reactionGroups", "shared_location", "updatedAt", "user"];
var mapStorableToMessage = exports.mapStorableToMessage = function mapStorableToMessage(_ref) {
  var currentUserId = _ref.currentUserId,
    messageRow = _ref.messageRow,
    pollRow = _ref.pollRow,
    reactionRows = _ref.reactionRows,
    reminderRow = _ref.reminderRow;
  var createdAt = messageRow.createdAt,
    deletedAt = messageRow.deletedAt,
    deletedForMe = messageRow.deletedForMe,
    extraData = messageRow.extraData,
    messageTextUpdatedAt = messageRow.messageTextUpdatedAt,
    poll_id = messageRow.poll_id,
    reactionGroups = messageRow.reactionGroups,
    shared_location = messageRow.shared_location,
    updatedAt = messageRow.updatedAt,
    user = messageRow.user,
    rest = (0, _objectWithoutProperties2.default)(messageRow, _excluded);
  var latestReactions = (reactionRows == null ? void 0 : reactionRows.map(function (reaction) {
    return (0, _mapStorableToReaction.mapStorableToReaction)(reaction);
  })) || [];
  var ownReactions = latestReactions.filter(function (reaction) {
    var _reaction$user;
    return ((_reaction$user = reaction.user) == null ? void 0 : _reaction$user.id) === currentUserId;
  });
  return Object.assign({}, rest, {
    attachments: messageRow.attachments ? JSON.parse(messageRow.attachments) : [],
    created_at: createdAt,
    deleted_at: deletedAt,
    deleted_for_me: deletedForMe,
    latest_reactions: latestReactions,
    message_text_updated_at: messageTextUpdatedAt,
    own_reactions: ownReactions,
    poll_id: poll_id,
    reaction_groups: reactionGroups ? JSON.parse(reactionGroups) : {},
    shared_location: shared_location ? JSON.parse(shared_location) : null,
    updated_at: updatedAt,
    user: (0, _mapStorableToUser.mapStorableToUser)(user)
  }, pollRow ? {
    poll: (0, _mapStorableToPoll.mapStorableToPoll)(pollRow)
  } : {}, extraData ? JSON.parse(extraData) : {}, reminderRow ? {
    reminder: (0, _mapStorableToReminder.mapStorableToReminder)(reminderRow)
  } : {});
};
