// stream-chat-react-native-core/lib/commonjs/components/ChannelPreview/hooks/useLatestMessagePreview.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLatestMessagePreview = exports.MessageReadStatus = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _react = require("react");
var _useMessageDeliveryStatus = require("./useMessageDeliveryStatus");
var _ChatContext = require("../../../contexts/chatContext/ChatContext");
var _TranslationContext = require("../../../contexts/translationContext/TranslationContext");
var _hooks = require("../../../hooks");
var _useTranslatedMessage = require("../../../hooks/useTranslatedMessage");
var _utils = require("../../../utils/utils");
var selector = function selector(nextValue) {
  return {
    createdBy: nextValue.created_by,
    latestVotesByOption: nextValue.latest_votes_by_option,
    name: nextValue.name
  };
};
var getMessageSenderName = function getMessageSenderName(message, currentUserId, t, membersLength) {
  var _message$user;
  if ((message == null || (_message$user = message.user) == null ? void 0 : _message$user.id) === currentUserId) {
    return t('You');
  }
  if (membersLength > 2) {
    var _message$user2, _message$user3, _message$user4;
    return (message == null || (_message$user2 = message.user) == null ? void 0 : _message$user2.name) || (message == null || (_message$user3 = message.user) == null ? void 0 : _message$user3.username) || (message == null || (_message$user4 = message.user) == null ? void 0 : _message$user4.id) || '';
  }
  return '';
};
var getMentionUsers = function getMentionUsers(mentionedUser) {
  if (Array.isArray(mentionedUser)) {
    var mentionUserString = mentionedUser.reduce(function (acc, cur) {
      var userName = cur.name || cur.id || '';
      if (userName) {
        acc += `${acc.length ? '|' : ''}@${userName}`;
      }
      return acc;
    }, '');
    return mentionUserString.replace(/[.*+?^${}()|[\]\\]/g, function (match) {
      return '\\' + match;
    });
  }
  return '';
};
var getLatestMessageDisplayText = function getLatestMessageDisplayText(channel, client, draftMessage, message, t, pollState) {
  var _message$attachments;
  if (!message) {
    return [{
      bold: false,
      text: t('Nothing yet...')
    }];
  }
  var isMessageTypeDeleted = message.type === 'deleted';
  if (isMessageTypeDeleted) {
    return [{
      bold: false,
      text: t('Message deleted')
    }];
  }
  var currentUserId = client == null ? void 0 : client.userID;
  var members = Object.keys(channel.state.members);
  var messageSender = getMessageSenderName(message, currentUserId, t, members.length);
  var messageSenderText = messageSender ? `${messageSender === t('You') ? '' : '@'}${messageSender}: ` : '';
  var boldOwner = messageSenderText.includes('@');
  if (draftMessage) {
    var _draftMessage$attachm;
    if ((_draftMessage$attachm = draftMessage.attachments) != null && _draftMessage$attachm.length) {
      return [{
        bold: true,
        draft: true,
        text: 'Draft:'
      }, {
        bold: false,
        text: t('🏙 Attachment...')
      }];
    }
    if (draftMessage.text) {
      return [{
        bold: true,
        draft: true,
        text: 'Draft:'
      }, {
        bold: false,
        text: draftMessage.text
      }];
    }
  }
  if (message.shared_location) {
    return [{
      bold: false,
      text: '📍'
    }, {
      bold: false,
      text: t('Location')
    }];
  }
  if (message.text) {
    var shortenedText = message.text.substring(0, 100).replace(/\n/g, ' ');
    var mentionedUsers = getMentionUsers(message.mentioned_users);
    var regEx = new RegExp(`^(${mentionedUsers})`);
    return [{
      bold: boldOwner,
      text: messageSenderText
    }].concat((0, _toConsumableArray2.default)(shortenedText.split('').reduce(function (acc, cur, index) {
      if (cur === '@' && mentionedUsers && regEx.test(shortenedText.substring(index))) {
        acc.push({
          bold: true,
          text: cur
        });
      } else if (mentionedUsers && regEx.test(acc[acc.length - 1].text)) {
        acc.push({
          bold: false,
          text: cur
        });
      } else {
        acc[acc.length - 1].text += cur;
      }
      return acc;
    }, [{
      bold: false,
      text: ''
    }])));
  }
  if (message.command) {
    return [{
      bold: boldOwner,
      text: messageSenderText
    }, {
      bold: false,
      text: `/${message.command}`
    }];
  }
  if ((_message$attachments = message.attachments) != null && _message$attachments.length) {
    return [{
      bold: boldOwner,
      text: messageSenderText
    }, {
      bold: false,
      text: t('🏙 Attachment...')
    }];
  }
  if (message.poll_id && pollState) {
    var _latestVotes$, _previewUser, _previewUser2;
    var createdBy = pollState.createdBy,
      latestVotesByOption = pollState.latestVotesByOption,
      name = pollState.name;
    var latestVotes;
    if (latestVotesByOption) {
      latestVotes = Object.values(latestVotesByOption).map(function (votes) {
        return votes == null ? void 0 : votes[0];
      }).sort(function (a, b) {
        return new Date(b.created_at).getTime() - new Date(a.created_at).getTime();
      });
    }
    var previewAction = 'created';
    var previewUser = createdBy;
    if (latestVotes && latestVotes.length && (_latestVotes$ = latestVotes[0]) != null && _latestVotes$.user) {
      var _latestVotes$2;
      previewAction = 'voted';
      previewUser = (_latestVotes$2 = latestVotes[0]) == null ? void 0 : _latestVotes$2.user;
    }
    var previewMessage = `${client.userID === ((_previewUser = previewUser) == null ? void 0 : _previewUser.id) ? 'You' : (_previewUser2 = previewUser) == null ? void 0 : _previewUser2.name} ${previewAction}: ${name}`;
    return [{
      bold: false,
      text: '📊 '
    }, {
      bold: false,
      text: previewMessage
    }];
  }
  return [{
    bold: boldOwner,
    text: messageSenderText
  }, {
    bold: false,
    text: t('Empty message...')
  }];
};
var MessageReadStatus = exports.MessageReadStatus = function (MessageReadStatus) {
  MessageReadStatus[MessageReadStatus["NOT_SENT_BY_CURRENT_USER"] = 0] = "NOT_SENT_BY_CURRENT_USER";
  MessageReadStatus[MessageReadStatus["UNREAD"] = 1] = "UNREAD";
  MessageReadStatus[MessageReadStatus["READ"] = 2] = "READ";
  MessageReadStatus[MessageReadStatus["DELIVERED"] = 3] = "DELIVERED";
  return MessageReadStatus;
}({});
var getLatestMessagePreview = function getLatestMessagePreview(params) {
  var channel = params.channel,
    client = params.client,
    draftMessage = params.draftMessage,
    lastMessage = params.lastMessage,
    pollState = params.pollState,
    status = params.status,
    t = params.t;
  var messages = channel.state.messages;
  if (!messages.length && !lastMessage) {
    return {
      created_at: '',
      messageObject: undefined,
      previews: [{
        bold: false,
        text: t('Nothing yet...')
      }],
      status: _useMessageDeliveryStatus.MessageDeliveryStatus.NOT_SENT_BY_CURRENT_USER
    };
  }
  var channelStateLastMessage = messages.length ? messages[messages.length - 1] : undefined;
  var message = lastMessage !== undefined ? lastMessage : channelStateLastMessage;
  return {
    created_at: message == null ? void 0 : message.created_at,
    messageObject: message,
    previews: getLatestMessageDisplayText(channel, client, draftMessage, message, t, pollState),
    status: status
  };
};
var textComposerStateSelector = function textComposerStateSelector(state) {
  return {
    text: state.text
  };
};
var stateSelector = function stateSelector(state) {
  return {
    attachments: state.attachments
  };
};
var useLatestMessagePreview = exports.useLatestMessagePreview = function useLatestMessagePreview(channel, forceUpdate, lastMessage) {
  var _lastMessage$poll_id, _useStateStore3;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  var _useTranslationContex = (0, _TranslationContext.useTranslationContext)(),
    t = _useTranslationContex.t;
  var _useStateStore = (0, _hooks.useStateStore)(channel.messageComposer.textComposer.state, textComposerStateSelector),
    draftText = _useStateStore.text;
  var _useStateStore2 = (0, _hooks.useStateStore)(channel.messageComposer.attachmentManager.state, stateSelector),
    attachments = _useStateStore2.attachments;
  var draftMessage = (0, _react.useMemo)(function () {
    return !channel.messageComposer.compositionIsEmpty ? {
      attachments: attachments,
      id: channel.messageComposer.id,
      text: draftText
    } : undefined;
  }, [channel.messageComposer, attachments, draftText]);
  var channelConfigExists = typeof (channel == null ? void 0 : channel.getConfig) === 'function';
  var translatedLastMessage = (0, _useTranslatedMessage.useTranslatedMessage)(lastMessage);
  var channelLastMessageString = translatedLastMessage ? (0, _utils.stringifyMessage)({
    message: translatedLastMessage
  }) : '';
  var readEvents = (0, _react.useMemo)(function () {
    var _channel$getConfig;
    if (!channelConfigExists) {
      return true;
    }
    var read_events = !channel.disconnected && !!(channel != null && channel.id) && ((_channel$getConfig = channel.getConfig()) == null ? void 0 : _channel$getConfig.read_events);
    if (typeof read_events !== 'boolean') {
      return true;
    }
    return read_events;
  }, [channelConfigExists, channel]);
  var _useMessageDeliverySt = (0, _useMessageDeliveryStatus.useMessageDeliveryStatus)({
      channel: channel,
      isReadEventsEnabled: readEvents,
      lastMessage: lastMessage
    }),
    status = _useMessageDeliverySt.status;
  var pollId = (_lastMessage$poll_id = lastMessage == null ? void 0 : lastMessage.poll_id) != null ? _lastMessage$poll_id : '';
  var poll = client.polls.fromState(pollId);
  var pollState = (_useStateStore3 = (0, _hooks.useStateStore)(poll == null ? void 0 : poll.state, selector)) != null ? _useStateStore3 : {};
  var createdBy = pollState.createdBy,
    latestVotesByOption = pollState.latestVotesByOption,
    name = pollState.name;
  var latestMessagePreview = (0, _react.useMemo)(function () {
    return getLatestMessagePreview({
      channel: channel,
      client: client,
      draftMessage: draftMessage,
      lastMessage: translatedLastMessage,
      pollState: pollState,
      status: status,
      t: t
    });
  }, [channelLastMessageString, status, draftMessage, forceUpdate, latestVotesByOption, createdBy, name]);
  return latestMessagePreview;
};
