// stream-chat-react-native-core/lib/commonjs/hooks/useAudioPlayerControl.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAudioPlayerControl = void 0;
var _react = require("react");
var _AudioPlayerContext = require("../contexts/audioPlayerContext/AudioPlayerContext");
var makeAudioPlayerId = function makeAudioPlayerId(_ref) {
  var requester = _ref.requester,
    src = _ref.src,
    id = _ref.id;
  return `${requester != null ? requester : 'requester-unknown'}:${src}:${id != null ? id : ''}`;
};
var useAudioPlayerControl = exports.useAudioPlayerControl = function useAudioPlayerControl(_ref2) {
  var duration = _ref2.duration,
    mimeType = _ref2.mimeType,
    playbackRates = _ref2.playbackRates,
    previewVoiceRecording = _ref2.previewVoiceRecording,
    _ref2$requester = _ref2.requester,
    requester = _ref2$requester === void 0 ? '' : _ref2$requester,
    type = _ref2.type,
    uri = _ref2.uri,
    fileId = _ref2.id;
  var _useAudioPlayerContex = (0, _AudioPlayerContext.useAudioPlayerContext)(),
    audioPlayerPool = _useAudioPlayerContex.audioPlayerPool;
  var id = makeAudioPlayerId({
    id: fileId,
    requester: requester,
    src: uri != null ? uri : ''
  });
  var audioPlayer = (0, _react.useMemo)(function () {
    return audioPlayerPool == null ? void 0 : audioPlayerPool.getOrAddPlayer({
      duration: duration != null ? duration : 0,
      id: id,
      mimeType: mimeType != null ? mimeType : '',
      playbackRates: playbackRates,
      previewVoiceRecording: previewVoiceRecording,
      type: type != null ? type : 'audio',
      uri: uri != null ? uri : ''
    });
  }, [audioPlayerPool, duration, id, mimeType, playbackRates, previewVoiceRecording, type, uri]);
  return audioPlayer;
};
