// stream-chat-react-native-core/lib/commonjs/hooks/usePrunableMessageList.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePrunableMessageList = usePrunableMessageList;
var _react = require("react");
var _useStableCallback = require("./useStableCallback");
var calculateSafeGap = function calculateSafeGap(maximumMessages) {
  return 0.2 * maximumMessages;
};
var isNearEnd = function isNearEnd(_ref) {
  var rangeConfig = _ref.rangeConfig,
    maximumMessageLimit = _ref.maximumMessageLimit;
  var first = rangeConfig.first,
    last = rangeConfig.last,
    inverted = rangeConfig.inverted;
  var safeGap = calculateSafeGap(maximumMessageLimit);
  if (!inverted) return first <= safeGap;
  return last >= maximumMessageLimit - 1 - safeGap;
};
function usePrunableMessageList(_ref2) {
  var rawSetMessages = _ref2.setMessages,
    maximumMessageLimit = _ref2.maximumMessageLimit;
  var visibleRangeConfigRef = (0, _react.useRef)({
    first: 0,
    inverted: true,
    last: -1
  });
  var viewabilityChangedCallback = (0, _useStableCallback.useStableCallback)(function (_ref3) {
    var viewableItems = _ref3.viewableItems,
      _ref3$inverted = _ref3.inverted,
      inverted = _ref3$inverted === void 0 ? true : _ref3$inverted;
    if (!(viewableItems != null && viewableItems.length) || maximumMessageLimit == null) return;
    var first = Infinity;
    var last = -1;
    for (var v of viewableItems) {
      if (v.index == null) continue;
      if (v.index < first) first = v.index;
      if (v.index > last) last = v.index;
    }
    if (first !== Infinity) visibleRangeConfigRef.current = {
      first: first,
      inverted: inverted,
      last: last
    };
  });
  var setMessages = (0, _useStableCallback.useStableCallback)(function (channel) {
    var rangeConfig = visibleRangeConfigRef.current;
    if (maximumMessageLimit == null || channel.state.messages.length <= maximumMessageLimit || isNearEnd({
      maximumMessageLimit: maximumMessageLimit,
      rangeConfig: rangeConfig
    })) {
      rawSetMessages(channel);
      return;
    }
    channel.state.pruneOldest(maximumMessageLimit);
    rawSetMessages(channel);
  });
  return {
    setMessages: setMessages,
    viewabilityChangedCallback: viewabilityChangedCallback
  };
}
