// stream-chat-react-native-core/lib/module/state-store/audio-player.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AudioPlayer = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _streamChat = require("stream-chat");
var _native = require("../native");
var DEFAULT_PLAYBACK_RATES = [1.0, 1.5, 2.0];
var DEFAULT_PLAYER_SETTINGS = {
  pitchCorrectionQuality: 'high',
  progressUpdateIntervalMillis: 100,
  shouldCorrectPitch: true
};
var INITIAL_STATE = {
  currentPlaybackRate: 1.0,
  duration: 0,
  isPlaying: false,
  playbackRates: DEFAULT_PLAYBACK_RATES,
  position: 0,
  progress: 0
};
var AudioPlayer = exports.AudioPlayer = function () {
  function AudioPlayer(options) {
    var _this = this,
      _options$previewVoice,
      _options$playbackRate;
    (0, _classCallCheck2.default)(this, AudioPlayer);
    this.playerRef = null;
    this._pool = null;
    this.initPlayer = function () {
      var _ref2 = (0, _asyncToGenerator2.default)(function* (_ref) {
        var _NativeHandlers$Sound;
        var uri = _ref.uri,
          playerRef = _ref.playerRef;
        if (playerRef) {
          _this.playerRef = playerRef;
          return;
        }
        if (_this.previewVoiceRecording) {
          var _NativeHandlers$Audio;
          if ((_NativeHandlers$Audio = _native.NativeHandlers.Audio) != null && _NativeHandlers$Audio.startPlayer) {
            var _NativeHandlers$Audio2;
            yield _native.NativeHandlers.Audio.startPlayer(uri, {}, _this.onVoiceRecordingPreviewPlaybackStatusUpdate);
            if ((_NativeHandlers$Audio2 = _native.NativeHandlers.Audio) != null && _NativeHandlers$Audio2.pausePlayer) {
              yield _native.NativeHandlers.Audio.pausePlayer();
            }
          }
          return;
        }
        if (!_this.isExpoCLI || !uri) {
          return;
        }
        if ((_NativeHandlers$Sound = _native.NativeHandlers.Sound) != null && _NativeHandlers$Sound.initializeSound) {
          var _NativeHandlers$Sound2;
          _this.playerRef = yield (_NativeHandlers$Sound2 = _native.NativeHandlers.Sound) == null ? void 0 : _NativeHandlers$Sound2.initializeSound({
            uri: uri
          }, DEFAULT_PLAYER_SETTINGS, _this.onPlaybackStatusUpdate);
        }
      });
      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }();
    this.onVoiceRecordingPreviewPlaybackStatusUpdate = function () {
      var _ref3 = (0, _asyncToGenerator2.default)(function* (playbackStatus) {
        var currentProgress = playbackStatus.currentPosition / playbackStatus.duration;
        if (currentProgress === 1) {
          yield _this.stop();
        } else {
          _this.progress = currentProgress;
        }
      });
      return function (_x2) {
        return _ref3.apply(this, arguments);
      };
    }();
    this.onPlaybackStatusUpdate = function () {
      var _ref4 = (0, _asyncToGenerator2.default)(function* (playbackStatus) {
        if (!playbackStatus.isLoaded) {
          if (playbackStatus.error) {
            console.log(`Encountered a fatal error during playback: ${playbackStatus.error}`);
          }
        } else {
          var durationMillis = playbackStatus.durationMillis,
            positionMillis = playbackStatus.positionMillis;
          if (_this.type !== 'voiceRecording') {
            _this.duration = durationMillis;
          }
          if (playbackStatus.isPlaying) {
            var duration = _this.type === 'voiceRecording' ? _this.duration : durationMillis;
            if (positionMillis <= duration) {
              _this.position = positionMillis;
            }
          }
          if (playbackStatus.didJustFinish && !playbackStatus.isLooping) {
            yield _this.stop();
          }
        }
      });
      return function (_x3) {
        return _ref4.apply(this, arguments);
      };
    }();
    this.isExpoCLI = _native.NativeHandlers.SDK === 'stream-chat-expo';
    this._id = options.id;
    this.type = options.type;
    this.previewVoiceRecording = (_options$previewVoice = options.previewVoiceRecording) != null ? _options$previewVoice : false;
    var playbackRates = (_options$playbackRate = options.playbackRates) != null ? _options$playbackRate : DEFAULT_PLAYBACK_RATES;
    this.state = new _streamChat.StateStore(Object.assign({}, INITIAL_STATE, {
      currentPlaybackRate: playbackRates[0],
      duration: options.duration * 1000,
      playbackRates: playbackRates
    }));
    this.initPlayer({
      uri: options.uri
    });
  }
  return (0, _createClass2.default)(AudioPlayer, [{
    key: "isPlaying",
    get: function get() {
      return this.state.getLatestValue().isPlaying;
    },
    set: function set(isPlaying) {
      this.state.partialNext({
        isPlaying: isPlaying
      });
    }
  }, {
    key: "duration",
    get: function get() {
      return this.state.getLatestValue().duration;
    },
    set: function set(duration) {
      this.state.partialNext({
        duration: duration
      });
    }
  }, {
    key: "position",
    get: function get() {
      return this.state.getLatestValue().position;
    },
    set: function set(position) {
      this.state.partialNext({
        position: position,
        progress: position / this.duration
      });
    }
  }, {
    key: "progress",
    get: function get() {
      return this.state.getLatestValue().progress;
    },
    set: function set(progress) {
      this.state.partialNext({
        position: progress * this.duration,
        progress: progress
      });
    }
  }, {
    key: "playbackRates",
    get: function get() {
      return this.state.getLatestValue().playbackRates;
    }
  }, {
    key: "currentPlaybackRate",
    get: function get() {
      return this.state.getLatestValue().currentPlaybackRate;
    }
  }, {
    key: "id",
    get: function get() {
      return this._id;
    }
  }, {
    key: "pool",
    set: function set(pool) {
      this._pool = pool;
    }
  }, {
    key: "changePlaybackRate",
    value: function () {
      var _changePlaybackRate = (0, _asyncToGenerator2.default)(function* () {
        var _this$playerRef;
        var currentPlaybackRateIndex = this.state.getLatestValue().playbackRates.indexOf(this.currentPlaybackRate);
        if (currentPlaybackRateIndex === -1) {
          currentPlaybackRateIndex = 0;
        }
        var nextPlayBackIndex = currentPlaybackRateIndex === this.playbackRates.length - 1 ? 0 : currentPlaybackRateIndex + 1;
        var nextPlaybackRate = this.playbackRates[nextPlayBackIndex];
        this.state.partialNext({
          currentPlaybackRate: nextPlaybackRate
        });
        if (!this.playerRef) {
          return;
        }
        if ((_this$playerRef = this.playerRef) != null && _this$playerRef.setRateAsync) {
          yield this.playerRef.setRateAsync(nextPlaybackRate, true, 'high');
        }
      });
      function changePlaybackRate() {
        return _changePlaybackRate.apply(this, arguments);
      }
      return changePlaybackRate;
    }()
  }, {
    key: "play",
    value: function play() {
      if (this.isPlaying) {
        return;
      }
      if (this._pool) {
        this._pool.requestPlay(this.id);
      }
      if (this.previewVoiceRecording) {
        var _NativeHandlers$Audio3;
        if ((_NativeHandlers$Audio3 = _native.NativeHandlers.Audio) != null && _NativeHandlers$Audio3.resumePlayer) {
          _native.NativeHandlers.Audio.resumePlayer();
        }
        this.state.partialNext({
          isPlaying: true
        });
        return;
      }
      if (!this.playerRef) {
        return;
      }
      if (this.isExpoCLI) {
        var _this$playerRef2;
        if ((_this$playerRef2 = this.playerRef) != null && _this$playerRef2.playAsync) {
          this.playerRef.playAsync();
        }
      } else {
        var _this$playerRef3;
        if ((_this$playerRef3 = this.playerRef) != null && _this$playerRef3.resume) {
          this.playerRef.resume();
        }
      }
      this.state.partialNext({
        isPlaying: true
      });
    }
  }, {
    key: "pause",
    value: function pause() {
      if (!this.isPlaying) {
        return;
      }
      if (this.previewVoiceRecording) {
        var _NativeHandlers$Audio4;
        if ((_NativeHandlers$Audio4 = _native.NativeHandlers.Audio) != null && _NativeHandlers$Audio4.pausePlayer) {
          _native.NativeHandlers.Audio.pausePlayer();
        }
        this.state.partialNext({
          isPlaying: false
        });
        return;
      }
      if (!this.playerRef) {
        return;
      }
      if (this.isExpoCLI) {
        var _this$playerRef4;
        if ((_this$playerRef4 = this.playerRef) != null && _this$playerRef4.pauseAsync) {
          this.playerRef.pauseAsync();
        }
      } else {
        var _this$playerRef5;
        if ((_this$playerRef5 = this.playerRef) != null && _this$playerRef5.pause) {
          this.playerRef.pause();
        }
      }
      this.state.partialNext({
        isPlaying: false
      });
      if (this._pool) {
        this._pool.notifyPaused();
      }
    }
  }, {
    key: "toggle",
    value: function toggle() {
      if (this.isPlaying) {
        this.pause();
      } else {
        this.play();
      }
    }
  }, {
    key: "seek",
    value: function () {
      var _seek = (0, _asyncToGenerator2.default)(function* (positionInSeconds) {
        if (this.previewVoiceRecording) {
          var _NativeHandlers$Audio5;
          this.position = positionInSeconds;
          if ((_NativeHandlers$Audio5 = _native.NativeHandlers.Audio) != null && _NativeHandlers$Audio5.seekToPlayer) {
            _native.NativeHandlers.Audio.seekToPlayer(positionInSeconds * 1000);
          }
          return;
        }
        if (!this.playerRef) {
          return;
        }
        this.position = positionInSeconds;
        if (this.isExpoCLI) {
          if (positionInSeconds === 0) {
            var _this$playerRef6;
            if ((_this$playerRef6 = this.playerRef) != null && _this$playerRef6.replayAsync) {
              yield this.playerRef.replayAsync({});
            }
          } else {
            var _this$playerRef7;
            if ((_this$playerRef7 = this.playerRef) != null && _this$playerRef7.setPositionAsync) {
              yield this.playerRef.setPositionAsync(positionInSeconds);
            }
          }
        } else {
          var _this$playerRef8;
          if ((_this$playerRef8 = this.playerRef) != null && _this$playerRef8.seek) {
            this.playerRef.seek(positionInSeconds);
          }
        }
      });
      function seek(_x4) {
        return _seek.apply(this, arguments);
      }
      return seek;
    }()
  }, {
    key: "stop",
    value: function () {
      var _stop = (0, _asyncToGenerator2.default)(function* () {
        yield this.seek(0);
        this.pause();
      });
      function stop() {
        return _stop.apply(this, arguments);
      }
      return stop;
    }()
  }, {
    key: "onRemove",
    value: function onRemove() {
      if (this.previewVoiceRecording) {
        var _NativeHandlers$Audio6;
        if ((_NativeHandlers$Audio6 = _native.NativeHandlers.Audio) != null && _NativeHandlers$Audio6.stopPlayer) {
          _native.NativeHandlers.Audio.stopPlayer();
        }
        this.state.partialNext(Object.assign({}, INITIAL_STATE, {
          currentPlaybackRate: this.playbackRates[0],
          playbackRates: DEFAULT_PLAYBACK_RATES
        }));
        return;
      }
      if (this.isExpoCLI) {
        var _this$playerRef9;
        if ((_this$playerRef9 = this.playerRef) != null && _this$playerRef9.stopAsync && this.playerRef.unloadAsync) {
          this.playerRef.stopAsync();
          this.playerRef.unloadAsync();
        }
      }
      this.playerRef = null;
      this.state.partialNext(Object.assign({}, INITIAL_STATE, {
        currentPlaybackRate: this.playbackRates[0],
        playbackRates: DEFAULT_PLAYBACK_RATES
      }));
    }
  }]);
}();
