// stream-chat-react-native-core/lib/module/state-store/message-list-prev-next-state.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessagePreviousAndNextMessageStore = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _streamChat = require("stream-chat");
var INITIAL_STATE = {
  messageList: {}
};
var MessagePreviousAndNextMessageStore = exports.MessagePreviousAndNextMessageStore = function () {
  function MessagePreviousAndNextMessageStore() {
    (0, _classCallCheck2.default)(this, MessagePreviousAndNextMessageStore);
    this.state = new _streamChat.StateStore(INITIAL_STATE);
  }
  return (0, _createClass2.default)(MessagePreviousAndNextMessageStore, [{
    key: "setMessageListPreviousAndNextMessage",
    value: function setMessageListPreviousAndNextMessage(_ref) {
      var messages = _ref.messages,
        _ref$isFlashList = _ref.isFlashList,
        isFlashList = _ref$isFlashList === void 0 ? true : _ref$isFlashList;
      var currentValue = this.state.getLatestValue();
      var prevMessageList = currentValue.messageList;
      var newMessageList = {};
      for (var i = 0; i < messages.length; i++) {
        var message = messages[i];
        var previousMessage = isFlashList ? messages[i - 1] : messages[i + 1];
        var nextMessage = isFlashList ? messages[i + 1] : messages[i - 1];
        var existing = prevMessageList[message.id];
        if (existing && existing.previousMessage === previousMessage && existing.nextMessage === nextMessage) {
          newMessageList[message.id] = existing;
        } else {
          newMessageList[message.id] = {
            nextMessage: nextMessage,
            previousMessage: previousMessage
          };
        }
      }
      this.state.partialNext({
        messageList: newMessageList
      });
    }
  }]);
}();
