// stream-chat-react-native-core/lib/commonjs/store/apis/softDeleteMessage.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.softDeleteMessage = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _createUpdateQuery = require("../sqlite-utils/createUpdateQuery");
var _SqliteClient = require("../SqliteClient");
var softDeleteMessage = exports.softDeleteMessage = function () {
  var _ref2 = (0, _asyncToGenerator2.default)(function* (_ref) {
    var _ref$deleteForMe = _ref.deleteForMe,
      deleteForMe = _ref$deleteForMe === void 0 ? false : _ref$deleteForMe,
      _ref$execute = _ref.execute,
      execute = _ref$execute === void 0 ? true : _ref$execute,
      id = _ref.id;
    var query = (0, _createUpdateQuery.createUpdateQuery)('messages', {
      deletedAt: deleteForMe ? undefined : new Date().toISOString(),
      deletedForMe: deleteForMe,
      type: 'deleted'
    }, {
      id: id
    });
    _SqliteClient.SqliteClient.logger == null || _SqliteClient.SqliteClient.logger('info', 'softDeleteMessage', {
      deleteForMe: deleteForMe,
      execute: execute,
      id: id
    });
    if (execute) {
      yield _SqliteClient.SqliteClient.executeSql.apply(null, query);
    }
    return [query];
  });
  return function softDeleteMessage(_x) {
    return _ref2.apply(this, arguments);
  };
}();
