// stream-chat-react-native-core/lib/commonjs/components/Attachment/AudioAttachment.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AudioAttachment = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _dayjs = _interopRequireDefault(require("dayjs"));
var _duration = _interopRequireDefault(require("dayjs/plugin/duration"));
var _streamChat = require("stream-chat");
var _contexts = require("../../contexts");
var _hooks = require("../../hooks");
var _useAudioPlayerControl = require("../../hooks/useAudioPlayerControl");
var _icons = require("../../icons");
var _native = require("../../native");
var _getTrimmedAttachmentTitle = require("../../utils/getTrimmedAttachmentTitle");
var _ProgressControl = require("../ProgressControl/ProgressControl");
var _WaveProgressBar = require("../ProgressControl/WaveProgressBar");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Attachment/AudioAttachment.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ONE_HOUR_IN_MILLISECONDS = 3600 * 1000;
var ONE_SECOND_IN_MILLISECONDS = 1000;
_dayjs.default.extend(_duration.default);
var audioPlayerSelector = function audioPlayerSelector(state) {
  return {
    currentPlaybackRate: state.currentPlaybackRate,
    duration: state.duration,
    isPlaying: state.isPlaying,
    position: state.position,
    progress: state.progress
  };
};
var AudioAttachment = exports.AudioAttachment = function AudioAttachment(props) {
  var _item$duration, _item$mime_type, _message$parent_id, _item$asset_url, _NativeHandlers$Sound;
  var soundRef = _react.default.useRef(null);
  var _props$hideProgressBa = props.hideProgressBar,
    hideProgressBar = _props$hideProgressBa === void 0 ? false : _props$hideProgressBa,
    item = props.item,
    message = props.message,
    _props$showSpeedSetti = props.showSpeedSettings,
    showSpeedSettings = _props$showSpeedSetti === void 0 ? false : _props$showSpeedSetti,
    testID = props.testID,
    titleMaxLength = props.titleMaxLength,
    _props$isPreview = props.isPreview,
    isPreview = _props$isPreview === void 0 ? false : _props$isPreview;
  var isVoiceRecording = (0, _streamChat.isVoiceRecordingAttachment)(item);
  var audioPlayer = (0, _useAudioPlayerControl.useAudioPlayerControl)({
    duration: (_item$duration = item.duration) != null ? _item$duration : 0,
    mimeType: (_item$mime_type = item.mime_type) != null ? _item$mime_type : '',
    requester: isPreview ? 'preview' : (message == null ? void 0 : message.id) && `${(_message$parent_id = message == null ? void 0 : message.parent_id) != null ? _message$parent_id : message == null ? void 0 : message.id}${message == null ? void 0 : message.id}`,
    type: isVoiceRecording ? 'voiceRecording' : 'audio',
    uri: (_item$asset_url = item.asset_url) != null ? _item$asset_url : ''
  });
  var _useStateStore = (0, _hooks.useStateStore)(audioPlayer.state, audioPlayerSelector),
    duration = _useStateStore.duration,
    isPlaying = _useStateStore.isPlaying,
    position = _useStateStore.position,
    progress = _useStateStore.progress,
    currentPlaybackRate = _useStateStore.currentPlaybackRate;
  (0, _react.useEffect)(function () {
    if (soundRef.current) {
      audioPlayer.initPlayer({
        playerRef: soundRef.current
      });
    }
  }, [audioPlayer]);
  (0, _react.useEffect)(function () {
    return function () {
      if (isPreview) {
        audioPlayer.onRemove();
      }
    };
  }, [audioPlayer, isPreview]);
  var handleLoad = function handleLoad(payload) {
    if (isVoiceRecording) {
      return;
    }
    audioPlayer.duration = payload.duration * ONE_SECOND_IN_MILLISECONDS;
  };
  var handleProgress = function handleProgress(data) {
    var currentTime = data.currentTime;
    audioPlayer.position = currentTime * ONE_SECOND_IN_MILLISECONDS;
  };
  var onSeek = function onSeek(seekResponse) {
    audioPlayer.position = seekResponse.currentTime * ONE_SECOND_IN_MILLISECONDS;
  };
  var handlePlayPause = function handlePlayPause() {
    audioPlayer.toggle();
  };
  var handleEnd = function () {
    var _ref = (0, _asyncToGenerator2.default)(function* () {
      yield audioPlayer.stop();
    });
    return function handleEnd() {
      return _ref.apply(this, arguments);
    };
  }();
  var dragStart = function dragStart() {
    audioPlayer.pause();
  };
  var dragProgress = function dragProgress(currentProgress) {
    audioPlayer.progress = currentProgress;
  };
  var dragEnd = function () {
    var _ref2 = (0, _asyncToGenerator2.default)(function* (currentProgress) {
      var positionInSeconds = currentProgress * duration / ONE_SECOND_IN_MILLISECONDS;
      yield audioPlayer.seek(positionInSeconds);
      audioPlayer.play();
    });
    return function dragEnd(_x) {
      return _ref2.apply(this, arguments);
    };
  }();
  var onSpeedChangeHandler = function () {
    var _ref3 = (0, _asyncToGenerator2.default)(function* () {
      yield audioPlayer.changePlaybackRate();
    });
    return function onSpeedChangeHandler() {
      return _ref3.apply(this, arguments);
    };
  }();
  var _useTheme = (0, _contexts.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$audio = _useTheme$theme.audioAttachment,
    container = _useTheme$theme$audio.container,
    leftContainer = _useTheme$theme$audio.leftContainer,
    playPauseButton = _useTheme$theme$audio.playPauseButton,
    progressControlContainer = _useTheme$theme$audio.progressControlContainer,
    progressDurationText = _useTheme$theme$audio.progressDurationText,
    rightContainer = _useTheme$theme$audio.rightContainer,
    speedChangeButton = _useTheme$theme$audio.speedChangeButton,
    speedChangeButtonText = _useTheme$theme$audio.speedChangeButtonText,
    _useTheme$theme$color = _useTheme$theme.colors,
    accent_blue = _useTheme$theme$color.accent_blue,
    black = _useTheme$theme$color.black,
    grey_dark = _useTheme$theme$color.grey_dark,
    grey_whisper = _useTheme$theme$color.grey_whisper,
    static_black = _useTheme$theme$color.static_black,
    static_white = _useTheme$theme$color.static_white,
    white = _useTheme$theme$color.white,
    filenameText = _useTheme$theme.messageInput.fileAttachmentUploadPreview.filenameText;
  var progressDuration = (0, _react.useMemo)(function () {
    return position ? position / ONE_HOUR_IN_MILLISECONDS >= 1 ? _dayjs.default.duration(position, 'milliseconds').format('HH:mm:ss') : _dayjs.default.duration(position, 'milliseconds').format('mm:ss') : _dayjs.default.duration(duration, 'milliseconds').format('mm:ss');
  }, [duration, position]);
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    accessibilityLabel: "audio-attachment-preview",
    style: [styles.container, {
      backgroundColor: white,
      borderColor: grey_whisper
    }, container],
    testID: testID,
    children: [(0, _jsxRuntime.jsx)(_reactNative.View, {
      style: [styles.leftContainer, leftContainer],
      children: (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
        accessibilityLabel: "Play Pause Button",
        onPress: handlePlayPause,
        style: [styles.playPauseButton, {
          backgroundColor: static_white,
          shadowColor: black
        }, playPauseButton],
        children: !isPlaying ? (0, _jsxRuntime.jsx)(_icons.Play, {
          fill: static_black,
          height: 32,
          width: 32
        }) : (0, _jsxRuntime.jsx)(_icons.Pause, {
          fill: static_black,
          height: 32,
          width: 32
        })
      })
    }), (0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: [styles.centerContainer],
      children: [(0, _jsxRuntime.jsx)(_reactNative.Text, {
        accessibilityLabel: "File Name",
        numberOfLines: 1,
        style: [styles.filenameText, {
          color: black
        }, _reactNative.I18nManager.isRTL ? {
          writingDirection: 'rtl'
        } : {
          writingDirection: 'ltr'
        }, filenameText],
        children: (0, _streamChat.isVoiceRecordingAttachment)(item) ? 'Recording' : (0, _getTrimmedAttachmentTitle.getTrimmedAttachmentTitle)(item.title, titleMaxLength)
      }), (0, _jsxRuntime.jsxs)(_reactNative.View, {
        style: styles.audioInfo,
        children: [(0, _jsxRuntime.jsx)(_reactNative.Text, {
          style: [styles.progressDurationText, {
            color: grey_dark
          }, progressDurationText],
          children: progressDuration
        }), !hideProgressBar && (0, _jsxRuntime.jsx)(_reactNative.View, {
          style: [styles.progressControlContainer, progressControlContainer],
          children: item.waveform_data ? (0, _jsxRuntime.jsx)(_WaveProgressBar.WaveProgressBar, {
            amplitudesCount: 30,
            onEndDrag: dragEnd,
            onProgressDrag: dragProgress,
            onStartDrag: dragStart,
            progress: progress,
            waveformData: item.waveform_data
          }) : (0, _jsxRuntime.jsx)(_ProgressControl.ProgressControl, {
            duration: duration,
            filledColor: accent_blue,
            onEndDrag: dragEnd,
            onProgressDrag: dragProgress,
            onStartDrag: dragStart,
            progress: progress,
            testID: "progress-control"
          })
        })]
      }), ((_NativeHandlers$Sound = _native.NativeHandlers.Sound) == null ? void 0 : _NativeHandlers$Sound.Player) && (0, _jsxRuntime.jsx)(_native.NativeHandlers.Sound.Player, {
        onEnd: handleEnd,
        onLoad: handleLoad,
        onProgress: handleProgress,
        onSeek: onSeek,
        paused: !isPlaying,
        rate: currentPlaybackRate,
        soundRef: soundRef,
        testID: "sound-player",
        uri: item.asset_url
      })]
    }), showSpeedSettings ? (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: [styles.rightContainer, rightContainer],
      children: !isPlaying ? (0, _jsxRuntime.jsx)(_icons.Audio, {
        fill: '#ffffff'
      }) : (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
        onPress: onSpeedChangeHandler,
        style: [styles.speedChangeButton, {
          backgroundColor: static_white,
          shadowColor: black
        }, speedChangeButton],
        children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
          style: [styles.speedChangeButtonText, speedChangeButtonText],
          children: `x${currentPlaybackRate.toFixed(1)}`
        })
      })
    }) : null]
  });
};
var styles = _reactNative.StyleSheet.create({
  audioInfo: {
    alignItems: 'center',
    flexDirection: 'row'
  },
  centerContainer: {
    flexGrow: 1
  },
  container: {
    alignItems: 'center',
    borderRadius: 12,
    borderWidth: 1,
    flex: 1,
    flexDirection: 'row',
    paddingLeft: 8,
    paddingRight: 16,
    paddingVertical: 12
  },
  filenameText: {
    fontSize: 14,
    fontWeight: 'bold',
    marginBottom: 8
  },
  leftContainer: {
    marginRight: 8
  },
  playPauseButton: {
    alignItems: 'center',
    borderRadius: 50,
    elevation: 4,
    justifyContent: 'center',
    marginRight: 8,
    padding: 4,
    shadowOffset: {
      height: 2,
      width: 0
    },
    shadowOpacity: 0.23,
    shadowRadius: 2.62
  },
  progressControlContainer: {
    flexGrow: 1,
    justifyContent: 'center'
  },
  progressDurationText: {
    fontSize: 12,
    marginRight: 8
  },
  rightContainer: {
    marginLeft: 16
  },
  speedChangeButton: {
    alignItems: 'center',
    alignSelf: 'center',
    borderRadius: 50,
    elevation: 4,
    justifyContent: 'center',
    paddingHorizontal: 8,
    paddingVertical: 4,
    shadowOffset: {
      height: 2,
      width: 0
    },
    shadowOpacity: 0.23,
    shadowRadius: 2.62
  },
  speedChangeButtonText: {
    fontSize: 12
  }
});
AudioAttachment.displayName = 'AudioAttachment{messageInput{audioAttachment}}';
