// stream-chat-react-native-core/lib/commonjs/components/ImageGallery/components/ImageGalleryFooter.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageGalleryFooterWithContext = exports.ImageGalleryFooter = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _ImageGalleryVideoControl = require("./ImageGalleryVideoControl");
var _ThemeContext = require("../../../contexts/themeContext/ThemeContext");
var _TranslationContext = require("../../../contexts/translationContext/TranslationContext");
var _icons = require("../../../icons");
var _native = require("../../../native");
var _types = require("../../../types/types");
var _SafeAreaViewWrapper = require("../../UIComponents/SafeAreaViewWrapper");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/ImageGallery/components/ImageGalleryFooter.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ReanimatedSafeAreaView = _reactNativeReanimated.default.createAnimatedComponent ? _reactNativeReanimated.default.createAnimatedComponent(_SafeAreaViewWrapper.SafeAreaView) : _SafeAreaViewWrapper.SafeAreaView;
var ImageGalleryFooterWithContext = exports.ImageGalleryFooterWithContext = function ImageGalleryFooterWithContext(props) {
  var accessibilityLabel = props.accessibilityLabel,
    centerElement = props.centerElement,
    duration = props.duration,
    GridIcon = props.GridIcon,
    leftElement = props.leftElement,
    onPlayPause = props.onPlayPause,
    opacity = props.opacity,
    openGridView = props.openGridView,
    paused = props.paused,
    photo = props.photo,
    photoLength = props.photoLength,
    progress = props.progress,
    rightElement = props.rightElement,
    selectedIndex = props.selectedIndex,
    ShareIcon = props.ShareIcon,
    videoControlElement = props.videoControlElement,
    videoRef = props.videoRef,
    visible = props.visible;
  var _useState = (0, _react.useState)(200),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    height = _useState2[0],
    setHeight = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    savingInProgress = _useState4[0],
    setSavingInProgress = _useState4[1];
  var shareIsInProgressRef = (0, _react.useRef)(false);
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    black = _useTheme$theme$color.black,
    white = _useTheme$theme$color.white,
    _useTheme$theme$image = _useTheme$theme.imageGallery.footer,
    centerContainer = _useTheme$theme$image.centerContainer,
    container = _useTheme$theme$image.container,
    imageCountText = _useTheme$theme$image.imageCountText,
    innerContainer = _useTheme$theme$image.innerContainer,
    rightContainer = _useTheme$theme$image.rightContainer;
  var _useTranslationContex = (0, _TranslationContext.useTranslationContext)(),
    t = _useTranslationContex.t;
  var footerStyle = (0, _reactNativeReanimated.useAnimatedStyle)(function () {
    return {
      opacity: opacity.value,
      transform: [{
        translateY: (0, _reactNativeReanimated.interpolate)(visible.value, [0, 1], [height, 0], _reactNativeReanimated.Extrapolation.CLAMP)
      }]
    };
  }, [height]);
  var share = function () {
    var _ref = (0, _asyncToGenerator2.default)(function* () {
      if (shareIsInProgressRef.current) {
        return;
      }
      shareIsInProgressRef.current = true;
      try {
        var _photo$mime_type;
        if (!_native.NativeHandlers.shareImage || !_native.NativeHandlers.deleteFile) {
          return;
        }
        var extension = ((_photo$mime_type = photo.mime_type) == null ? void 0 : _photo$mime_type.split('/')[1]) || 'jpg';
        var shouldDownload = photo.uri && photo.uri.includes('http');
        var localFile;
        if (shouldDownload) {
          var _photo$user;
          setSavingInProgress(true);
          localFile = yield _native.NativeHandlers.saveFile({
            fileName: `${((_photo$user = photo.user) == null ? void 0 : _photo$user.id) || 'ChatPhoto'}-${photo.messageId}-${selectedIndex}.${extension}`,
            fromUrl: photo.uri
          });
          setSavingInProgress(false);
        } else {
          localFile = photo.uri;
        }
        yield _native.NativeHandlers.shareImage({
          type: photo.mime_type || 'image/jpeg',
          url: localFile
        });
        if (shouldDownload) {
          yield _native.NativeHandlers.deleteFile({
            uri: localFile
          });
        }
      } catch (error) {
        setSavingInProgress(false);
        console.log(error);
      }
      shareIsInProgressRef.current = false;
    });
    return function share() {
      return _ref.apply(this, arguments);
    };
  }();
  return (0, _jsxRuntime.jsx)(_reactNativeReanimated.default.View, {
    accessibilityLabel: accessibilityLabel,
    onLayout: function onLayout(event) {
      return setHeight(event.nativeEvent.layout.height);
    },
    pointerEvents: 'box-none',
    style: styles.wrapper,
    children: (0, _jsxRuntime.jsxs)(ReanimatedSafeAreaView, {
      edges: ['bottom'],
      style: [{
        backgroundColor: white
      }, footerStyle, container],
      children: [photo.type === _types.FileTypes.Video ? videoControlElement ? videoControlElement({
        duration: duration,
        onPlayPause: onPlayPause,
        paused: paused,
        progress: progress,
        videoRef: videoRef
      }) : (0, _jsxRuntime.jsx)(_ImageGalleryVideoControl.ImageGalleryVideoControl, {
        duration: duration,
        onPlayPause: onPlayPause,
        paused: paused,
        progress: progress,
        videoRef: videoRef
      }) : null, (0, _jsxRuntime.jsxs)(_reactNative.View, {
        style: [styles.innerContainer, {
          backgroundColor: white
        }, innerContainer],
        children: [leftElement ? leftElement({
          openGridView: openGridView,
          photo: photo,
          share: share,
          shareMenuOpen: savingInProgress
        }) : (0, _jsxRuntime.jsx)(ShareButton, {
          savingInProgress: savingInProgress,
          share: share,
          ShareIcon: ShareIcon
        }), centerElement ? centerElement({
          openGridView: openGridView,
          photo: photo,
          share: share,
          shareMenuOpen: savingInProgress
        }) : (0, _jsxRuntime.jsx)(_reactNative.View, {
          style: [styles.centerContainer, centerContainer],
          children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
            style: [styles.imageCountText, {
              color: black
            }, imageCountText],
            children: t('{{ index }} of {{ photoLength }}', {
              index: selectedIndex + 1,
              photoLength: photoLength
            })
          })
        }), rightElement ? rightElement({
          openGridView: openGridView,
          photo: photo,
          share: share,
          shareMenuOpen: savingInProgress
        }) : (0, _jsxRuntime.jsx)(_reactNative.TouchableOpacity, {
          onPress: openGridView,
          children: (0, _jsxRuntime.jsx)(_reactNative.View, {
            style: [styles.rightContainer, rightContainer],
            children: GridIcon ? GridIcon : (0, _jsxRuntime.jsx)(_icons.Grid, {})
          })
        })]
      })]
    })
  });
};
var ShareButton = function ShareButton(_ref2) {
  var share = _ref2.share,
    ShareIcon = _ref2.ShareIcon,
    savingInProgress = _ref2.savingInProgress;
  var _useTheme2 = (0, _ThemeContext.useTheme)(),
    _useTheme2$theme = _useTheme2.theme,
    black = _useTheme2$theme.colors.black,
    leftContainer = _useTheme2$theme.imageGallery.footer.leftContainer;
  if (!(0, _native.isShareImageAvailable)() || !(0, _native.isFileSystemAvailable)()) {
    return null;
  }
  return (0, _jsxRuntime.jsx)(_reactNative.TouchableOpacity, {
    accessibilityLabel: "Share Button",
    onPress: share,
    children: (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: [styles.leftContainer, leftContainer],
      children: savingInProgress ? (0, _jsxRuntime.jsx)(_reactNative.ActivityIndicator, {
        size: "small"
      }) : ShareIcon ? ShareIcon : (0, _jsxRuntime.jsx)(_icons.Share, {
        pathFill: black
      })
    })
  });
};
var areEqual = function areEqual(prevProps, nextProps) {
  var prevDuration = prevProps.duration,
    prevPaused = prevProps.paused,
    prevProgress = prevProps.progress,
    prevSelectedIndex = prevProps.selectedIndex;
  var nextDuration = nextProps.duration,
    nextPaused = nextProps.paused,
    nextProgress = nextProps.progress,
    nextSelectedIndex = nextProps.selectedIndex;
  var isDurationEqual = prevDuration === nextDuration;
  if (!isDurationEqual) {
    return false;
  }
  var isPausedEqual = prevPaused === nextPaused;
  if (!isPausedEqual) {
    return false;
  }
  var isProgressEqual = prevProgress === nextProgress;
  if (!isProgressEqual) {
    return false;
  }
  var isSelectedIndexEqual = prevSelectedIndex === nextSelectedIndex;
  if (!isSelectedIndexEqual) {
    return false;
  }
  return true;
};
var MemoizedImageGalleryFooter = _react.default.memo(ImageGalleryFooterWithContext, areEqual);
var ImageGalleryFooter = exports.ImageGalleryFooter = function ImageGalleryFooter(props) {
  return (0, _jsxRuntime.jsx)(MemoizedImageGalleryFooter, Object.assign({}, props));
};
ImageGalleryFooter.displayName = 'ImageGalleryFooter{imageGallery{footer}}';
var styles = _reactNative.StyleSheet.create({
  centerContainer: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center'
  },
  imageCountText: {
    fontSize: 16,
    fontWeight: '600'
  },
  innerContainer: {
    flexDirection: 'row',
    paddingVertical: 4
  },
  leftContainer: {
    flex: 1,
    justifyContent: 'center',
    marginLeft: 8
  },
  rightContainer: {
    flex: 1,
    justifyContent: 'center',
    marginRight: 8
  },
  wrapper: {
    bottom: 0,
    left: 0,
    position: 'absolute',
    right: 0
  }
});
