// stream-chat-react-native-core/lib/commonjs/contexts/liveLocationManagerContext/LiveLocationManagerContext.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLiveLocationManagerContext = exports.LiveLocationManagerProvider = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _streamChat = require("stream-chat");
var _ChatContext = require("../chatContext/ChatContext");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/contexts/liveLocationManagerContext/LiveLocationManagerContext.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var LiveLocationManagerContext = (0, _react.createContext)({
  liveLocationManager: null
});
var useLiveLocationManagerContext = exports.useLiveLocationManagerContext = function useLiveLocationManagerContext() {
  return (0, _react.useContext)(LiveLocationManagerContext);
};
var LiveLocationManagerProvider = exports.LiveLocationManagerProvider = function LiveLocationManagerProvider(props) {
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  var watchLocation = props.watchLocation,
    getDeviceId = props.getDeviceId,
    children = props.children;
  var liveLocationManager = (0, _react.useMemo)(function () {
    if (!client) {
      return null;
    }
    return new _streamChat.LiveLocationManager({
      client: client,
      getDeviceId: getDeviceId != null ? getDeviceId : function () {
        return `react-native-${_reactNative.Platform.OS}-${client.userID}`;
      },
      watchLocation: watchLocation
    });
  }, [client, getDeviceId, watchLocation]);
  (0, _react.useEffect)(function () {
    if (!liveLocationManager) {
      return;
    }
    liveLocationManager.init();
    return function () {
      liveLocationManager.unregisterSubscriptions();
    };
  }, [liveLocationManager]);
  return (0, _jsxRuntime.jsx)(LiveLocationManagerContext.Provider, {
    value: {
      liveLocationManager: liveLocationManager
    },
    children: children
  });
};
