// react-native-compressor/lib/commonjs/utils/Uploader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cancelUpload = exports.backgroundUpload = exports.UploaderHttpMethod = exports.UploadType = void 0;
var _reactNative = require("react-native");
var _Main = require("../Main");
var _helpers = require("./helpers");
let UploadType = exports.UploadType = /*#__PURE__*/function (UploadType) {
  UploadType[UploadType["BINARY_CONTENT"] = 0] = "BINARY_CONTENT";
  UploadType[UploadType["MULTIPART"] = 1] = "MULTIPART";
  return UploadType;
}({});
let UploaderHttpMethod = exports.UploaderHttpMethod = /*#__PURE__*/function (UploaderHttpMethod) {
  UploaderHttpMethod["POST"] = "POST";
  UploaderHttpMethod["PUT"] = "PUT";
  UploaderHttpMethod["PATCH"] = "PATCH";
  return UploaderHttpMethod;
}({});
const cancelUpload = (uuid = '', shouldCancelAll = false) => {
  return _Main.Compressor.cancelUpload(uuid, shouldCancelAll);
};
exports.cancelUpload = cancelUpload;
const backgroundUpload = async (url, fileUrl, options, onProgress, abortSignal) => {
  const uuid = (0, _helpers.uuidv4)();
  try {
    if (_reactNative.Platform.OS === 'android' && fileUrl.includes('file://')) {
      fileUrl = fileUrl.replace('file://', '');
    }
    if (options?.getCancellationId) {
      options?.getCancellationId(uuid);
    }
    abortSignal?.addEventListener('abort', () => cancelUpload(uuid));
    const result = await _Main.Compressor.upload(fileUrl, (0, _helpers.toNativeOptions)({
      ...options,
      uuid,
      method: options.httpMethod,
      url
    }), onProgress);
    return result;
  } finally {
    abortSignal?.removeEventListener('abort', () => cancelUpload(uuid));
  }
};
exports.backgroundUpload = backgroundUpload;
