// stream-chat-react-native-core/lib/commonjs/components/MessageInput/ShowThreadMessageInChannelButton.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShowThreadMessageInChannelButtonWithContext = exports.ShowThreadMessageInChannelButton = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _useMessageComposer = require("../../contexts/messageInputContext/hooks/useMessageComposer");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _ThreadContext = require("../../contexts/threadContext/ThreadContext");
var _TranslationContext = require("../../contexts/translationContext/TranslationContext");
var _useStateStore2 = require("../../hooks/useStateStore");
var _icons = require("../../icons");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageInput/ShowThreadMessageInChannelButton.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var stateSelector = function stateSelector(state) {
  return {
    showReplyInChannel: state.showReplyInChannel
  };
};
var ShowThreadMessageInChannelButtonWithContext = exports.ShowThreadMessageInChannelButtonWithContext = function ShowThreadMessageInChannelButtonWithContext(props) {
  var allowThreadMessagesInChannel = props.allowThreadMessagesInChannel,
    t = props.t,
    threadList = props.threadList;
  var messageComposer = (0, _useMessageComposer.useMessageComposer)();
  var _useStateStore = (0, _useStateStore2.useStateStore)(messageComposer.state, stateSelector),
    showReplyInChannel = _useStateStore.showReplyInChannel;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    accent_blue = _useTheme$theme$color.accent_blue,
    grey = _useTheme$theme$color.grey,
    white = _useTheme$theme$color.white,
    _useTheme$theme$messa = _useTheme$theme.messageInput.showThreadMessageInChannelButton,
    check = _useTheme$theme$messa.check,
    checkBoxActive = _useTheme$theme$messa.checkBoxActive,
    checkBoxInactive = _useTheme$theme$messa.checkBoxInactive,
    container = _useTheme$theme$messa.container,
    innerContainer = _useTheme$theme$messa.innerContainer,
    text = _useTheme$theme$messa.text;
  var onPressHandler = (0, _react.useCallback)(function () {
    messageComposer.toggleShowReplyInChannel();
  }, [messageComposer]);
  if (!threadList || !allowThreadMessagesInChannel) {
    return null;
  }
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: [styles.container, container],
    testID: "show-thread-message-in-channel-button",
    children: (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
      onPress: onPressHandler,
      style: function style(_ref) {
        var pressed = _ref.pressed;
        return {
          opacity: pressed ? 0.8 : 1
        };
      },
      children: (0, _jsxRuntime.jsxs)(_reactNative.View, {
        style: [styles.innerContainer, innerContainer],
        children: [(0, _jsxRuntime.jsx)(_reactNative.View, {
          style: [styles.checkBox, showReplyInChannel ? Object.assign({
            backgroundColor: accent_blue,
            borderColor: accent_blue
          }, checkBoxActive) : Object.assign({
            borderColor: grey
          }, checkBoxInactive)],
          children: showReplyInChannel && (0, _jsxRuntime.jsx)(_icons.Check, Object.assign({
            height: 16,
            pathFill: white,
            width: 16
          }, check))
        }), (0, _jsxRuntime.jsx)(_reactNative.Text, {
          style: [styles.text, {
            color: grey
          }, text],
          children: t('Also send to channel')
        })]
      })
    })
  });
};
var areEqual = function areEqual(prevProps, nextProps) {
  var prevAllowThreadMessagesInChannel = prevProps.allowThreadMessagesInChannel,
    prevT = prevProps.t,
    prevThreadList = prevProps.threadList;
  var nextAllowThreadMessagesInChannel = nextProps.allowThreadMessagesInChannel,
    nextT = nextProps.t,
    nextThreadList = nextProps.threadList;
  var tEqual = prevT === nextT;
  if (!tEqual) {
    return false;
  }
  var threadListEqual = prevThreadList === nextThreadList;
  if (!threadListEqual) {
    return false;
  }
  var allowThreadMessagesInChannelEqual = prevAllowThreadMessagesInChannel === nextAllowThreadMessagesInChannel;
  if (!allowThreadMessagesInChannelEqual) {
    return false;
  }
  return true;
};
var MemoizedShowThreadMessageInChannelButton = _react.default.memo(ShowThreadMessageInChannelButtonWithContext, areEqual);
var ShowThreadMessageInChannelButton = exports.ShowThreadMessageInChannelButton = function ShowThreadMessageInChannelButton(props) {
  var _useTranslationContex = (0, _TranslationContext.useTranslationContext)(),
    t = _useTranslationContex.t;
  var _useThreadContext = (0, _ThreadContext.useThreadContext)(),
    allowThreadMessagesInChannel = _useThreadContext.allowThreadMessagesInChannel;
  return (0, _jsxRuntime.jsx)(MemoizedShowThreadMessageInChannelButton, Object.assign({
    allowThreadMessagesInChannel: allowThreadMessagesInChannel,
    t: t
  }, props));
};
ShowThreadMessageInChannelButton.displayName = 'ShowThreadMessageInChannelButton{messageInput{showThreadMessageInChannelButton}}';
var styles = _reactNative.StyleSheet.create({
  checkBox: {
    alignItems: 'center',
    borderRadius: 3,
    borderWidth: 2,
    height: 16,
    justifyContent: 'center',
    width: 16
  },
  container: {
    flexDirection: 'row',
    marginHorizontal: 2,
    marginTop: 8
  },
  innerContainer: {
    flexDirection: 'row'
  },
  text: {
    fontSize: 13,
    marginLeft: 12
  }
});
