// stream-chat-react-native-core/lib/module/index.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  iOS14RefreshGallerySelection: true,
  registerNativeHandlers: true,
  SoundReturnType: true,
  PlaybackStatus: true,
  RecordingStatus: true,
  enTranslations: true,
  esTranslations: true,
  frTranslations: true,
  heTranslations: true,
  hiTranslations: true,
  itTranslations: true,
  jaTranslations: true,
  koTranslations: true,
  nlTranslations: true,
  ptBRTranslations: true,
  ruTranslations: true,
  trTranslations: true,
  SqliteClient: true,
  OfflineDB: true,
  version: true,
  OfflineStoreApis: true
};
Object.defineProperty(exports, "OfflineDB", {
  enumerable: true,
  get: function get() {
    return _OfflineDB.OfflineDB;
  }
});
exports.OfflineStoreApis = void 0;
Object.defineProperty(exports, "PlaybackStatus", {
  enumerable: true,
  get: function get() {
    return _native.PlaybackStatus;
  }
});
Object.defineProperty(exports, "RecordingStatus", {
  enumerable: true,
  get: function get() {
    return _native.RecordingStatus;
  }
});
Object.defineProperty(exports, "SoundReturnType", {
  enumerable: true,
  get: function get() {
    return _native.SoundReturnType;
  }
});
Object.defineProperty(exports, "SqliteClient", {
  enumerable: true,
  get: function get() {
    return _SqliteClient.SqliteClient;
  }
});
Object.defineProperty(exports, "enTranslations", {
  enumerable: true,
  get: function get() {
    return _en.default;
  }
});
Object.defineProperty(exports, "esTranslations", {
  enumerable: true,
  get: function get() {
    return _es.default;
  }
});
Object.defineProperty(exports, "frTranslations", {
  enumerable: true,
  get: function get() {
    return _fr.default;
  }
});
Object.defineProperty(exports, "heTranslations", {
  enumerable: true,
  get: function get() {
    return _he.default;
  }
});
Object.defineProperty(exports, "hiTranslations", {
  enumerable: true,
  get: function get() {
    return _hi.default;
  }
});
exports.iOS14RefreshGallerySelection = void 0;
Object.defineProperty(exports, "itTranslations", {
  enumerable: true,
  get: function get() {
    return _it.default;
  }
});
Object.defineProperty(exports, "jaTranslations", {
  enumerable: true,
  get: function get() {
    return _ja.default;
  }
});
Object.defineProperty(exports, "koTranslations", {
  enumerable: true,
  get: function get() {
    return _ko.default;
  }
});
Object.defineProperty(exports, "nlTranslations", {
  enumerable: true,
  get: function get() {
    return _nl.default;
  }
});
Object.defineProperty(exports, "ptBRTranslations", {
  enumerable: true,
  get: function get() {
    return _ptBr.default;
  }
});
Object.defineProperty(exports, "registerNativeHandlers", {
  enumerable: true,
  get: function get() {
    return _native.registerNativeHandlers;
  }
});
Object.defineProperty(exports, "ruTranslations", {
  enumerable: true,
  get: function get() {
    return _ru.default;
  }
});
Object.defineProperty(exports, "trTranslations", {
  enumerable: true,
  get: function get() {
    return _tr.default;
  }
});
Object.defineProperty(exports, "version", {
  enumerable: true,
  get: function get() {
    return _version.version;
  }
});
require("intl-pluralrules");
require("./polyfills");
var _components = require("./components");
Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _components[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _components[key];
    }
  });
});
var _hooks = require("./hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _hooks[key];
    }
  });
});
var _native = require("./native");
var _contexts = require("./contexts");
Object.keys(_contexts).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _contexts[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _contexts[key];
    }
  });
});
var _icons = require("./icons");
Object.keys(_icons).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _icons[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _icons[key];
    }
  });
});
var _middlewares = require("./middlewares");
Object.keys(_middlewares).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _middlewares[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _middlewares[key];
    }
  });
});
var _types = require("./types/types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});
var _patchMessageTextCommand = require("./utils/patchMessageTextCommand");
Object.keys(_patchMessageTextCommand).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _patchMessageTextCommand[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _patchMessageTextCommand[key];
    }
  });
});
var _Streami18n = require("./utils/i18n/Streami18n");
Object.keys(_Streami18n).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Streami18n[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _Streami18n[key];
    }
  });
});
var _setupCommandUIMiddlewares = require("./utils/setupCommandUIMiddlewares");
Object.keys(_setupCommandUIMiddlewares).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _setupCommandUIMiddlewares[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _setupCommandUIMiddlewares[key];
    }
  });
});
var _utils = require("./utils/utils");
Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _utils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _utils[key];
    }
  });
});
var _en = _interopRequireDefault(require("./i18n/en.json"));
var _es = _interopRequireDefault(require("./i18n/es.json"));
var _fr = _interopRequireDefault(require("./i18n/fr.json"));
var _he = _interopRequireDefault(require("./i18n/he.json"));
var _hi = _interopRequireDefault(require("./i18n/hi.json"));
var _it = _interopRequireDefault(require("./i18n/it.json"));
var _ja = _interopRequireDefault(require("./i18n/ja.json"));
var _ko = _interopRequireDefault(require("./i18n/ko.json"));
var _nl = _interopRequireDefault(require("./i18n/nl.json"));
var _ptBr = _interopRequireDefault(require("./i18n/pt-br.json"));
var _ru = _interopRequireDefault(require("./i18n/ru.json"));
var _tr = _interopRequireDefault(require("./i18n/tr.json"));
var _stateStore = require("./state-store");
Object.keys(_stateStore).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _stateStore[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _stateStore[key];
    }
  });
});
var _SqliteClient = require("./store/SqliteClient");
var _OfflineDB = require("./store/OfflineDB");
var _version = require("./version.json");
var OfflineStoreApis = _interopRequireWildcard(require("./store/apis"));
exports.OfflineStoreApis = OfflineStoreApis;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var iOS14RefreshGallerySelection = exports.iOS14RefreshGallerySelection = _native.NativeHandlers.iOS14RefreshGallerySelection;
