// stream-chat-react-native-core/lib/module/components/Message/MessageSimple/utils/renderText.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderText = exports.MarkdownReactiveScrollView = exports.ListOutput = void 0;
var _toArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toArray"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _objectDestructuringEmpty2 = _interopRequireDefault(require("@babel/runtime/helpers/objectDestructuringEmpty"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _reactNativeMarkdownPackage = _interopRequireDefault(require("react-native-markdown-package"));
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _simpleMarkdown = require("simple-markdown");
var _generateMarkdownText = require("./generateMarkdownText");
var _utils = require("../../../../utils/utils");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Message/MessageSimple/utils/renderText.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MarkdownReactiveScrollView = exports.MarkdownReactiveScrollView = function MarkdownReactiveScrollView(_ref) {
  var children = _ref.children;
  var scrollViewRef = (0, _reactNativeReanimated.useAnimatedRef)();
  var contentWidth = (0, _reactNativeReanimated.useSharedValue)(0);
  var visibleContentWidth = (0, _reactNativeReanimated.useSharedValue)(0);
  var offsetBeforeScroll = (0, _reactNativeReanimated.useSharedValue)(0);
  var panGesture = _reactNativeGestureHandler.Gesture.Pan().activeOffsetX([-5, 5]).onUpdate(function (event) {
    var translationX = event.translationX;
    (0, _reactNativeReanimated.scrollTo)(scrollViewRef, offsetBeforeScroll.value - translationX, 0, false);
  }).onEnd(function (event) {
    var translationX = event.translationX;
    var velocityEffect = event.velocityX * 0.3;
    var finalPosition = (0, _reactNativeReanimated.clamp)(offsetBeforeScroll.value - translationX - velocityEffect, 0, contentWidth.value - visibleContentWidth.value);
    offsetBeforeScroll.value = finalPosition;
    (0, _reactNativeReanimated.scrollTo)(scrollViewRef, finalPosition, 0, true);
  });
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: {
      width: '100%'
    },
    children: (0, _jsxRuntime.jsx)(_reactNativeGestureHandler.GestureDetector, {
      gesture: panGesture,
      children: (0, _jsxRuntime.jsx)(_reactNativeReanimated.default.ScrollView, {
        contentContainerStyle: {
          flexGrow: 1
        },
        horizontal: true,
        nestedScrollEnabled: true,
        onContentSizeChange: function onContentSizeChange(width) {
          contentWidth.value = width;
        },
        onLayout: function onLayout(e) {
          visibleContentWidth.value = e.nativeEvent.layout.width;
        },
        ref: scrollViewRef,
        scrollEnabled: false,
        children: children
      })
    })
  });
};
var defaultMarkdownStyles = {
  codeBlock: {
    fontFamily: _reactNative.Platform.OS === 'ios' ? 'Courier' : 'Monospace',
    fontWeight: '500',
    marginVertical: 8
  },
  inlineCode: {
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5
  },
  list: {
    marginBottom: 8,
    marginTop: 8
  },
  listItemNumber: {
    fontWeight: 'bold'
  },
  listItemText: {
    flex: 0
  },
  listRow: {
    flexDirection: 'row'
  },
  mentions: {
    fontWeight: '700'
  },
  paragraph: {
    marginBottom: 8,
    marginTop: 8
  },
  paragraphCenter: {
    marginBottom: 8,
    marginTop: 8
  },
  paragraphWithImage: {
    marginBottom: 8,
    marginTop: 8
  },
  table: {
    borderRadius: 3,
    borderWidth: 1,
    flex: 1,
    flexDirection: 'row'
  },
  tableHeader: {
    flexDirection: 'row',
    justifyContent: 'space-around'
  },
  tableHeaderCell: {
    fontWeight: '500'
  },
  tableRow: {
    alignItems: 'center',
    justifyContent: 'space-around'
  },
  tableRowCell: {
    flex: 1
  }
};
var mentionsParseFunction = function mentionsParseFunction(capture, parse, state) {
  return {
    content: (0, _simpleMarkdown.parseInline)(parse, capture[0], state)
  };
};
var renderText = exports.renderText = function renderText(params) {
  var colors = params.colors,
    markdownRules = params.markdownRules,
    markdownStyles = params.markdownStyles,
    message = params.message,
    messageOverlay = params.messageOverlay,
    messageTextNumberOfLines = params.messageTextNumberOfLines,
    onLinkParams = params.onLink,
    onLongPressParam = params.onLongPress,
    onlyEmojis = params.onlyEmojis,
    onPressParam = params.onPress,
    preventPress = params.preventPress;
  var text = message.text;
  var markdownText = (0, _generateMarkdownText.generateMarkdownText)(text);
  var styles = Object.assign({}, defaultMarkdownStyles, markdownStyles, {
    autolink: Object.assign({}, defaultMarkdownStyles.autolink, {
      color: colors.accent_blue
    }, markdownStyles == null ? void 0 : markdownStyles.autolink),
    blockQuoteSection: Object.assign({}, defaultMarkdownStyles.blockQuoteSection, {
      flexDirection: 'row',
      padding: 8
    }, markdownStyles == null ? void 0 : markdownStyles.blockQuoteSection),
    blockQuoteSectionBar: Object.assign({}, defaultMarkdownStyles.blockQuoteSectionBar, {
      backgroundColor: colors.grey_gainsboro,
      marginRight: 8,
      width: 2
    }, markdownStyles == null ? void 0 : markdownStyles.blockQuoteSectionBar),
    blockQuoteText: Object.assign({}, defaultMarkdownStyles.blockQuoteText, markdownStyles == null ? void 0 : markdownStyles.blockQuoteText),
    codeBlock: Object.assign({}, defaultMarkdownStyles.codeBlock, {
      backgroundColor: colors.code_block,
      color: colors.black,
      padding: 8
    }, markdownStyles == null ? void 0 : markdownStyles.codeBlock),
    inlineCode: Object.assign({}, defaultMarkdownStyles.inlineCode, {
      backgroundColor: colors.white_smoke,
      borderColor: colors.grey_gainsboro,
      color: colors.accent_red
    }, markdownStyles == null ? void 0 : markdownStyles.inlineCode),
    mentions: Object.assign({}, defaultMarkdownStyles.mentions, {
      color: colors.accent_blue
    }, markdownStyles == null ? void 0 : markdownStyles.mentions),
    table: Object.assign({}, defaultMarkdownStyles.table, {
      borderColor: colors.grey_dark,
      marginVertical: 8
    }, markdownStyles == null ? void 0 : markdownStyles.table),
    tableHeader: Object.assign({}, defaultMarkdownStyles.tableHeader, {
      backgroundColor: colors.grey
    }, markdownStyles == null ? void 0 : markdownStyles.tableHeader),
    tableHeaderCell: Object.assign({}, defaultMarkdownStyles.tableHeaderCell, {
      padding: 5
    }, markdownStyles == null ? void 0 : markdownStyles.tableHeaderCell),
    tableRow: Object.assign({}, defaultMarkdownStyles.tableRow, markdownStyles == null ? void 0 : markdownStyles.tableRow),
    tableRowCell: Object.assign({}, defaultMarkdownStyles.tableRowCell, {
      borderColor: colors.grey_dark,
      padding: 5
    }, markdownStyles == null ? void 0 : markdownStyles.tableRowCell),
    tableRowLast: Object.assign({}, markdownStyles == null ? void 0 : markdownStyles.tableRowLast),
    text: Object.assign({}, defaultMarkdownStyles.text, {
      color: colors.black
    }, markdownStyles == null ? void 0 : markdownStyles.text)
  });
  var onLink = function onLink(url) {
    return onLinkParams ? onLinkParams(url) : _reactNative.Linking.canOpenURL(url).then(function (canOpenUrl) {
      return canOpenUrl && _reactNative.Linking.openURL(url);
    });
  };
  var previousLink;
  var linkReact = function linkReact(node, output, _ref2) {
    var state = Object.assign({}, ((0, _objectDestructuringEmpty2.default)(_ref2), _ref2));
    var url;
    if (state != null && state.withinLink && previousLink) {
      url = previousLink;
    } else {
      url = node.target;
      previousLink = node.target;
    }
    var onPress = function onPress(event) {
      if (!preventPress && onPressParam) {
        onPressParam({
          additionalInfo: {
            url: url
          },
          defaultHandler: function defaultHandler() {
            onLink(url);
          },
          emitter: 'textLink',
          event: event
        });
      }
    };
    var onLongPress = function onLongPress(event) {
      if (!preventPress && onLongPressParam) {
        onLongPressParam({
          additionalInfo: {
            url: url
          },
          emitter: 'textLink',
          event: event
        });
      }
    };
    return (0, _jsxRuntime.jsx)(_reactNative.Text, {
      onLongPress: onLongPress,
      onPress: onPress,
      style: styles.autolink,
      suppressHighlighting: true,
      children: output(node.content, Object.assign({}, state, {
        withinLink: true
      }))
    }, state.key);
  };
  var paragraphTextReact = function paragraphTextReact(node, output, _ref3) {
    var state = Object.assign({}, ((0, _objectDestructuringEmpty2.default)(_ref3), _ref3));
    if (messageTextNumberOfLines !== undefined) {
      if (state.key === '0' || state.key === 0) {
        return (0, _jsxRuntime.jsx)(_reactNative.Text, {
          numberOfLines: messageTextNumberOfLines,
          style: styles.paragraph,
          children: output(node.content, state)
        }, state.key);
      } else {
        return null;
      }
    }
    return (0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: styles.paragraph,
      children: output(node.content, state)
    }, state.key);
  };
  var mentioned_users = message.mentioned_users;
  var mentionedUsernames = (mentioned_users || []).map(function (user) {
    return user.name || user.id;
  }).filter(Boolean).sort(function (a, b) {
    return b.length - a.length;
  }).map(_utils.escapeRegExp);
  var mentionedUsers = mentionedUsernames.map(function (username) {
    return `@${username}`;
  }).join('|');
  var regEx = new RegExp(`^\\B(${mentionedUsers})`, 'g');
  var mentionsMatchFunction = function mentionsMatchFunction(source) {
    return regEx.exec(source);
  };
  var mentionsReact = function mentionsReact(node, output, _ref4) {
    var _node$content$;
    var state = Object.assign({}, ((0, _objectDestructuringEmpty2.default)(_ref4), _ref4));
    var userName = (_node$content$ = node.content[0]) == null || (_node$content$ = _node$content$.content) == null ? void 0 : _node$content$.substring(1);
    var onPress = function onPress(event) {
      if (!preventPress && onPressParam) {
        onPressParam({
          additionalInfo: {
            user: mentioned_users == null ? void 0 : mentioned_users.find(function (user) {
              return userName === user.name;
            })
          },
          emitter: 'textMention',
          event: event
        });
      }
    };
    var onLongPress = function onLongPress(event) {
      if (!preventPress && onLongPressParam) {
        onLongPressParam({
          emitter: 'textMention',
          event: event
        });
      }
    };
    return (0, _jsxRuntime.jsx)(_reactNative.Text, {
      onLongPress: onLongPress,
      onPress: onPress,
      style: styles.mentions,
      children: Array.isArray(node.content) ? node.content.reduce(function (acc, current) {
        return acc + current.content;
      }, '') || '' : output(node.content, state)
    }, state.key);
  };
  var listReact = function listReact(node, output, state) {
    return (0, _jsxRuntime.jsx)(ListOutput, {
      node: node,
      output: output,
      state: state,
      styles: styles
    }, `list-${state.key}`);
  };
  var codeBlockReact = function codeBlockReact(node, _, state) {
    var _node$content;
    return (0, _jsxRuntime.jsx)(MarkdownReactiveScrollView, {
      children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: styles.codeBlock,
        children: node == null || (_node$content = node.content) == null ? void 0 : _node$content.trim()
      })
    }, state.key);
  };
  var tableReact = function tableReact(node, output, state) {
    return (0, _jsxRuntime.jsx)(MarkdownReactiveScrollView, {
      children: (0, _jsxRuntime.jsx)(MarkdownTable, {
        node: node,
        output: output,
        state: state,
        styles: styles
      })
    }, state.key);
  };
  var blockQuoteReact = function blockQuoteReact(node, output, state) {
    return (0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: styles.blockQuoteSection,
      children: [(0, _jsxRuntime.jsx)(_reactNative.View, {
        style: styles.blockQuoteSectionBar
      }), (0, _jsxRuntime.jsx)(_reactNative.View, {
        style: styles.blockQuoteText,
        children: output(node.content, state)
      })]
    }, state.key);
  };
  var customRules = Object.assign({
    blockQuote: {
      react: blockQuoteReact
    },
    codeBlock: {
      react: codeBlockReact
    },
    image: {
      match: function match() {
        return null;
      }
    },
    link: {
      react: linkReact
    },
    list: {
      react: listReact
    },
    paragraph: messageTextNumberOfLines ? {
      react: paragraphTextReact
    } : {},
    reflink: {
      match: function match() {
        return null;
      }
    },
    sublist: {
      react: listReact
    }
  }, mentionedUsers ? {
    mentions: {
      match: mentionsMatchFunction,
      order: _simpleMarkdown.defaultRules.text.order - 0.5,
      parse: mentionsParseFunction,
      react: mentionsReact
    }
  } : {}, {
    table: {
      react: tableReact
    }
  });
  return (0, _jsxRuntime.jsx)(_reactNativeMarkdownPackage.default, {
    onLink: onLink,
    rules: Object.assign({}, customRules, markdownRules),
    styles: styles,
    children: markdownText
  }, `${JSON.stringify(mentioned_users)}-${onlyEmojis}-${messageOverlay ? JSON.stringify(markdownStyles) : undefined}-${JSON.stringify(colors)}`);
};
var ListOutput = exports.ListOutput = function ListOutput(_ref5) {
  var node = _ref5.node,
    output = _ref5.output,
    state = _ref5.state,
    styles = _ref5.styles;
  var isSublist = state.withinList;
  var parentTypes = ['text', 'paragraph', 'strong'];
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: isSublist ? styles == null ? void 0 : styles.sublist : styles == null ? void 0 : styles.list,
    children: node.items.map(function (item, index) {
      var _item$;
      var indexAfterStart = node.start + index;
      if (item === null) {
        return (0, _jsxRuntime.jsx)(ListRow, {
          style: styles == null ? void 0 : styles.listRow,
          testID: "list-item",
          children: (0, _jsxRuntime.jsx)(Bullet, {
            index: node.ordered && indexAfterStart,
            style: node.ordered ? styles == null ? void 0 : styles.listItemNumber : styles == null ? void 0 : styles.listItemBullet
          })
        }, index);
      }
      isSublist = item.length > 1 && item[1].type === 'list';
      var isSublistWithinText = parentTypes.includes(((_item$ = item[0]) != null ? _item$ : {}).type) && isSublist;
      var style = isSublistWithinText ? {
        marginBottom: 0
      } : {};
      return (0, _jsxRuntime.jsxs)(ListRow, {
        style: styles == null ? void 0 : styles.listRow,
        testID: "list-item",
        children: [(0, _jsxRuntime.jsx)(Bullet, {
          index: node.ordered && indexAfterStart,
          style: node.ordered ? styles == null ? void 0 : styles.listItemNumber : styles == null ? void 0 : styles.listItemBullet
        }), (0, _jsxRuntime.jsx)(ListItem, {
          style: [styles == null ? void 0 : styles.listItemText, style],
          children: output(item, state)
        }, 1)]
      }, index);
    })
  }, state.key);
};
var Bullet = function Bullet(_ref6) {
  var index = _ref6.index,
    style = _ref6.style;
  return (0, _jsxRuntime.jsx)(_reactNative.Text, {
    style: style,
    children: index ? `${index}. ` : "\u2022 "
  }, 0);
};
var ListRow = function ListRow(_ref7) {
  var children = _ref7.children,
    style = _ref7.style;
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: style,
    children: children
  });
};
var ListItem = function ListItem(_ref8) {
  var children = _ref8.children,
    style = _ref8.style;
  return (0, _jsxRuntime.jsx)(_reactNative.Text, {
    style: style,
    children: children
  });
};
var transpose = function transpose(matrix) {
  return matrix[0].map(function (_, colIndex) {
    return matrix.map(function (row) {
      return row[colIndex];
    });
  });
};
var MarkdownTable = function MarkdownTable(_ref9) {
  var node = _ref9.node,
    output = _ref9.output,
    state = _ref9.state,
    styles = _ref9.styles;
  var content = (0, _react.useMemo)(function () {
    var _node$cells;
    var nodeContent = [node == null ? void 0 : node.header].concat((0, _toConsumableArray2.default)((_node$cells = node == null ? void 0 : node.cells) != null ? _node$cells : null));
    return transpose(nodeContent);
  }, [node == null ? void 0 : node.cells, node == null ? void 0 : node.header]);
  var columns = content == null ? void 0 : content.map(function (column, idx) {
    return (0, _jsxRuntime.jsx)(MarkdownTableColumn, {
      items: column,
      output: output,
      state: state,
      styles: styles
    }, `column-${idx}`);
  });
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: styles.table,
    children: columns
  }, state.key);
};
var MarkdownTableColumn = function MarkdownTableColumn(_ref10) {
  var items = _ref10.items,
    output = _ref10.output,
    state = _ref10.state,
    styles = _ref10.styles;
  var _items = (0, _toArray2.default)(items),
    headerCellContent = _items[0],
    columnCellContents = _items.slice(1);
  var ColumnCell = (0, _react.useCallback)(function (_ref11) {
    var content = _ref11.content;
    return content ? (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: styles.tableRow,
      children: (0, _jsxRuntime.jsx)(_reactNative.View, {
        style: styles.tableRowCell,
        children: output(content, state)
      })
    }) : null;
  }, [output, state, styles]);
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: {
      flex: 1,
      flexDirection: 'column'
    },
    children: [headerCellContent ? (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: styles.tableHeader,
      children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: styles.tableHeaderCell,
        children: output(headerCellContent, state)
      })
    }, -1) : null, columnCellContents && columnCellContents.map(function (content, idx) {
      return (0, _jsxRuntime.jsx)(ColumnCell, {
        content: content
      }, `cell-${idx}`);
    })]
  });
};
