// stream-chat-react-native-core/lib/module/mock-builders/mock.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUser = exports.getTestClientWithUser = exports.getTestClient = exports.getRandomInt = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _streamChat = require("stream-chat");
var apiKey = 'API_KEY';
var token = 'dummy_token';
var setUser = exports.setUser = function setUser(client, user) {
  return new Promise(function (resolve) {
    client.connectionId = 'dumm_connection_id';
    client.user = user;
    client.user.mutes = [];
    client._user = Object.assign({}, user);
    client.userID = user.id;
    client.userToken = token;
    resolve();
  });
};
function mockClient(client) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var _options$disableAppSe = options.disableAppSettings,
    disableAppSettings = _options$disableAppSe === void 0 ? true : _options$disableAppSe;
  jest.spyOn(client, '_setToken').mockImplementation();
  jest.spyOn(client, '_setupConnection').mockImplementation();
  client.tokenManager = {
    getToken: jest.fn(function () {
      return token;
    }),
    tokenReady: jest.fn(function () {
      return true;
    })
  };
  client.setUser = setUser.bind(null, client);
  if (disableAppSettings) {
    client.getAppSettings = jest.fn(function () {
      return {};
    });
  }
  return client;
}
var getTestClient = exports.getTestClient = function getTestClient() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return mockClient(new _streamChat.StreamChat(apiKey), options);
};
var getTestClientWithUser = exports.getTestClientWithUser = function () {
  var _ref = (0, _asyncToGenerator2.default)(function* (user) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var _options$disableAppSe2 = options.disableAppSettings,
      disableAppSettings = _options$disableAppSe2 === void 0 ? true : _options$disableAppSe2;
    var client = mockClient(new _streamChat.StreamChat(apiKey));
    yield setUser(client, user);
    client.wsPromise = Promise.resolve();
    if (disableAppSettings) {
      client.getAppSettings = jest.fn(function () {
        return {};
      });
    }
    return client;
  });
  return function getTestClientWithUser(_x) {
    return _ref.apply(this, arguments);
  };
}();
var getRandomInt = exports.getRandomInt = function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
};
