// stream-chat-react-native-core/lib/module/components/Poll/components/PollResults/PollVote.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PollVote = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _streamChat = require("stream-chat");
var _contexts = require("../../../../contexts");
var _getDateString = require("../../../../utils/i18n/getDateString");
var _Avatar = require("../../../Avatar/Avatar");
var _usePollState2 = require("../../hooks/usePollState");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Poll/components/PollResults/PollVote.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var PollVote = exports.PollVote = function PollVote(_ref) {
  var _vote$user, _vote$user$name, _vote$user2, _vote$user3;
  var vote = _ref.vote;
  var _useTranslationContex = (0, _contexts.useTranslationContext)(),
    t = _useTranslationContex.t,
    tDateTimeParser = _useTranslationContex.tDateTimeParser;
  var _usePollState = (0, _usePollState2.usePollState)(),
    votingVisibility = _usePollState.votingVisibility;
  var _useTheme = (0, _contexts.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    black = _useTheme$theme$color.black,
    text_low_emphasis = _useTheme$theme$color.text_low_emphasis,
    _useTheme$theme$poll$ = _useTheme$theme.poll.results.vote,
    container = _useTheme$theme$poll$.container,
    dateText = _useTheme$theme$poll$.dateText,
    userName = _useTheme$theme$poll$.userName;
  var dateString = (0, _react.useMemo)(function () {
    return (0, _getDateString.getDateString)({
      date: vote.created_at,
      t: t,
      tDateTimeParser: tDateTimeParser,
      timestampTranslationKey: 'timestamp/PollVote'
    });
  }, [vote.created_at, t, tDateTimeParser]);
  var isAnonymous = (0, _react.useMemo)(function () {
    return votingVisibility === _streamChat.VotingVisibility.anonymous;
  }, [votingVisibility]);
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.voteContainer, container],
    children: [(0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: {
        flexDirection: 'row'
      },
      children: [!isAnonymous && (_vote$user = vote.user) != null && _vote$user.image ? (0, _jsxRuntime.jsx)(_Avatar.Avatar, {
        image: vote.user.image,
        size: 20
      }, vote.id) : null, (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: [styles.voteUserName, {
          color: black
        }, userName],
        children: isAnonymous ? t('Anonymous') : (_vote$user$name = (_vote$user2 = vote.user) == null ? void 0 : _vote$user2.name) != null ? _vote$user$name : (_vote$user3 = vote.user) == null ? void 0 : _vote$user3.id
      })]
    }), (0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: [styles.voteDate, {
        color: text_low_emphasis
      }, dateText],
      children: dateString
    })]
  });
};
var styles = _reactNative.StyleSheet.create({
  voteContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 8,
    paddingVertical: 8
  },
  voteCount: {
    fontSize: 16,
    marginLeft: 16
  },
  voteDate: {
    fontSize: 14
  },
  voteUserName: {
    fontSize: 14,
    marginLeft: 2
  }
});
