// stream-chat-react-native-core/lib/module/contexts/overlayContext/OverlayProvider.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverlayProvider = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeReanimated = require("react-native-reanimated");
var _OverlayContext = require("./OverlayContext");
var _ImageGallery = require("../../components/ImageGallery/ImageGallery");
var _useStreami18n = require("../../hooks/useStreami18n");
var _ImageGalleryContext = require("../imageGalleryContext/ImageGalleryContext");
var _ThemeContext = require("../themeContext/ThemeContext");
var _TranslationContext = require("../translationContext/TranslationContext");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/contexts/overlayContext/OverlayProvider.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var OverlayProvider = exports.OverlayProvider = function OverlayProvider(props) {
  var autoPlayVideo = props.autoPlayVideo,
    children = props.children,
    giphyVersion = props.giphyVersion,
    i18nInstance = props.i18nInstance,
    imageGalleryCustomComponents = props.imageGalleryCustomComponents,
    _props$imageGalleryGr = props.imageGalleryGridHandleHeight,
    imageGalleryGridHandleHeight = _props$imageGalleryGr === void 0 ? 40 : _props$imageGalleryGr,
    imageGalleryGridSnapPoints = props.imageGalleryGridSnapPoints,
    numberOfImageGalleryGridColumns = props.numberOfImageGalleryGridColumns,
    value = props.value;
  var _useState = (0, _react.useState)((value == null ? void 0 : value.overlay) || 'none'),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    overlay = _useState2[0],
    setOverlay = _useState2[1];
  var overlayOpacity = (0, _reactNativeReanimated.useSharedValue)(0);
  var translators = (0, _useStreami18n.useStreami18n)(i18nInstance);
  (0, _react.useEffect)(function () {
    var backAction = function backAction() {
      if (overlay !== 'none') {
        setOverlay('none');
        return true;
      }
      return false;
    };
    var backHandler = _reactNative.BackHandler.addEventListener('hardwareBackPress', backAction);
    return function () {
      return backHandler.remove();
    };
  }, [overlay]);
  (0, _react.useEffect)(function () {
    (0, _reactNativeReanimated.cancelAnimation)(overlayOpacity);
    if (overlay !== 'none') {
      overlayOpacity.value = (0, _reactNativeReanimated.withTiming)(1);
    } else {
      overlayOpacity.value = (0, _reactNativeReanimated.withTiming)(0);
    }
  }, [overlay]);
  var overlayContext = {
    overlay: overlay,
    setOverlay: setOverlay,
    style: value == null ? void 0 : value.style
  };
  return (0, _jsxRuntime.jsx)(_TranslationContext.TranslationProvider, {
    value: Object.assign({}, translators, {
      userLanguage: _TranslationContext.DEFAULT_USER_LANGUAGE
    }),
    children: (0, _jsxRuntime.jsx)(_OverlayContext.OverlayContext.Provider, {
      value: overlayContext,
      children: (0, _jsxRuntime.jsx)(_ImageGalleryContext.ImageGalleryProvider, {
        children: (0, _jsxRuntime.jsxs)(_ThemeContext.ThemeProvider, {
          style: overlayContext.style,
          children: [children, overlay === 'gallery' && (0, _jsxRuntime.jsx)(_ImageGallery.ImageGallery, {
            autoPlayVideo: autoPlayVideo,
            giphyVersion: giphyVersion,
            imageGalleryCustomComponents: imageGalleryCustomComponents,
            imageGalleryGridHandleHeight: imageGalleryGridHandleHeight,
            imageGalleryGridSnapPoints: imageGalleryGridSnapPoints,
            numberOfImageGalleryGridColumns: numberOfImageGalleryGridColumns,
            overlayOpacity: overlayOpacity
          })]
        })
      })
    })
  });
};
