// stream-chat-react-native-core/lib/module/components/MessageInput/AttachButton.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AttachButton = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _NativeAttachmentPicker = require("./components/NativeAttachmentPicker");
var _AttachmentPickerContext = require("../../contexts/attachmentPickerContext/AttachmentPickerContext");
var _MessageInputContext = require("../../contexts/messageInputContext/MessageInputContext");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _Attach = require("../../icons/Attach");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageInput/AttachButton.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var AttachButtonWithContext = function AttachButtonWithContext(props) {
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    showAttachButtonPicker = _useState2[0],
    setShowAttachButtonPicker = _useState2[1];
  var _useState3 = (0, _react.useState)(),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    attachButtonLayoutRectangle = _useState4[0],
    setAttachButtonLayoutRectangle = _useState4[1];
  var disableAttachmentPicker = props.disableAttachmentPicker,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    handleAttachButtonPress = props.handleAttachButtonPress,
    handleOnPress = props.handleOnPress,
    selectedPicker = props.selectedPicker,
    toggleAttachmentPicker = props.toggleAttachmentPicker;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    accent_blue = _useTheme$theme$color.accent_blue,
    grey = _useTheme$theme$color.grey,
    attachButton = _useTheme$theme.messageInput.attachButton;
  var onAttachButtonLayout = function onAttachButtonLayout(event) {
    var layout = event.nativeEvent.layout;
    setAttachButtonLayoutRectangle(function (prev) {
      if (prev && prev.width === layout.width && prev.height === layout.height && prev.x === layout.x && prev.y === layout.y) {
        return prev;
      }
      return layout;
    });
  };
  var attachButtonHandler = function attachButtonHandler() {
    setShowAttachButtonPicker(function (prevShowAttachButtonPicker) {
      return !prevShowAttachButtonPicker;
    });
  };
  var onPressHandler = function onPressHandler() {
    if (disabled) {
      return;
    }
    if (handleOnPress) {
      handleOnPress();
      return;
    }
    if (handleAttachButtonPress) {
      handleAttachButtonPress();
      return;
    }
    if (!disableAttachmentPicker) {
      toggleAttachmentPicker();
    } else {
      attachButtonHandler();
    }
  };
  return (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [(0, _jsxRuntime.jsx)(_reactNative.Pressable, {
      disabled: disabled,
      onLayout: onAttachButtonLayout,
      onPress: onPressHandler,
      style: [attachButton],
      testID: "attach-button",
      children: (0, _jsxRuntime.jsx)(_Attach.Attach, {
        fill: selectedPicker === 'images' ? accent_blue : grey,
        size: 32
      })
    }), showAttachButtonPicker ? (0, _jsxRuntime.jsx)(_NativeAttachmentPicker.NativeAttachmentPicker, {
      attachButtonLayoutRectangle: attachButtonLayoutRectangle,
      onRequestedClose: function onRequestedClose() {
        return setShowAttachButtonPicker(false);
      }
    }) : null]
  });
};
var areEqual = function areEqual(prevProps, nextProps) {
  var prevHandleOnPress = prevProps.handleOnPress,
    prevSelectedPicker = prevProps.selectedPicker;
  var nextHandleOnPress = nextProps.handleOnPress,
    nextSelectedPicker = nextProps.selectedPicker;
  var handleOnPressEqual = prevHandleOnPress === nextHandleOnPress;
  if (!handleOnPressEqual) {
    return false;
  }
  var selectedPickerEqual = prevSelectedPicker === nextSelectedPicker;
  if (!selectedPickerEqual) {
    return false;
  }
  return true;
};
var MemoizedAttachButton = _react.default.memo(AttachButtonWithContext, areEqual);
var AttachButton = exports.AttachButton = function AttachButton(props) {
  var _useAttachmentPickerC = (0, _AttachmentPickerContext.useAttachmentPickerContext)(),
    disableAttachmentPicker = _useAttachmentPickerC.disableAttachmentPicker,
    selectedPicker = _useAttachmentPickerC.selectedPicker;
  var _useMessageInputConte = (0, _MessageInputContext.useMessageInputContext)(),
    handleAttachButtonPress = _useMessageInputConte.handleAttachButtonPress,
    toggleAttachmentPicker = _useMessageInputConte.toggleAttachmentPicker;
  return (0, _jsxRuntime.jsx)(MemoizedAttachButton, Object.assign({
    disableAttachmentPicker: disableAttachmentPicker,
    handleAttachButtonPress: handleAttachButtonPress,
    selectedPicker: selectedPicker,
    toggleAttachmentPicker: toggleAttachmentPicker
  }, props));
};
AttachButton.displayName = 'AttachButton{messageInput}';
