// stream-chat-react-native-core/lib/module/components/MessageInput/StopMessageStreamingButton.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StopMessageStreamingButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _icons = require("../../icons");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageInput/StopMessageStreamingButton.tsx";
var StopMessageStreamingButton = exports.StopMessageStreamingButton = function StopMessageStreamingButton(props) {
  var onPress = props.onPress;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    accent_blue = _useTheme$theme.colors.accent_blue,
    _useTheme$theme$messa = _useTheme$theme.messageInput,
    stopMessageStreamingButton = _useTheme$theme$messa.stopMessageStreamingButton,
    stopMessageStreamingIcon = _useTheme$theme$messa.stopMessageStreamingIcon;
  return (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
    hitSlop: {
      bottom: 15,
      left: 15,
      right: 15,
      top: 15
    },
    onPress: onPress,
    style: [stopMessageStreamingButton],
    testID: "more-options-button",
    children: (0, _jsxRuntime.jsx)(_icons.CircleStop, Object.assign({
      fill: accent_blue,
      size: 32
    }, stopMessageStreamingIcon))
  });
};
StopMessageStreamingButton.displayName = 'StopMessageStreamingButton{messageInput}';
