// stream-chat-react-native-core/lib/commonjs/components/MessageList/MessageList.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageList = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _debounce = _interopRequireDefault(require("lodash/debounce"));
var _useMessageList2 = require("./hooks/useMessageList");
var _useShouldScrollToRecentOnNewOwnMessage = require("./hooks/useShouldScrollToRecentOnNewOwnMessage");
var _InlineLoadingMoreIndicator = require("./InlineLoadingMoreIndicator");
var _InlineLoadingMoreRecentIndicator = require("./InlineLoadingMoreRecentIndicator");
var _InlineLoadingMoreRecentThreadIndicator = require("./InlineLoadingMoreRecentThreadIndicator");
var _AttachmentPickerContext = require("../../contexts/attachmentPickerContext/AttachmentPickerContext");
var _ChannelContext = require("../../contexts/channelContext/ChannelContext");
var _ChatContext = require("../../contexts/chatContext/ChatContext");
var _DebugContext = require("../../contexts/debugContext/DebugContext");
var _ImageGalleryContext = require("../../contexts/imageGalleryContext/ImageGalleryContext");
var _MessageListItemContext = require("../../contexts/messageListItemContext/MessageListItemContext");
var _MessagesContext = require("../../contexts/messagesContext/MessagesContext");
var _OwnCapabilitiesContext = require("../../contexts/ownCapabilitiesContext/OwnCapabilitiesContext");
var _PaginatedMessageListContext = require("../../contexts/paginatedMessageListContext/PaginatedMessageListContext");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _ThreadContext = require("../../contexts/threadContext/ThreadContext");
var _hooks = require("../../hooks");
var _types = require("../../types/types");
var _MessageWrapper = require("../Message/MessageSimple/MessageWrapper");
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["contentContainerStyle", "style"];
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageList/MessageList.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var WAIT_FOR_SCROLL_TIMEOUT = 0;
var MAX_RETRIES_AFTER_SCROLL_FAILURE = 10;
var styles = _reactNative.StyleSheet.create({
  container: {
    alignItems: 'center',
    flex: 1,
    width: '100%'
  },
  contentContainer: {
    paddingBottom: 4
  },
  flex: {
    flex: 1
  },
  listContainer: {
    flex: 1,
    width: '100%'
  },
  stickyHeader: {
    position: 'absolute',
    top: 0
  }
});
var keyExtractor = function keyExtractor(item) {
  if (item.id) {
    return item.id;
  }
  if (item.created_at) {
    return typeof item.created_at === 'string' ? item.created_at : item.created_at.toISOString();
  }
  return Date.now().toString();
};
var flatListViewabilityConfig = {
  viewAreaCoveragePercentThreshold: 1
};
var hasReadLastMessage = function hasReadLastMessage(channel, userId) {
  var _channel$state$latest, _channel$state$read$u;
  var latestMessageIdInChannel = (_channel$state$latest = channel.state.latestMessages.slice(-1)[0]) == null ? void 0 : _channel$state$latest.id;
  var lastReadMessageIdServer = (_channel$state$read$u = channel.state.read[userId]) == null ? void 0 : _channel$state$read$u.last_read_message_id;
  return latestMessageIdInChannel === lastReadMessageIdServer;
};
var getPreviousLastMessage = function getPreviousLastMessage(messages, newMessage) {
  if (!newMessage) return;
  var previousLastMessage;
  for (var i = messages.length - 1; i >= 0; i--) {
    var msg = messages[i];
    if (!(msg != null && msg.id)) break;
    if (msg.id !== newMessage.id) {
      previousLastMessage = msg;
      break;
    }
  }
  return previousLastMessage;
};
var MessageListWithContext = function MessageListWithContext(props) {
  var _client$user4;
  var LoadingMoreRecentIndicator = props.threadList ? _InlineLoadingMoreRecentThreadIndicator.InlineLoadingMoreRecentThreadIndicator : _InlineLoadingMoreRecentIndicator.InlineLoadingMoreRecentIndicator;
  var additionalFlatListProps = props.additionalFlatListProps,
    channel = props.channel,
    channelUnreadStateStore = props.channelUnreadStateStore,
    client = props.client,
    closePicker = props.closePicker,
    DateHeader = props.DateHeader,
    disabled = props.disabled,
    disableTypingIndicator = props.disableTypingIndicator,
    EmptyStateIndicator = props.EmptyStateIndicator,
    FlatList = props.FlatList,
    _props$FooterComponen = props.FooterComponent,
    FooterComponent = _props$FooterComponen === void 0 ? _InlineLoadingMoreIndicator.InlineLoadingMoreIndicator : _props$FooterComponen,
    _props$HeaderComponen = props.HeaderComponent,
    HeaderComponent = _props$HeaderComponen === void 0 ? LoadingMoreRecentIndicator : _props$HeaderComponen,
    hideStickyDateHeader = props.hideStickyDateHeader,
    _props$inverted = props.inverted,
    inverted = _props$inverted === void 0 ? true : _props$inverted,
    _props$isListActive = props.isListActive,
    isListActive = _props$isListActive === void 0 ? false : _props$isListActive,
    _props$isLiveStreamin = props.isLiveStreaming,
    isLiveStreaming = _props$isLiveStreamin === void 0 ? false : _props$isLiveStreamin,
    legacyImageViewerSwipeBehaviour = props.legacyImageViewerSwipeBehaviour,
    loadChannelAroundMessage = props.loadChannelAroundMessage,
    loading = props.loading,
    LoadingIndicator = props.LoadingIndicator,
    loadMore = props.loadMore,
    loadMoreRecent = props.loadMoreRecent,
    loadMoreRecentThread = props.loadMoreRecentThread,
    loadMoreThread = props.loadMoreThread,
    markRead = props.markRead,
    maximumMessageLimit = props.maximumMessageLimit,
    myMessageTheme = props.myMessageTheme,
    NetworkDownIndicator = props.NetworkDownIndicator,
    noGroupByUser = props.noGroupByUser,
    onListScroll = props.onListScroll,
    onThreadSelect = props.onThreadSelect,
    readEvents = props.readEvents,
    reloadChannel = props.reloadChannel,
    ScrollToBottomButton = props.ScrollToBottomButton,
    selectedPicker = props.selectedPicker,
    setChannelUnreadState = props.setChannelUnreadState,
    setFlatListRef = props.setFlatListRef,
    setMessages = props.setMessages,
    setSelectedPicker = props.setSelectedPicker,
    setTargetedMessage = props.setTargetedMessage,
    StickyHeader = props.StickyHeader,
    targetedMessage = props.targetedMessage,
    thread = props.thread,
    threadInstance = props.threadInstance,
    _props$threadList = props.threadList,
    threadList = _props$threadList === void 0 ? false : _props$threadList,
    TypingIndicator = props.TypingIndicator,
    TypingIndicatorContainer = props.TypingIndicatorContainer,
    UnreadMessagesNotification = props.UnreadMessagesNotification,
    hasMore = props.hasMore,
    threadHasMore = props.threadHasMore;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isUnreadNotificationOpen = _useState2[0],
    setIsUnreadNotificationOpen = _useState2[1];
  var _useTheme = (0, _ThemeContext.useTheme)(),
    theme = _useTheme.theme;
  var white_snow = theme.colors.white_snow,
    _theme$messageList = theme.messageList,
    container = _theme$messageList.container,
    contentContainer = _theme$messageList.contentContainer,
    listContainer = _theme$messageList.listContainer;
  var myMessageThemeString = (0, _react.useMemo)(function () {
    return JSON.stringify(myMessageTheme);
  }, [myMessageTheme]);
  var modifiedTheme = (0, _react.useMemo)(function () {
    return (0, _ThemeContext.mergeThemes)({
      style: myMessageTheme,
      theme: theme
    });
  }, [myMessageThemeString, theme]);
  var _useMessageList = (0, _useMessageList2.useMessageList)({
      isLiveStreaming: isLiveStreaming,
      noGroupByUser: noGroupByUser,
      threadList: threadList
    }),
    processedMessageList = _useMessageList.processedMessageList,
    rawMessageList = _useMessageList.rawMessageList,
    viewabilityChangedCallback = _useMessageList.viewabilityChangedCallback;
  var processedMessageListRef = (0, _react.useRef)(processedMessageList);
  processedMessageListRef.current = processedMessageList;
  var renderItem = (0, _react.useCallback)(function (_ref) {
    var message = _ref.item,
      index = _ref.index;
    var previousMessage = processedMessageList[index + 1];
    var nextMessage = processedMessageList[index - 1];
    return (0, _jsxRuntime.jsx)(_MessageWrapper.MessageWrapper, {
      message: message,
      nextMessage: nextMessage,
      previousMessage: previousMessage
    });
  }, [processedMessageList]);
  var messageListLengthBeforeUpdate = (0, _react.useRef)(0);
  var messageListLengthAfterUpdate = processedMessageList.length;
  var topMessageBeforeUpdate = (0, _react.useRef)(undefined);
  var latestNonCurrentMessageBeforeUpdateRef = (0, _react.useRef)(undefined);
  var topMessageAfterUpdate = rawMessageList[0];
  var shouldScrollToRecentOnNewOwnMessageRef = (0, _useShouldScrollToRecentOnNewOwnMessage.useShouldScrollToRecentOnNewOwnMessage)(rawMessageList, client.userID);
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    autoscrollToRecent = _useState4[0],
    setAutoscrollToRecent = _useState4[1];
  var minIndexForVisible = Math.min(1, processedMessageList.length);
  var autoscrollToTopThreshold = autoscrollToRecent ? isLiveStreaming ? 300 : 10 : undefined;
  var maintainVisibleContentPosition = (0, _react.useMemo)(function () {
    return {
      autoscrollToTopThreshold: autoscrollToTopThreshold,
      minIndexForVisible: minIndexForVisible
    };
  }, [autoscrollToTopThreshold, minIndexForVisible]);
  var onStartReachedTracker = (0, _react.useRef)({});
  var onEndReachedTracker = (0, _react.useRef)({});
  var onStartReachedInPromise = (0, _react.useRef)(null);
  var onEndReachedInPromise = (0, _react.useRef)(null);
  var flatListRef = (0, _react.useRef)(null);
  var channelResyncScrollSet = (0, _react.useRef)(true);
  var scrollToDebounceTimeoutRef = (0, _react.useRef)(undefined);
  var onScrollEventTimeoutRef = (0, _react.useRef)(undefined);
  var messageIdLastScrolledToRef = (0, _react.useRef)(undefined);
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    hasMoved = _useState6[0],
    setHasMoved = _useState6[1];
  var _useState7 = (0, _react.useState)(false),
    _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
    scrollToBottomButtonVisible = _useState8[0],
    setScrollToBottomButtonVisible = _useState8[1];
  var _useState9 = (0, _react.useState)(),
    _useState10 = (0, _slicedToArray2.default)(_useState9, 2),
    stickyHeaderDate = _useState10[0],
    setStickyHeaderDate = _useState10[1];
  var stickyHeaderDateRef = (0, _react.useRef)(undefined);
  var channelRef = (0, _react.useRef)(channel);
  channelRef.current = channel;
  var updateStickyHeaderDateIfNeeded = (0, _hooks.useStableCallback)(function (viewableItems) {
    if (!viewableItems.length) {
      return;
    }
    var lastItem = viewableItems[viewableItems.length - 1];
    if (lastItem) {
      var _lastItem$item, _stickyHeaderDateRef$;
      if (!channel.state.messagePagination.hasPrev && processedMessageList[processedMessageList.length - 1].id === lastItem.item.id) {
        setStickyHeaderDate(undefined);
        return;
      }
      var isMessageTypeDeleted = lastItem.item.type === 'deleted';
      if (lastItem != null && (_lastItem$item = lastItem.item) != null && _lastItem$item.created_at && !isMessageTypeDeleted && typeof lastItem.item.created_at !== 'string' && lastItem.item.created_at.toDateString() !== ((_stickyHeaderDateRef$ = stickyHeaderDateRef.current) == null ? void 0 : _stickyHeaderDateRef$.toDateString())) {
        stickyHeaderDateRef.current = lastItem.item.created_at;
        setStickyHeaderDate(lastItem.item.created_at);
      }
    }
  });
  var updateStickyUnreadIndicator = (0, _hooks.useStableCallback)(function (viewableItems) {
    var channelUnreadState = channelUnreadStateStore.channelUnreadState;
    var lastReadMessageId = channelUnreadState == null ? void 0 : channelUnreadState.last_read_message_id;
    var lastReadMessageVisible = viewableItems.some(function (item) {
      return item.item.id === lastReadMessageId;
    });
    if (!viewableItems.length || !readEvents || lastReadMessageVisible || selectedPicker === 'images') {
      setIsUnreadNotificationOpen(false);
      return;
    }
    var lastItem = viewableItems[viewableItems.length - 1];
    if (lastItem) {
      var _channelUnreadState$l;
      var lastItemMessage = lastItem.item;
      var lastItemCreatedAt = lastItemMessage.created_at;
      var unreadIndicatorDate = channelUnreadState == null || (_channelUnreadState$l = channelUnreadState.last_read) == null ? void 0 : _channelUnreadState$l.getTime();
      var lastItemDate = lastItemCreatedAt.getTime();
      if (!channel.state.messagePagination.hasPrev && processedMessageList[processedMessageList.length - 1].id === lastItemMessage.id) {
        setIsUnreadNotificationOpen(false);
        return;
      }
      if (viewableItems.length === 1 && channel.countUnread() === 0 && lastItemMessage.user.id === client.userID) {
        setIsUnreadNotificationOpen(false);
        return;
      }
      if (unreadIndicatorDate && lastItemDate > unreadIndicatorDate) {
        setIsUnreadNotificationOpen(true);
      } else {
        setIsUnreadNotificationOpen(false);
      }
    }
  });
  var unstableOnViewableItemsChanged = function unstableOnViewableItemsChanged(_ref2) {
    var viewableItems = _ref2.viewableItems;
    viewabilityChangedCallback({
      inverted: inverted,
      viewableItems: viewableItems
    });
    if (!viewableItems) {
      return;
    }
    if (!hideStickyDateHeader) {
      updateStickyHeaderDateIfNeeded(viewableItems);
    }
    updateStickyUnreadIndicator(viewableItems);
  };
  var onViewableItemsChanged = (0, _react.useRef)(unstableOnViewableItemsChanged);
  onViewableItemsChanged.current = unstableOnViewableItemsChanged;
  var stableOnViewableItemsChanged = (0, _react.useCallback)(function (_ref3) {
    var viewableItems = _ref3.viewableItems;
    onViewableItemsChanged.current({
      viewableItems: viewableItems
    });
  }, []);
  var resetPaginationTrackersRef = (0, _react.useRef)(function () {
    onStartReachedTracker.current = {};
    onEndReachedTracker.current = {};
  });
  (0, _react.useEffect)(function () {
    if (disabled) {
      setScrollToBottomButtonVisible(false);
    }
  }, [disabled]);
  (0, _react.useEffect)(function () {
    var shouldMarkRead = function shouldMarkRead() {
      var _client$user, _client$user2;
      var channelUnreadState = channelUnreadStateStore.channelUnreadState;
      return !(channelUnreadState != null && channelUnreadState.first_unread_message_id) && !scrollToBottomButtonVisible && ((_client$user = client.user) == null ? void 0 : _client$user.id) && !hasReadLastMessage(channel, (_client$user2 = client.user) == null ? void 0 : _client$user2.id);
    };
    var handleEvent = function () {
      var _ref4 = (0, _asyncToGenerator2.default)(function* (event) {
        var _event$message, _event$message2, _event$message3, _client$user3;
        var mainChannelUpdated = !((_event$message = event.message) != null && _event$message.parent_id) || ((_event$message2 = event.message) == null ? void 0 : _event$message2.show_in_channel);
        var isMyOwnMessage = ((_event$message3 = event.message) == null || (_event$message3 = _event$message3.user) == null ? void 0 : _event$message3.id) === ((_client$user3 = client.user) == null ? void 0 : _client$user3.id);
        var channelUnreadState = channelUnreadStateStore.channelUnreadState;
        if ((scrollToBottomButtonVisible || channelUnreadState != null && channelUnreadState.first_unread_message_id) && !isMyOwnMessage) {
          var _channelUnreadState$u, _channelUnreadState$l2;
          var previousUnreadCount = (_channelUnreadState$u = channelUnreadState == null ? void 0 : channelUnreadState.unread_messages) != null ? _channelUnreadState$u : 0;
          var previousLastMessage = getPreviousLastMessage(channel.state.messages, event.message);
          setChannelUnreadState(Object.assign({}, channelUnreadState, {
            last_read: (_channelUnreadState$l2 = channelUnreadState == null ? void 0 : channelUnreadState.last_read) != null ? _channelUnreadState$l2 : previousUnreadCount === 0 && previousLastMessage != null && previousLastMessage.created_at ? new Date(previousLastMessage.created_at) : new Date(0),
            unread_messages: previousUnreadCount + 1
          }));
        } else if (mainChannelUpdated && shouldMarkRead()) {
          yield markRead();
        }
      });
      return function handleEvent(_x) {
        return _ref4.apply(this, arguments);
      };
    }();
    var listener = channel.on('message.new', handleEvent);
    return function () {
      listener == null || listener.unsubscribe();
    };
  }, [channel, channelUnreadStateStore, (_client$user4 = client.user) == null ? void 0 : _client$user4.id, markRead, scrollToBottomButtonVisible, setChannelUnreadState, threadList]);
  (0, _react.useEffect)(function () {
    var isMessageRemovedFromMessageList = messageListLengthAfterUpdate < messageListLengthBeforeUpdate.current;
    var scrollToBottomIfNeeded = function scrollToBottomIfNeeded() {
      var _topMessageBeforeUpda;
      if (!client || !channel || rawMessageList.length === 0) {
        return;
      }
      if (isMessageRemovedFromMessageList || (_topMessageBeforeUpda = topMessageBeforeUpdate.current) != null && _topMessageBeforeUpda.created_at && topMessageAfterUpdate != null && topMessageAfterUpdate.created_at && topMessageBeforeUpdate.current.created_at < topMessageAfterUpdate.created_at) {
        channelResyncScrollSet.current = false;
        setScrollToBottomButtonVisible(false);
        resetPaginationTrackersRef.current();
        setTimeout(function () {
          var _flatListRef$current;
          (_flatListRef$current = flatListRef.current) == null || _flatListRef$current.scrollToOffset({
            offset: 0
          });
        }, WAIT_FOR_SCROLL_TIMEOUT);
        setTimeout(function () {
          channelResyncScrollSet.current = true;
          if (channel.countUnread() > 0) {
            markRead();
          }
        }, 500);
      }
    };
    if (threadList || isMessageRemovedFromMessageList) {
      if (maximumMessageLimit) {
        resetPaginationTrackersRef.current();
      } else {
        scrollToBottomIfNeeded();
      }
    }
    messageListLengthBeforeUpdate.current = messageListLengthAfterUpdate;
    topMessageBeforeUpdate.current = topMessageAfterUpdate;
  }, [threadList, messageListLengthAfterUpdate, topMessageAfterUpdate == null ? void 0 : topMessageAfterUpdate.id, maximumMessageLimit]);
  (0, _react.useEffect)(function () {
    if (threadList) {
      setAutoscrollToRecent(true);
      return;
    }
    if (!processedMessageList.length) {
      return;
    }
    var notLatestSet = channel.state.messages !== channel.state.latestMessages;
    if (notLatestSet) {
      latestNonCurrentMessageBeforeUpdateRef.current = channel.state.latestMessages[channel.state.latestMessages.length - 1];
      setAutoscrollToRecent(false);
      setScrollToBottomButtonVisible(true);
      return;
    }
    var latestNonCurrentMessageBeforeUpdate = latestNonCurrentMessageBeforeUpdateRef.current;
    latestNonCurrentMessageBeforeUpdateRef.current = undefined;
    var latestCurrentMessageAfterUpdate = processedMessageList[0];
    if (!latestCurrentMessageAfterUpdate) {
      setAutoscrollToRecent(true);
      return;
    }
    var didMergeMessageSetsWithNoUpdates = (latestNonCurrentMessageBeforeUpdate == null ? void 0 : latestNonCurrentMessageBeforeUpdate.id) === latestCurrentMessageAfterUpdate.id;
    var shouldForceScrollToRecent = !didMergeMessageSetsWithNoUpdates || processedMessageList.length - messageListLengthBeforeUpdate.current > 0;
    if (maximumMessageLimit && shouldForceScrollToRecent || !maximumMessageLimit) {
      setAutoscrollToRecent(shouldForceScrollToRecent);
    }
    if (!didMergeMessageSetsWithNoUpdates) {
      var shouldScrollToRecentOnNewOwnMessage = shouldScrollToRecentOnNewOwnMessageRef.current();
      if (shouldScrollToRecentOnNewOwnMessage) {
        setTimeout(function () {
          var _flatListRef$current2;
          (_flatListRef$current2 = flatListRef.current) == null || _flatListRef$current2.scrollToOffset({
            animated: true,
            offset: 0
          });
        }, WAIT_FOR_SCROLL_TIMEOUT);
      }
    }
  }, [channel, threadList, processedMessageList, shouldScrollToRecentOnNewOwnMessageRef, maximumMessageLimit]);
  var goToMessage = (0, _hooks.useStableCallback)(function () {
    var _ref5 = (0, _asyncToGenerator2.default)(function* (messageId) {
      var indexOfParentInMessageList = processedMessageList.findIndex(function (message) {
        return (message == null ? void 0 : message.id) === messageId;
      });
      try {
        if (indexOfParentInMessageList === -1) {
          yield loadChannelAroundMessage({
            messageId: messageId
          });
          return;
        } else {
          if (!flatListRef.current) {
            return;
          }
          clearTimeout(failScrollTimeoutId.current);
          scrollToIndexFailedRetryCountRef.current = 0;
          messageIdLastScrolledToRef.current = messageId;
          setTargetedMessage(messageId);
          flatListRef.current.scrollToIndex({
            animated: true,
            index: indexOfParentInMessageList,
            viewPosition: 0.5
          });
          return;
        }
      } catch (e) {
        console.warn('Error while scrolling to message', e);
      }
    });
    return function (_x2) {
      return _ref5.apply(this, arguments);
    };
  }());
  (0, _react.useEffect)(function () {
    if (!targetedMessage) {
      return;
    }
    scrollToDebounceTimeoutRef.current = setTimeout((0, _asyncToGenerator2.default)(function* () {
      var indexOfParentInMessageList = processedMessageList.findIndex(function (message) {
        return (message == null ? void 0 : message.id) === targetedMessage;
      });
      if (indexOfParentInMessageList === -1) {
        yield loadChannelAroundMessage({
          messageId: targetedMessage,
          setTargetedMessage: setTargetedMessage
        });
      } else {
        if (!flatListRef.current) {
          return;
        }
        clearTimeout(scrollToDebounceTimeoutRef.current);
        clearTimeout(failScrollTimeoutId.current);
        scrollToIndexFailedRetryCountRef.current = 0;
        flatListRef.current.scrollToIndex({
          animated: true,
          index: indexOfParentInMessageList,
          viewPosition: 0.5
        });
        setTargetedMessage(undefined);
      }
    }), WAIT_FOR_SCROLL_TIMEOUT);
  }, [targetedMessage]);
  var messageListItemContextValue = (0, _react.useMemo)(function () {
    return {
      goToMessage: goToMessage,
      modifiedTheme: modifiedTheme,
      noGroupByUser: noGroupByUser,
      onThreadSelect: onThreadSelect
    };
  }, [goToMessage, modifiedTheme, noGroupByUser, onThreadSelect]);
  var maybeCallOnStartReached = (0, _hooks.useStableCallback)((0, _asyncToGenerator2.default)(function* () {
    if (processedMessageList != null && processedMessageList.length && onStartReachedTracker.current[processedMessageList.length]) {
      return;
    }
    if (processedMessageList != null && processedMessageList.length) {
      onStartReachedTracker.current[processedMessageList.length] = true;
    }
    var callback = function callback() {
      onStartReachedInPromise.current = null;
      return Promise.resolve();
    };
    var onError = function onError() {
      setTimeout(function () {
        onStartReachedTracker.current = {};
      }, 2000);
    };
    if (onEndReachedInPromise.current) {
      yield onEndReachedInPromise.current;
    }
    onStartReachedInPromise.current = (threadList && !!threadInstance && loadMoreRecentThread ? loadMoreRecentThread({}) : loadMoreRecent()).then(callback).catch(onError);
  }));
  var maybeCallOnEndReached = (0, _hooks.useStableCallback)((0, _asyncToGenerator2.default)(function* () {
    var shouldQuery = threadList && threadHasMore || !threadList && hasMore;
    if (processedMessageList != null && processedMessageList.length && onEndReachedTracker.current[processedMessageList.length] || !shouldQuery) {
      return;
    }
    if (processedMessageList != null && processedMessageList.length) {
      onEndReachedTracker.current[processedMessageList.length] = true;
    }
    var callback = function callback() {
      onEndReachedInPromise.current = null;
      return Promise.resolve();
    };
    var onError = function onError() {
      setTimeout(function () {
        onEndReachedTracker.current = {};
      }, 2000);
    };
    if (onStartReachedInPromise.current) {
      yield onStartReachedInPromise.current;
    }
    onEndReachedInPromise.current = (threadList ? loadMoreThread() : loadMore()).then(callback).catch(onError);
  }));
  var onUserScrollEvent = (0, _hooks.useStableCallback)(function (event) {
    var nativeEvent = event.nativeEvent;
    clearTimeout(onScrollEventTimeoutRef.current);
    var offset = nativeEvent.contentOffset.y;
    var visibleLength = nativeEvent.layoutMeasurement.height;
    var contentLength = nativeEvent.contentSize.height;
    if (!channel || !channelResyncScrollSet.current) {
      return;
    }
    var isScrollAtStart = offset < 100;
    var isScrollAtEnd = contentLength - visibleLength - offset < 100;
    if (isScrollAtStart) {
      maybeCallOnStartReached();
    }
    if (isScrollAtEnd) {
      maybeCallOnEndReached();
    }
  });
  var handleScroll = (0, _hooks.useStableCallback)(function (event) {
    var messageListHasMessages = processedMessageList.length > 0;
    var offset = event.nativeEvent.contentOffset.y;
    var isScrollAtBottom = offset <= 150;
    var notLatestSet = channel.state.messages !== channel.state.latestMessages;
    var showScrollToBottomButton = messageListHasMessages && (!threadList && notLatestSet || !isScrollAtBottom);
    setScrollToBottomButtonVisible(showScrollToBottomButton);
    if (onListScroll) {
      onListScroll(event);
    }
  });
  var goToNewMessages = (0, _hooks.useStableCallback)((0, _asyncToGenerator2.default)(function* () {
    var isNotLatestSet = channel.state.messages !== channel.state.latestMessages;
    if (isNotLatestSet) {
      resetPaginationTrackersRef.current();
      yield reloadChannel();
    } else if (flatListRef.current) {
      flatListRef.current.scrollToOffset({
        animated: true,
        offset: 0
      });
    }
    setScrollToBottomButtonVisible(false);
    yield markRead({
      updateChannelUnreadState: false
    });
  }));
  var scrollToIndexFailedRetryCountRef = (0, _react.useRef)(0);
  var failScrollTimeoutId = (0, _react.useRef)(undefined);
  var onScrollToIndexFailedRef = (0, _react.useRef)(function (info) {
    if (!flatListRef.current) {
      return;
    }
    failScrollTimeoutId.current = setTimeout(function () {
      try {
        var _flatListRef$current3;
        (_flatListRef$current3 = flatListRef.current) == null || _flatListRef$current3.scrollToIndex({
          animated: true,
          index: info.index,
          viewPosition: 0.5
        });
        if (messageIdLastScrolledToRef.current) {
          setTargetedMessage(messageIdLastScrolledToRef.current);
        }
        scrollToIndexFailedRetryCountRef.current = 0;
      } catch (e) {
        if (!onScrollToIndexFailedRef.current || scrollToIndexFailedRetryCountRef.current > MAX_RETRIES_AFTER_SCROLL_FAILURE) {
          scrollToIndexFailedRetryCountRef.current = 0;
          return;
        }
        scrollToIndexFailedRetryCountRef.current += 1;
        onScrollToIndexFailedRef.current(info);
      }
    }, WAIT_FOR_SCROLL_TIMEOUT);
  });
  var messagesWithImages = legacyImageViewerSwipeBehaviour && processedMessageList.filter(function (message) {
    var isMessageTypeDeleted = message.type === 'deleted';
    if (!isMessageTypeDeleted && message.attachments) {
      return message.attachments.some(function (attachment) {
        return attachment.type === _types.FileTypes.Image && !attachment.title_link && !attachment.og_scrape_url && (attachment.image_url || attachment.thumb_url);
      });
    }
    return false;
  });
  var imageString = legacyImageViewerSwipeBehaviour && messagesWithImages && messagesWithImages.map(function (message) {
    var _message$attachments;
    return (_message$attachments = message.attachments) == null ? void 0 : _message$attachments.map(function (attachment) {
      return attachment.image_url || attachment.thumb_url || '';
    }).join();
  }).join();
  var numberOfMessagesWithImages = legacyImageViewerSwipeBehaviour && messagesWithImages && messagesWithImages.length;
  var threadExists = !!thread;
  (0, _react.useEffect)(function () {
    if (legacyImageViewerSwipeBehaviour && isListActive && (threadList && thread || !threadList && !thread)) {
      setMessages(messagesWithImages);
    }
  }, [imageString, isListActive, legacyImageViewerSwipeBehaviour, numberOfMessagesWithImages, threadExists, threadList]);
  var dismissImagePicker = (0, _hooks.useStableCallback)(function () {
    if (selectedPicker) {
      setSelectedPicker(undefined);
      closePicker();
    }
  });
  var onScrollBeginDrag = (0, _hooks.useStableCallback)(function (event) {
    !hasMoved && selectedPicker && setHasMoved(true);
    onUserScrollEvent(event);
  });
  var onScrollEndDrag = (0, _hooks.useStableCallback)(function (event) {
    hasMoved && selectedPicker && setHasMoved(false);
    onUserScrollEvent(event);
  });
  var refCallback = (0, _hooks.useStableCallback)(function (ref) {
    flatListRef.current = ref;
    if (setFlatListRef) {
      setFlatListRef(ref);
    }
  });
  var onUnreadNotificationClose = (0, _hooks.useStableCallback)((0, _asyncToGenerator2.default)(function* () {
    yield markRead();
    setIsUnreadNotificationOpen(false);
  }));
  var debugRef = (0, _DebugContext.useDebugContext)();
  var isDebugModeEnabled = __DEV__ && debugRef && debugRef.current;
  if (isDebugModeEnabled) {
    if (debugRef.current.setEventType) {
      debugRef.current.setEventType('send');
    }
    if (debugRef.current.setSendEventParams) {
      debugRef.current.setSendEventParams({
        action: thread ? 'ThreadList' : 'Messages',
        data: processedMessageList
      });
    }
  }
  var additionalFlatListPropsExcludingStyle;
  if (additionalFlatListProps) {
    var contentContainerStyle = additionalFlatListProps.contentContainerStyle,
      style = additionalFlatListProps.style,
      rest = (0, _objectWithoutProperties2.default)(additionalFlatListProps, _excluded);
    additionalFlatListPropsExcludingStyle = rest;
  }
  var flatListStyle = (0, _react.useMemo)(function () {
    return [styles.listContainer, listContainer, additionalFlatListProps == null ? void 0 : additionalFlatListProps.style];
  }, [additionalFlatListProps == null ? void 0 : additionalFlatListProps.style, listContainer]);
  var flatListContentContainerStyle = (0, _react.useMemo)(function () {
    return [styles.contentContainer, additionalFlatListProps == null ? void 0 : additionalFlatListProps.contentContainerStyle, contentContainer];
  }, [additionalFlatListProps == null ? void 0 : additionalFlatListProps.contentContainerStyle, contentContainer]);
  var viewportHeightRef = (0, _react.useRef)(undefined);
  var debouncedPrefillMessages = (0, _react.useMemo)(function () {
    return (0, _debounce.default)(function (viewportHeight, contentHeight) {
      if (viewportHeight >= contentHeight) {
        maybeCallOnEndReached();
      }
    }, 500, {
      leading: false,
      trailing: true
    });
  }, [maybeCallOnEndReached]);
  var onContentSizeChange = (0, _hooks.useStableCallback)(function (width, height) {
    var _viewportHeightRef$cu;
    if (additionalFlatListProps != null && additionalFlatListProps.onContentSizeChange) {
      additionalFlatListProps.onContentSizeChange(width, height);
    }
    debouncedPrefillMessages((_viewportHeightRef$cu = viewportHeightRef.current) != null ? _viewportHeightRef$cu : 0, height);
  });
  var onLayout = (0, _hooks.useStableCallback)(function (event) {
    if (additionalFlatListProps != null && additionalFlatListProps.onLayout) {
      additionalFlatListProps.onLayout(event);
    }
    viewportHeightRef.current = event.nativeEvent.layout.height;
  });
  if (!FlatList) {
    return null;
  }
  if (loading) {
    return (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: [styles.container, {
        backgroundColor: white_snow
      }, container],
      children: (0, _jsxRuntime.jsx)(LoadingIndicator, {
        listType: "message"
      })
    });
  }
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: white_snow
    }, container],
    testID: "message-flat-list-wrapper",
    children: [processedMessageList.length === 0 && !thread ? (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: [styles.flex, {
        backgroundColor: white_snow
      }],
      testID: "empty-state",
      children: EmptyStateIndicator ? (0, _jsxRuntime.jsx)(EmptyStateIndicator, {
        listType: "message"
      }) : null
    }) : (0, _jsxRuntime.jsx)(_MessageListItemContext.MessageListItemProvider, {
      value: messageListItemContextValue,
      children: (0, _jsxRuntime.jsx)(FlatList, Object.assign({
        contentContainerStyle: flatListContentContainerStyle,
        data: processedMessageList,
        extraData: disabled,
        inverted: inverted,
        keyboardShouldPersistTaps: "handled",
        keyExtractor: keyExtractor,
        ListFooterComponent: FooterComponent,
        ListHeaderComponent: HeaderComponent,
        maintainVisibleContentPosition: maintainVisibleContentPosition,
        maxToRenderPerBatch: 30,
        onContentSizeChange: onContentSizeChange,
        onLayout: onLayout,
        onMomentumScrollEnd: onUserScrollEvent,
        onScroll: handleScroll,
        onScrollBeginDrag: onScrollBeginDrag,
        onScrollEndDrag: onScrollEndDrag,
        onScrollToIndexFailed: onScrollToIndexFailedRef.current,
        onTouchEnd: dismissImagePicker,
        onViewableItemsChanged: stableOnViewableItemsChanged,
        ref: refCallback,
        renderItem: renderItem,
        scrollEventThrottle: isLiveStreaming ? 16 : undefined,
        showsVerticalScrollIndicator: false,
        strictMode: isLiveStreaming,
        style: flatListStyle,
        testID: "message-flat-list",
        viewabilityConfig: flatListViewabilityConfig
      }, additionalFlatListPropsExcludingStyle))
    }), (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: styles.stickyHeader,
      children: messageListLengthAfterUpdate && StickyHeader ? (0, _jsxRuntime.jsx)(StickyHeader, {
        date: stickyHeaderDate,
        DateHeader: DateHeader
      }) : null
    }), !disableTypingIndicator && TypingIndicator && (0, _jsxRuntime.jsx)(TypingIndicatorContainer, {
      children: (0, _jsxRuntime.jsx)(TypingIndicator, {})
    }), (0, _jsxRuntime.jsx)(ScrollToBottomButton, {
      onPress: goToNewMessages,
      showNotification: scrollToBottomButtonVisible,
      unreadCount: threadList ? 0 : channel == null ? void 0 : channel.countUnread()
    }), (0, _jsxRuntime.jsx)(NetworkDownIndicator, {}), isUnreadNotificationOpen && !threadList ? (0, _jsxRuntime.jsx)(UnreadMessagesNotification, {
      onCloseHandler: onUnreadNotificationClose
    }) : null]
  });
};
var MessageList = exports.MessageList = function MessageList(props) {
  var _useAttachmentPickerC = (0, _AttachmentPickerContext.useAttachmentPickerContext)(),
    closePicker = _useAttachmentPickerC.closePicker,
    selectedPicker = _useAttachmentPickerC.selectedPicker,
    setSelectedPicker = _useAttachmentPickerC.setSelectedPicker;
  var _useChannelContext = (0, _ChannelContext.useChannelContext)(),
    channel = _useChannelContext.channel,
    channelUnreadState = _useChannelContext.channelUnreadState,
    channelUnreadStateStore = _useChannelContext.channelUnreadStateStore,
    disabled = _useChannelContext.disabled,
    EmptyStateIndicator = _useChannelContext.EmptyStateIndicator,
    enableMessageGroupingByUser = _useChannelContext.enableMessageGroupingByUser,
    error = _useChannelContext.error,
    hideStickyDateHeader = _useChannelContext.hideStickyDateHeader,
    highlightedMessageId = _useChannelContext.highlightedMessageId,
    isChannelActive = _useChannelContext.isChannelActive,
    loadChannelAroundMessage = _useChannelContext.loadChannelAroundMessage,
    loading = _useChannelContext.loading,
    LoadingIndicator = _useChannelContext.LoadingIndicator,
    maximumMessageLimit = _useChannelContext.maximumMessageLimit,
    markRead = _useChannelContext.markRead,
    NetworkDownIndicator = _useChannelContext.NetworkDownIndicator,
    reloadChannel = _useChannelContext.reloadChannel,
    scrollToFirstUnreadThreshold = _useChannelContext.scrollToFirstUnreadThreshold,
    setChannelUnreadState = _useChannelContext.setChannelUnreadState,
    setTargetedMessage = _useChannelContext.setTargetedMessage,
    StickyHeader = _useChannelContext.StickyHeader,
    targetedMessage = _useChannelContext.targetedMessage,
    threadList = _useChannelContext.threadList;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  var _useImageGalleryConte = (0, _ImageGalleryContext.useImageGalleryContext)(),
    setMessages = _useImageGalleryConte.setMessages;
  var _useOwnCapabilitiesCo = (0, _OwnCapabilitiesContext.useOwnCapabilitiesContext)(),
    readEvents = _useOwnCapabilitiesCo.readEvents;
  var _useMessagesContext = (0, _MessagesContext.useMessagesContext)(),
    DateHeader = _useMessagesContext.DateHeader,
    disableTypingIndicator = _useMessagesContext.disableTypingIndicator,
    FlatList = _useMessagesContext.FlatList,
    InlineDateSeparator = _useMessagesContext.InlineDateSeparator,
    InlineUnreadIndicator = _useMessagesContext.InlineUnreadIndicator,
    legacyImageViewerSwipeBehaviour = _useMessagesContext.legacyImageViewerSwipeBehaviour,
    Message = _useMessagesContext.Message,
    MessageSystem = _useMessagesContext.MessageSystem,
    myMessageTheme = _useMessagesContext.myMessageTheme,
    ScrollToBottomButton = _useMessagesContext.ScrollToBottomButton,
    shouldShowUnreadUnderlay = _useMessagesContext.shouldShowUnreadUnderlay,
    TypingIndicator = _useMessagesContext.TypingIndicator,
    TypingIndicatorContainer = _useMessagesContext.TypingIndicatorContainer,
    UnreadMessagesNotification = _useMessagesContext.UnreadMessagesNotification;
  var _usePaginatedMessageL = (0, _PaginatedMessageListContext.usePaginatedMessageListContext)(),
    loadMore = _usePaginatedMessageL.loadMore,
    loadMoreRecent = _usePaginatedMessageL.loadMoreRecent,
    hasMore = _usePaginatedMessageL.hasMore;
  var _useThreadContext = (0, _ThreadContext.useThreadContext)(),
    loadMoreRecentThread = _useThreadContext.loadMoreRecentThread,
    loadMoreThread = _useThreadContext.loadMoreThread,
    threadHasMore = _useThreadContext.threadHasMore,
    thread = _useThreadContext.thread,
    threadInstance = _useThreadContext.threadInstance;
  return (0, _jsxRuntime.jsx)(MessageListWithContext, Object.assign({
    channel: channel,
    channelUnreadState: channelUnreadState,
    channelUnreadStateStore: channelUnreadStateStore,
    client: client,
    closePicker: closePicker,
    DateHeader: DateHeader,
    disabled: disabled,
    disableTypingIndicator: disableTypingIndicator,
    EmptyStateIndicator: EmptyStateIndicator,
    enableMessageGroupingByUser: enableMessageGroupingByUser,
    error: error,
    FlatList: FlatList,
    hasMore: hasMore,
    hideStickyDateHeader: hideStickyDateHeader,
    highlightedMessageId: highlightedMessageId,
    InlineDateSeparator: InlineDateSeparator,
    InlineUnreadIndicator: InlineUnreadIndicator,
    isListActive: isChannelActive,
    legacyImageViewerSwipeBehaviour: legacyImageViewerSwipeBehaviour,
    loadChannelAroundMessage: loadChannelAroundMessage,
    loading: loading,
    LoadingIndicator: LoadingIndicator,
    loadMore: loadMore,
    loadMoreRecent: loadMoreRecent,
    loadMoreRecentThread: loadMoreRecentThread,
    loadMoreThread: loadMoreThread,
    markRead: markRead,
    maximumMessageLimit: maximumMessageLimit,
    Message: Message,
    MessageSystem: MessageSystem,
    myMessageTheme: myMessageTheme,
    NetworkDownIndicator: NetworkDownIndicator,
    readEvents: readEvents,
    reloadChannel: reloadChannel,
    ScrollToBottomButton: ScrollToBottomButton,
    scrollToFirstUnreadThreshold: scrollToFirstUnreadThreshold,
    selectedPicker: selectedPicker,
    setChannelUnreadState: setChannelUnreadState,
    setMessages: setMessages,
    setSelectedPicker: setSelectedPicker,
    setTargetedMessage: setTargetedMessage,
    shouldShowUnreadUnderlay: shouldShowUnreadUnderlay,
    StickyHeader: StickyHeader,
    targetedMessage: targetedMessage,
    thread: thread,
    threadHasMore: threadHasMore,
    threadInstance: threadInstance,
    threadList: threadList,
    TypingIndicator: TypingIndicator,
    TypingIndicatorContainer: TypingIndicatorContainer,
    UnreadMessagesNotification: UnreadMessagesNotification
  }, props, {
    noGroupByUser: !enableMessageGroupingByUser || props.noGroupByUser
  }));
};
