// stream-chat-react-native-core/lib/commonjs/components/Message/MessageSimple/MessageBubble.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwipableMessageBubble = exports.MessageBubble = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _contexts = require("../../../contexts");
var _hooks = require("../../../hooks");
var _native = require("../../../native");
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["MessageSwipeContent", "messageSwipeToReplyHitSlop", "onSwipe"];
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Message/MessageSimple/MessageBubble.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MessageBubble = exports.MessageBubble = _react.default.memo(function (_ref) {
  var reactionListPosition = _ref.reactionListPosition,
    messageContentWidth = _ref.messageContentWidth,
    setMessageContentWidth = _ref.setMessageContentWidth,
    MessageContent = _ref.MessageContent,
    ReactionListTop = _ref.ReactionListTop,
    backgroundColor = _ref.backgroundColor,
    isVeryLastMessage = _ref.isVeryLastMessage,
    messageGroupedSingleOrBottom = _ref.messageGroupedSingleOrBottom,
    noBorder = _ref.noBorder;
  var _useTheme = (0, _contexts.useTheme)(),
    contentWrapper = _useTheme.theme.messageSimple.contentWrapper;
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.contentWrapper, contentWrapper],
    children: [(0, _jsxRuntime.jsx)(MessageContent, {
      backgroundColor: backgroundColor,
      isVeryLastMessage: isVeryLastMessage,
      messageGroupedSingleOrBottom: messageGroupedSingleOrBottom,
      noBorder: noBorder,
      setMessageContentWidth: setMessageContentWidth
    }), reactionListPosition === 'top' && ReactionListTop ? (0, _jsxRuntime.jsx)(ReactionListTop, {
      messageContentWidth: messageContentWidth
    }) : null]
  });
});
var AnimatedWrapper = _reactNativeReanimated.default.createAnimatedComponent(_reactNative.View);
var SwipableMessageBubble = exports.SwipableMessageBubble = _react.default.memo(function (props) {
  var MessageSwipeContent = props.MessageSwipeContent,
    messageSwipeToReplyHitSlop = props.messageSwipeToReplyHitSlop,
    onSwipe = props.onSwipe,
    messageBubbleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var _useTheme2 = (0, _contexts.useTheme)(),
    _useTheme2$theme$mess = _useTheme2.theme.messageSimple,
    contentWrapper = _useTheme2$theme$mess.contentWrapper,
    swipeContentContainer = _useTheme2$theme$mess.swipeContentContainer;
  var translateX = (0, _reactNativeReanimated.useSharedValue)(0);
  var touchStart = (0, _reactNativeReanimated.useSharedValue)(null);
  var isSwiping = (0, _reactNativeReanimated.useSharedValue)(false);
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    shouldRenderAnimatedWrapper = _useState2[0],
    setShouldRenderAnimatedWrapper = _useState2[1];
  var SWIPABLE_THRESHOLD = 25;
  var MINIMUM_DISTANCE = 8;
  var triggerHaptic = _native.NativeHandlers.triggerHaptic;
  var setMessageContentWidth = (0, _hooks.useStableCallback)(function (valueOrCallback) {
    if (typeof valueOrCallback === 'number') {
      props.setMessageContentWidth(Math.ceil(valueOrCallback));
      return;
    }
    props.setMessageContentWidth(valueOrCallback);
  });
  var swipeGesture = (0, _react.useMemo)(function () {
    return _reactNativeGestureHandler.Gesture.Pan().hitSlop(messageSwipeToReplyHitSlop).onBegin(function (event) {
      touchStart.value = {
        x: event.x,
        y: event.y
      };
    }).onTouchesMove(function (event, state) {
      if (!touchStart.value || !event.changedTouches.length) {
        state.fail();
        return;
      }
      var xDiff = Math.abs(event.changedTouches[0].x - touchStart.value.x);
      var yDiff = Math.abs(event.changedTouches[0].y - touchStart.value.y);
      var isHorizontalPanning = xDiff > yDiff;
      var hasMinimumDistance = xDiff > MINIMUM_DISTANCE || yDiff > MINIMUM_DISTANCE;
      if (isHorizontalPanning && hasMinimumDistance) {
        state.activate();
        if (!isSwiping.value) {
          (0, _reactNativeReanimated.runOnJS)(setShouldRenderAnimatedWrapper)(true);
        }
        isSwiping.value = true;
      } else if (hasMinimumDistance) {
        state.fail();
      }
    }).onStart(function () {
      translateX.value = 0;
    }).onChange(function (_ref2) {
      var translationX = _ref2.translationX;
      if (translationX > 0) {
        translateX.value = translationX;
      }
    }).onEnd(function () {
      if (translateX.value >= SWIPABLE_THRESHOLD) {
        (0, _reactNativeReanimated.runOnJS)(onSwipe)();
        if (triggerHaptic) {
          (0, _reactNativeReanimated.runOnJS)(triggerHaptic)('impactMedium');
        }
      }
      isSwiping.value = false;
      translateX.value = (0, _reactNativeReanimated.withSpring)(0, {
        dampingRatio: 1,
        duration: 500,
        overshootClamping: true,
        stiffness: 1
      }, function () {
        (0, _reactNativeReanimated.runOnJS)(setShouldRenderAnimatedWrapper)(false);
      });
    });
  }, [messageSwipeToReplyHitSlop, touchStart, isSwiping, translateX, onSwipe, triggerHaptic]);
  var swipeContentAnimatedStyle = (0, _reactNativeReanimated.useAnimatedStyle)(function () {
    return {
      opacity: (0, _reactNativeReanimated.interpolate)(translateX.value, [0, SWIPABLE_THRESHOLD], [0, 1]),
      width: translateX.value
    };
  }, []);
  return (0, _jsxRuntime.jsx)(_reactNativeGestureHandler.GestureDetector, {
    gesture: swipeGesture,
    children: (0, _jsxRuntime.jsxs)(_reactNative.View, {
      hitSlop: messageSwipeToReplyHitSlop,
      style: [styles.contentWrapper, contentWrapper, props.messageContentWidth > 0 && shouldRenderAnimatedWrapper ? {
        width: props.messageContentWidth
      } : {}],
      children: [shouldRenderAnimatedWrapper ? (0, _jsxRuntime.jsx)(AnimatedWrapper, {
        style: [styles.swipeContentContainer, swipeContentAnimatedStyle, swipeContentContainer],
        children: MessageSwipeContent ? (0, _jsxRuntime.jsx)(MessageSwipeContent, {}) : null
      }) : null, (0, _jsxRuntime.jsx)(MessageBubble, Object.assign({}, messageBubbleProps, {
        setMessageContentWidth: setMessageContentWidth
      }))]
    })
  });
});
var styles = _reactNative.StyleSheet.create({
  contentWrapper: {
    alignItems: 'center',
    flexDirection: 'row'
  },
  swipeContentContainer: {
    flexShrink: 0,
    overflow: 'hidden',
    position: 'relative'
  }
});
