// stream-chat-react-native-core/lib/commonjs/components/Message/MessageSimple/MessageSwipeContent.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageSwipeContent = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _ThemeContext = require("../../../contexts/themeContext/ThemeContext");
var _icons = require("../../../icons");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Message/MessageSimple/MessageSwipeContent.tsx";
var MessageSwipeContent = exports.MessageSwipeContent = function MessageSwipeContent() {
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    grey = _useTheme$theme.colors.grey,
    container = _useTheme$theme.messageSimple.swipeLeftContent.container;
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: [styles.container, container],
    children: (0, _jsxRuntime.jsx)(_icons.CurveLineLeftUp, {
      pathFill: grey
    })
  });
};
var styles = _reactNative.StyleSheet.create({
  container: {
    justifyContent: 'center',
    paddingRight: 16
  }
});
