// stream-chat-react-native-core/lib/commonjs/components/Poll/components/Button.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericPollButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _contexts = require("../../../contexts");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Poll/components/Button.tsx";
var GenericPollButton = exports.GenericPollButton = function GenericPollButton(_ref) {
  var onPress = _ref.onPress,
    title = _ref.title;
  var _useTheme = (0, _contexts.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    accent_dark_blue = _useTheme$theme.colors.accent_dark_blue,
    _useTheme$theme$poll$ = _useTheme$theme.poll.button,
    container = _useTheme$theme$poll$.container,
    text = _useTheme$theme$poll$.text;
  return (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
    onPress: onPress,
    style: function style(_ref2) {
      var pressed = _ref2.pressed;
      return [{
        opacity: pressed ? 0.5 : 1
      }, styles.container, container];
    },
    children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: [styles.text, {
        color: accent_dark_blue
      }, text],
      children: title
    })
  });
};
var styles = _reactNative.StyleSheet.create({
  container: {
    alignItems: 'center',
    marginHorizontal: 16,
    paddingVertical: 11
  },
  text: {
    fontSize: 16
  }
});
