// stream-chat-react-native-core/lib/commonjs/components/Poll/components/PollAnswersList.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PollAnswersListContent = exports.PollAnswersList = exports.PollAnswerListItem = exports.AnswerListAddCommentButton = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _streamChat = require("stream-chat");
var _PollInputDialog = require("./PollInputDialog");
var _contexts = require("../../../contexts");
var _getDateString = require("../../../utils/i18n/getDateString");
var _Avatar = require("../../Avatar/Avatar");
var _usePollAnswersPagination = require("../hooks/usePollAnswersPagination");
var _usePollState3 = require("../hooks/usePollState");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Poll/components/PollAnswersList.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var AnswerListAddCommentButton = exports.AnswerListAddCommentButton = function AnswerListAddCommentButton(props) {
  var _ownAnswer$answer_tex;
  var _useTranslationContex = (0, _contexts.useTranslationContext)(),
    t = _useTranslationContex.t;
  var _usePollContext = (0, _contexts.usePollContext)(),
    message = _usePollContext.message,
    poll = _usePollContext.poll;
  var _usePollState = (0, _usePollState3.usePollState)(),
    addComment = _usePollState.addComment,
    ownAnswer = _usePollState.ownAnswer;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    showAddCommentDialog = _useState2[0],
    setShowAddCommentDialog = _useState2[1];
  var onPress = props.onPress;
  var onPressHandler = (0, _react.useCallback)(function () {
    if (onPress) {
      onPress({
        message: message,
        poll: poll
      });
      return;
    }
    setShowAddCommentDialog(true);
  }, [message, onPress, poll]);
  var _useTheme = (0, _contexts.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    accent_dark_blue = _useTheme$theme$color.accent_dark_blue,
    bg_user = _useTheme$theme$color.bg_user,
    _useTheme$theme$poll = _useTheme$theme.poll,
    buttonContainer = _useTheme$theme$poll.answersList.buttonContainer,
    text = _useTheme$theme$poll.button.text;
  return (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [(0, _jsxRuntime.jsx)(_reactNative.Pressable, {
      onPress: onPressHandler,
      style: function style(_ref) {
        var pressed = _ref.pressed;
        return [{
          opacity: pressed ? 0.5 : 1
        }, styles.addCommentButtonContainer, {
          backgroundColor: bg_user
        }, buttonContainer];
      },
      children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: [styles.addCommentButtonText, {
          color: accent_dark_blue
        }, text],
        children: ownAnswer ? t('Update your comment') : t('Add a comment')
      })
    }), showAddCommentDialog ? (0, _jsxRuntime.jsx)(_PollInputDialog.PollInputDialog, {
      closeDialog: function closeDialog() {
        return setShowAddCommentDialog(false);
      },
      initialValue: (_ownAnswer$answer_tex = ownAnswer == null ? void 0 : ownAnswer.answer_text) != null ? _ownAnswer$answer_tex : '',
      onSubmit: addComment,
      title: t('Add a comment'),
      visible: showAddCommentDialog
    }) : null]
  });
};
var PollAnswerListItem = exports.PollAnswerListItem = function PollAnswerListItem(_ref2) {
  var _answer$user, _answer$user2, _answer$user3;
  var answer = _ref2.answer;
  var _useTranslationContex2 = (0, _contexts.useTranslationContext)(),
    t = _useTranslationContex2.t,
    tDateTimeParser = _useTranslationContex2.tDateTimeParser;
  var _usePollState2 = (0, _usePollState3.usePollState)(),
    votingVisibility = _usePollState2.votingVisibility;
  var _useTheme2 = (0, _contexts.useTheme)(),
    _useTheme2$theme = _useTheme2.theme,
    _useTheme2$theme$colo = _useTheme2$theme.colors,
    bg_user = _useTheme2$theme$colo.bg_user,
    black = _useTheme2$theme$colo.black,
    itemStyle = _useTheme2$theme.poll.answersList.item;
  var dateString = (0, _react.useMemo)(function () {
    return (0, _getDateString.getDateString)({
      date: answer.updated_at,
      t: t,
      tDateTimeParser: tDateTimeParser,
      timestampTranslationKey: 'timestamp/PollVote'
    });
  }, [answer.updated_at, t, tDateTimeParser]);
  var isAnonymous = (0, _react.useMemo)(function () {
    return votingVisibility === _streamChat.VotingVisibility.anonymous;
  }, [votingVisibility]);
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.listItemContainer, {
      backgroundColor: bg_user
    }, itemStyle.container],
    children: [(0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: [styles.listItemAnswerText, {
        color: black
      }, itemStyle.answerText],
      children: answer.answer_text
    }), (0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: [styles.listItemInfoContainer, itemStyle.infoContainer],
      children: [(0, _jsxRuntime.jsxs)(_reactNative.View, {
        style: [styles.listItemUserInfoContainer, itemStyle.userInfoContainer],
        children: [!isAnonymous && (_answer$user = answer.user) != null && _answer$user.image ? (0, _jsxRuntime.jsx)(_Avatar.Avatar, {
          image: (_answer$user2 = answer.user) == null ? void 0 : _answer$user2.image,
          size: 20
        }) : null, (0, _jsxRuntime.jsx)(_reactNative.Text, {
          style: {
            color: black,
            fontSize: 14,
            marginLeft: 2
          },
          children: isAnonymous ? t('Anonymous') : (_answer$user3 = answer.user) == null ? void 0 : _answer$user3.name
        })]
      }), (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: {
          color: black
        },
        children: dateString
      })]
    })]
  });
};
var renderPollAnswerListItem = function renderPollAnswerListItem(_ref3) {
  var item = _ref3.item;
  return (0, _jsxRuntime.jsx)(PollAnswerListItem, {
    answer: item
  });
};
var PollAnswersListContent = exports.PollAnswersListContent = function PollAnswersListContent(_ref4) {
  var additionalFlatListProps = _ref4.additionalFlatListProps;
  var _usePollAnswersPagina = (0, _usePollAnswersPagination.usePollAnswersPagination)(),
    hasNextPage = _usePollAnswersPagina.hasNextPage,
    loadMore = _usePollAnswersPagina.loadMore,
    pollAnswers = _usePollAnswersPagina.pollAnswers;
  var _useTheme3 = (0, _contexts.useTheme)(),
    _useTheme3$theme = _useTheme3.theme,
    white = _useTheme3$theme.colors.white,
    container = _useTheme3$theme.poll.answersList.container;
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: white
    }, container],
    children: [(0, _jsxRuntime.jsx)(_reactNative.FlatList, Object.assign({
      data: pollAnswers,
      keyExtractor: function keyExtractor(item) {
        return `poll_answer_${item.id}`;
      },
      onEndReached: function onEndReached() {
        return hasNextPage && loadMore();
      },
      renderItem: renderPollAnswerListItem
    }, additionalFlatListProps)), (0, _jsxRuntime.jsx)(AnswerListAddCommentButton, {})]
  });
};
var PollAnswersList = exports.PollAnswersList = function PollAnswersList(_ref5) {
  var additionalFlatListProps = _ref5.additionalFlatListProps,
    message = _ref5.message,
    poll = _ref5.poll,
    PollAnswersListOverride = _ref5.PollAnswersListContent;
  return (0, _jsxRuntime.jsx)(_contexts.PollContextProvider, {
    value: {
      message: message,
      poll: poll
    },
    children: PollAnswersListOverride ? (0, _jsxRuntime.jsx)(PollAnswersListOverride, {}) : (0, _jsxRuntime.jsx)(PollAnswersListContent, {
      additionalFlatListProps: additionalFlatListProps
    })
  });
};
var styles = _reactNative.StyleSheet.create({
  addCommentButtonContainer: {
    alignItems: 'center',
    borderRadius: 12,
    paddingHorizontal: 16,
    paddingVertical: 18
  },
  addCommentButtonText: {
    fontSize: 16
  },
  container: {
    flex: 1,
    margin: 16
  },
  listItemAnswerText: {
    fontSize: 16,
    fontWeight: '500'
  },
  listItemContainer: {
    borderRadius: 12,
    marginBottom: 8,
    paddingBottom: 20,
    paddingHorizontal: 16,
    paddingTop: 12
  },
  listItemInfoContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 24
  },
  listItemUserInfoContainer: {
    alignItems: 'center',
    flexDirection: 'row'
  }
});
