// stream-chat-react-native-core/lib/module/components/MessageInput/components/AudioRecorder/AudioRecordingPreview.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AudioRecordingPreview = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _dayjs = _interopRequireDefault(require("dayjs"));
var _ThemeContext = require("../../../../contexts/themeContext/ThemeContext");
var _useAudioPlayerControl = require("../../../../hooks/useAudioPlayerControl");
var _useStateStore2 = require("../../../../hooks/useStateStore");
var _icons = require("../../../../icons");
var _native = require("../../../../native");
var _WaveProgressBar = require("../../../ProgressControl/WaveProgressBar");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageInput/components/AudioRecorder/AudioRecordingPreview.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ONE_SECOND_IN_MILLISECONDS = 1000;
var ONE_HOUR_IN_MILLISECONDS = 3600 * 1000;
var audioPlayerSelector = function audioPlayerSelector(state) {
  return {
    duration: state.duration,
    isPlaying: state.isPlaying,
    position: state.position,
    progress: state.progress
  };
};
var AudioRecordingPreview = exports.AudioRecordingPreview = function AudioRecordingPreview(props) {
  var recordingDuration = props.recordingDuration,
    uri = props.uri,
    waveformData = props.waveformData;
  var audioPlayer = (0, _useAudioPlayerControl.useAudioPlayerControl)({
    duration: recordingDuration / ONE_SECOND_IN_MILLISECONDS,
    mimeType: 'audio/aac',
    previewVoiceRecording: !(_native.NativeHandlers.SDK === 'stream-chat-expo'),
    type: 'voiceRecording',
    uri: uri
  });
  var _useStateStore = (0, _useStateStore2.useStateStore)(audioPlayer.state, audioPlayerSelector),
    duration = _useStateStore.duration,
    isPlaying = _useStateStore.isPlaying,
    position = _useStateStore.position,
    progress = _useStateStore.progress;
  (0, _react.useEffect)(function () {
    return function () {
      audioPlayer.onRemove();
    };
  }, [audioPlayer]);
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    accent_blue = _useTheme$theme$color.accent_blue,
    grey_dark = _useTheme$theme$color.grey_dark,
    _useTheme$theme$messa = _useTheme$theme.messageInput.audioRecordingPreview,
    container = _useTheme$theme$messa.container,
    currentTime = _useTheme$theme$messa.currentTime,
    infoContainer = _useTheme$theme$messa.infoContainer,
    pauseIcon = _useTheme$theme$messa.pauseIcon,
    playIcon = _useTheme$theme$messa.playIcon,
    progressBar = _useTheme$theme$messa.progressBar;
  var handlePlayPause = function handlePlayPause() {
    audioPlayer.toggle();
  };
  var progressDuration = (0, _react.useMemo)(function () {
    return position ? position / ONE_HOUR_IN_MILLISECONDS >= 1 ? _dayjs.default.duration(position, 'milliseconds').format('HH:mm:ss') : _dayjs.default.duration(position, 'milliseconds').format('mm:ss') : _dayjs.default.duration(duration, 'milliseconds').format('mm:ss');
  }, [duration, position]);
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.container, container],
    children: [(0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: [styles.infoContainer, infoContainer],
      children: [(0, _jsxRuntime.jsx)(_reactNative.Pressable, {
        onPress: handlePlayPause,
        children: !isPlaying ? (0, _jsxRuntime.jsx)(_icons.Play, Object.assign({
          fill: accent_blue,
          height: 32,
          width: 32
        }, playIcon)) : (0, _jsxRuntime.jsx)(_icons.Pause, Object.assign({
          fill: accent_blue,
          height: 32,
          width: 32
        }, pauseIcon))
      }), (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: [styles.currentTime, {
          color: grey_dark
        }, currentTime],
        children: progressDuration
      })]
    }), (0, _jsxRuntime.jsx)(_reactNative.View, {
      style: [styles.progressBar, progressBar],
      children: (0, _jsxRuntime.jsx)(_WaveProgressBar.WaveProgressBar, {
        progress: progress,
        waveformData: waveformData
      })
    })]
  });
};
var styles = _reactNative.StyleSheet.create({
  container: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'row',
    paddingBottom: 8,
    paddingTop: 4
  },
  currentTime: {
    fontSize: 16,
    marginLeft: 4
  },
  infoContainer: {
    alignItems: 'center',
    display: 'flex',
    flex: 1,
    flexDirection: 'row'
  },
  progressBar: {
    flex: 3
  }
});
AudioRecordingPreview.displayName = 'AudioRecordingPreview{messageInput}';
