// stream-chat-react-native-core/lib/module/components/MessageInput/components/AttachmentPreview/DismissAttachmentUpload.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DismissAttachmentUpload = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _ThemeContext = require("../../../../contexts/themeContext/ThemeContext");
var _icons = require("../../../../icons");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageInput/components/AttachmentPreview/DismissAttachmentUpload.tsx";
var DismissAttachmentUpload = exports.DismissAttachmentUpload = function DismissAttachmentUpload(_ref) {
  var onPress = _ref.onPress;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    overlay = _useTheme$theme$color.overlay,
    white = _useTheme$theme$color.white,
    _useTheme$theme$messa = _useTheme$theme.messageInput.dismissAttachmentUpload,
    dismiss = _useTheme$theme$messa.dismiss,
    dismissIcon = _useTheme$theme$messa.dismissIcon,
    dismissIconColor = _useTheme$theme$messa.dismissIconColor;
  return (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
    onPress: onPress,
    style: function style(_ref2) {
      var pressed = _ref2.pressed;
      return [styles.dismiss, {
        backgroundColor: overlay,
        opacity: pressed ? 0.8 : 1
      }, dismiss];
    },
    testID: "remove-upload-preview",
    children: (0, _jsxRuntime.jsx)(_icons.Close, Object.assign({
      pathFill: dismissIconColor || white
    }, dismissIcon))
  });
};
var styles = _reactNative.StyleSheet.create({
  dismiss: {
    borderRadius: 24,
    position: 'absolute',
    right: 8,
    top: 8
  }
});
