// stream-chat-react-native-core/lib/commonjs/contexts/audioPlayerContext/AudioPlayerContext.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAudioPlayerContext = exports.useActiveAudioPlayer = exports.AudioPlayerProvider = exports.AudioPlayerContext = void 0;
var _react = _interopRequireWildcard(require("react"));
var _useStateStore2 = require("../../hooks/useStateStore");
var _audioPlayerPool = require("../../state-store/audio-player-pool");
var _defaultBaseContextValue = require("../utils/defaultBaseContextValue");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/contexts/audioPlayerContext/AudioPlayerContext.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var AudioPlayerContext = exports.AudioPlayerContext = (0, _react.createContext)(_defaultBaseContextValue.DEFAULT_BASE_CONTEXT_VALUE);
var AudioPlayerProvider = exports.AudioPlayerProvider = function AudioPlayerProvider(_ref) {
  var value = _ref.value,
    children = _ref.children;
  var allowConcurrentAudioPlayback = value.allowConcurrentAudioPlayback;
  var audioPlayerPool = (0, _react.useMemo)(function () {
    return new _audioPlayerPool.AudioPlayerPool({
      allowConcurrentAudioPlayback: allowConcurrentAudioPlayback
    });
  }, [allowConcurrentAudioPlayback]);
  var audioPlayerPoolContextValue = (0, _react.useMemo)(function () {
    return {
      audioPlayerPool: audioPlayerPool
    };
  }, [audioPlayerPool]);
  (0, _react.useEffect)(function () {
    return function () {
      audioPlayerPool.clear();
    };
  }, [audioPlayerPool]);
  return (0, _jsxRuntime.jsx)(AudioPlayerContext.Provider, {
    value: audioPlayerPoolContextValue,
    children: children
  });
};
var useAudioPlayerContext = exports.useAudioPlayerContext = function useAudioPlayerContext() {
  return (0, _react.useContext)(AudioPlayerContext);
};
var activeAudioPlayerSelector = function activeAudioPlayerSelector(_ref2) {
  var activeAudioPlayer = _ref2.activeAudioPlayer;
  return {
    activeAudioPlayer: activeAudioPlayer
  };
};
var useActiveAudioPlayer = exports.useActiveAudioPlayer = function useActiveAudioPlayer() {
  var _useStateStore;
  var _useContext = (0, _react.useContext)(AudioPlayerContext),
    audioPlayerPool = _useContext.audioPlayerPool;
  var _ref3 = (_useStateStore = (0, _useStateStore2.useStateStore)(audioPlayerPool.state, activeAudioPlayerSelector)) != null ? _useStateStore : {},
    activeAudioPlayer = _ref3.activeAudioPlayer;
  return activeAudioPlayer;
};
