// stream-chat-react-native-core/lib/commonjs/contexts/messageInputContext/hooks/useCreateMessageInputContext.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCreateMessageInputContext = void 0;
var _react = require("react");
var useCreateMessageInputContext = exports.useCreateMessageInputContext = function useCreateMessageInputContext(_ref) {
  var additionalTextInputProps = _ref.additionalTextInputProps,
    asyncMessagesLockDistance = _ref.asyncMessagesLockDistance,
    asyncMessagesMinimumPressDuration = _ref.asyncMessagesMinimumPressDuration,
    asyncMessagesMultiSendEnabled = _ref.asyncMessagesMultiSendEnabled,
    asyncMessagesSlideToCancelDistance = _ref.asyncMessagesSlideToCancelDistance,
    AttachButton = _ref.AttachButton,
    AttachmentPickerBottomSheetHandle = _ref.AttachmentPickerBottomSheetHandle,
    attachmentPickerBottomSheetHandleHeight = _ref.attachmentPickerBottomSheetHandleHeight,
    attachmentPickerBottomSheetHeight = _ref.attachmentPickerBottomSheetHeight,
    AttachmentPickerSelectionBar = _ref.AttachmentPickerSelectionBar,
    attachmentSelectionBarHeight = _ref.attachmentSelectionBarHeight,
    AttachmentUploadPreviewList = _ref.AttachmentUploadPreviewList,
    AttachmentUploadProgressIndicator = _ref.AttachmentUploadProgressIndicator,
    AudioAttachmentUploadPreview = _ref.AudioAttachmentUploadPreview,
    AudioRecorder = _ref.AudioRecorder,
    audioRecordingEnabled = _ref.audioRecordingEnabled,
    AudioRecordingInProgress = _ref.AudioRecordingInProgress,
    AudioRecordingLockIndicator = _ref.AudioRecordingLockIndicator,
    AudioRecordingPreview = _ref.AudioRecordingPreview,
    AudioRecordingWaveform = _ref.AudioRecordingWaveform,
    AutoCompleteSuggestionHeader = _ref.AutoCompleteSuggestionHeader,
    AutoCompleteSuggestionItem = _ref.AutoCompleteSuggestionItem,
    AutoCompleteSuggestionList = _ref.AutoCompleteSuggestionList,
    CameraSelectorIcon = _ref.CameraSelectorIcon,
    closeAttachmentPicker = _ref.closeAttachmentPicker,
    closePollCreationDialog = _ref.closePollCreationDialog,
    CommandInput = _ref.CommandInput,
    CommandsButton = _ref.CommandsButton,
    compressImageQuality = _ref.compressImageQuality,
    cooldownEndsAt = _ref.cooldownEndsAt,
    CooldownTimer = _ref.CooldownTimer,
    CreatePollContent = _ref.CreatePollContent,
    CreatePollIcon = _ref.CreatePollIcon,
    editMessage = _ref.editMessage,
    FileAttachmentUploadPreview = _ref.FileAttachmentUploadPreview,
    FileSelectorIcon = _ref.FileSelectorIcon,
    handleAttachButtonPress = _ref.handleAttachButtonPress,
    hasCameraPicker = _ref.hasCameraPicker,
    hasCommands = _ref.hasCommands,
    hasFilePicker = _ref.hasFilePicker,
    hasImagePicker = _ref.hasImagePicker,
    ImageAttachmentUploadPreview = _ref.ImageAttachmentUploadPreview,
    ImageSelectorIcon = _ref.ImageSelectorIcon,
    Input = _ref.Input,
    inputBoxRef = _ref.inputBoxRef,
    InputButtons = _ref.InputButtons,
    InputEditingStateHeader = _ref.InputEditingStateHeader,
    InputReplyStateHeader = _ref.InputReplyStateHeader,
    MoreOptionsButton = _ref.MoreOptionsButton,
    openAttachmentPicker = _ref.openAttachmentPicker,
    openPollCreationDialog = _ref.openPollCreationDialog,
    pickAndUploadImageFromNativePicker = _ref.pickAndUploadImageFromNativePicker,
    pickFile = _ref.pickFile,
    selectedPicker = _ref.selectedPicker,
    SendButton = _ref.SendButton,
    sendMessage = _ref.sendMessage,
    SendMessageDisallowedIndicator = _ref.SendMessageDisallowedIndicator,
    setInputBoxRef = _ref.setInputBoxRef,
    setInputRef = _ref.setInputRef,
    showPollCreationDialog = _ref.showPollCreationDialog,
    ShowThreadMessageInChannelButton = _ref.ShowThreadMessageInChannelButton,
    StartAudioRecordingButton = _ref.StartAudioRecordingButton,
    StopMessageStreamingButton = _ref.StopMessageStreamingButton,
    takeAndUploadImage = _ref.takeAndUploadImage,
    thread = _ref.thread,
    toggleAttachmentPicker = _ref.toggleAttachmentPicker,
    uploadNewFile = _ref.uploadNewFile,
    VideoAttachmentUploadPreview = _ref.VideoAttachmentUploadPreview,
    VideoRecorderSelectorIcon = _ref.VideoRecorderSelectorIcon;
  var threadId = thread == null ? void 0 : thread.id;
  var messageInputContext = (0, _react.useMemo)(function () {
    return {
      additionalTextInputProps: additionalTextInputProps,
      asyncMessagesLockDistance: asyncMessagesLockDistance,
      asyncMessagesMinimumPressDuration: asyncMessagesMinimumPressDuration,
      asyncMessagesMultiSendEnabled: asyncMessagesMultiSendEnabled,
      asyncMessagesSlideToCancelDistance: asyncMessagesSlideToCancelDistance,
      AttachButton: AttachButton,
      AttachmentPickerBottomSheetHandle: AttachmentPickerBottomSheetHandle,
      attachmentPickerBottomSheetHandleHeight: attachmentPickerBottomSheetHandleHeight,
      attachmentPickerBottomSheetHeight: attachmentPickerBottomSheetHeight,
      AttachmentPickerSelectionBar: AttachmentPickerSelectionBar,
      attachmentSelectionBarHeight: attachmentSelectionBarHeight,
      AttachmentUploadPreviewList: AttachmentUploadPreviewList,
      AttachmentUploadProgressIndicator: AttachmentUploadProgressIndicator,
      AudioAttachmentUploadPreview: AudioAttachmentUploadPreview,
      AudioRecorder: AudioRecorder,
      audioRecordingEnabled: audioRecordingEnabled,
      AudioRecordingInProgress: AudioRecordingInProgress,
      AudioRecordingLockIndicator: AudioRecordingLockIndicator,
      AudioRecordingPreview: AudioRecordingPreview,
      AudioRecordingWaveform: AudioRecordingWaveform,
      AutoCompleteSuggestionHeader: AutoCompleteSuggestionHeader,
      AutoCompleteSuggestionItem: AutoCompleteSuggestionItem,
      AutoCompleteSuggestionList: AutoCompleteSuggestionList,
      CameraSelectorIcon: CameraSelectorIcon,
      closeAttachmentPicker: closeAttachmentPicker,
      closePollCreationDialog: closePollCreationDialog,
      CommandInput: CommandInput,
      CommandsButton: CommandsButton,
      compressImageQuality: compressImageQuality,
      cooldownEndsAt: cooldownEndsAt,
      CooldownTimer: CooldownTimer,
      CreatePollContent: CreatePollContent,
      CreatePollIcon: CreatePollIcon,
      editMessage: editMessage,
      FileAttachmentUploadPreview: FileAttachmentUploadPreview,
      FileSelectorIcon: FileSelectorIcon,
      handleAttachButtonPress: handleAttachButtonPress,
      hasCameraPicker: hasCameraPicker,
      hasCommands: hasCommands,
      hasFilePicker: hasFilePicker,
      hasImagePicker: hasImagePicker,
      ImageAttachmentUploadPreview: ImageAttachmentUploadPreview,
      ImageSelectorIcon: ImageSelectorIcon,
      Input: Input,
      inputBoxRef: inputBoxRef,
      InputButtons: InputButtons,
      InputEditingStateHeader: InputEditingStateHeader,
      InputReplyStateHeader: InputReplyStateHeader,
      MoreOptionsButton: MoreOptionsButton,
      openAttachmentPicker: openAttachmentPicker,
      openPollCreationDialog: openPollCreationDialog,
      pickAndUploadImageFromNativePicker: pickAndUploadImageFromNativePicker,
      pickFile: pickFile,
      selectedPicker: selectedPicker,
      SendButton: SendButton,
      sendMessage: sendMessage,
      SendMessageDisallowedIndicator: SendMessageDisallowedIndicator,
      setInputBoxRef: setInputBoxRef,
      setInputRef: setInputRef,
      showPollCreationDialog: showPollCreationDialog,
      ShowThreadMessageInChannelButton: ShowThreadMessageInChannelButton,
      StartAudioRecordingButton: StartAudioRecordingButton,
      StopMessageStreamingButton: StopMessageStreamingButton,
      takeAndUploadImage: takeAndUploadImage,
      toggleAttachmentPicker: toggleAttachmentPicker,
      uploadNewFile: uploadNewFile,
      VideoAttachmentUploadPreview: VideoAttachmentUploadPreview,
      VideoRecorderSelectorIcon: VideoRecorderSelectorIcon
    };
  }, [cooldownEndsAt, threadId, showPollCreationDialog, selectedPicker]);
  return messageInputContext;
};
