// stream-chat-react-native-core/lib/module/components/Message/hooks/useMessageActions.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMessageActions = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = _interopRequireDefault(require("react"));
var _useMessageActionHandlers = require("./useMessageActionHandlers");
var _ThemeContext = require("../../../contexts/themeContext/ThemeContext");
var _icons = require("../../../icons");
var _removeReservedFields = require("../../../utils/removeReservedFields");
var _utils = require("../../../utils/utils");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Message/hooks/useMessageActions.tsx";
var useMessageActions = exports.useMessageActions = function useMessageActions(_ref) {
  var _message$user3;
  var channel = _ref.channel,
    client = _ref.client,
    deleteMessageFromContext = _ref.deleteMessage,
    deleteReaction = _ref.deleteReaction,
    dismissOverlay = _ref.dismissOverlay,
    enforceUniqueReaction = _ref.enforceUniqueReaction,
    handleBan = _ref.handleBan,
    handleCopy = _ref.handleCopy,
    handleDelete = _ref.handleDelete,
    handleDeleteForMe = _ref.handleDeleteForMe,
    handleEdit = _ref.handleEdit,
    handleFlag = _ref.handleFlag,
    handleMarkUnread = _ref.handleMarkUnread,
    handleMute = _ref.handleMute,
    handlePinMessage = _ref.handlePinMessage,
    handleQuotedReply = _ref.handleQuotedReply,
    handleReactionProp = _ref.handleReaction,
    handleRetry = _ref.handleRetry,
    handleThreadReply = _ref.handleThreadReply,
    message = _ref.message,
    onThreadSelect = _ref.onThreadSelect,
    openThread = _ref.openThread,
    retrySendMessage = _ref.retrySendMessage,
    selectReaction = _ref.selectReaction,
    sendReaction = _ref.sendReaction,
    setEditingState = _ref.setEditingState,
    supportedReactions = _ref.supportedReactions,
    t = _ref.t,
    setQuotedMessage = _ref.setQuotedMessage;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme$color = _useTheme.theme.colors,
    accent_red = _useTheme$theme$color.accent_red,
    grey = _useTheme$theme$color.grey;
  var _useMessageActionHand = (0, _useMessageActionHandlers.useMessageActionHandlers)({
      channel: channel,
      client: client,
      deleteMessage: deleteMessageFromContext,
      deleteReaction: deleteReaction,
      enforceUniqueReaction: enforceUniqueReaction,
      message: message,
      retrySendMessage: retrySendMessage,
      sendReaction: sendReaction,
      setEditingState: setEditingState,
      setQuotedMessage: setQuotedMessage,
      supportedReactions: supportedReactions
    }),
    handleCopyMessage = _useMessageActionHand.handleCopyMessage,
    handleDeleteMessage = _useMessageActionHand.handleDeleteMessage,
    handleDeleteForMeMessage = _useMessageActionHand.handleDeleteForMeMessage,
    handleEditMessage = _useMessageActionHand.handleEditMessage,
    handleFlagMessage = _useMessageActionHand.handleFlagMessage,
    handleMarkUnreadMessage = _useMessageActionHand.handleMarkUnreadMessage,
    handleQuotedReplyMessage = _useMessageActionHand.handleQuotedReplyMessage,
    handleResendMessage = _useMessageActionHand.handleResendMessage,
    handleToggleBanUser = _useMessageActionHand.handleToggleBanUser,
    handleToggleMuteUser = _useMessageActionHand.handleToggleMuteUser,
    handleTogglePinMessage = _useMessageActionHand.handleTogglePinMessage,
    handleToggleReaction = _useMessageActionHand.handleToggleReaction;
  var error = message.type === 'error' || message.status === _utils.MessageStatusTypes.FAILED;
  var onOpenThread = function onOpenThread() {
    if (onThreadSelect) {
      onThreadSelect(message);
    }
    if (openThread) {
      openThread(message);
    }
  };
  var isMuted = (client.mutedUsers || []).some(function (mute) {
    var _message$user;
    return mute.user.id === client.userID && mute.target.id === ((_message$user = message.user) == null ? void 0 : _message$user.id);
  });
  var banUser = {
    action: function () {
      var _action = (0, _asyncToGenerator2.default)(function* () {
        var _message$user2;
        dismissOverlay();
        if ((_message$user2 = message.user) != null && _message$user2.id) {
          if (handleBan) {
            handleBan(message);
          }
          yield handleToggleBanUser();
        }
      });
      function action() {
        return _action.apply(this, arguments);
      }
      return action;
    }(),
    actionType: 'banUser',
    icon: (0, _jsxRuntime.jsx)(_icons.UserDelete, {
      pathFill: grey
    }),
    title: (_message$user3 = message.user) != null && _message$user3.banned ? t('Unban User') : t('Ban User')
  };
  var copyMessage = {
    action: function action() {
      dismissOverlay();
      if (handleCopy) {
        handleCopy(message);
      }
      handleCopyMessage();
    },
    actionType: 'copyMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.Copy, {
      pathFill: grey
    }),
    title: t('Copy Message')
  };
  var deleteMessage = {
    action: function action() {
      dismissOverlay();
      if (handleDelete) {
        handleDelete(message);
      }
      handleDeleteMessage();
    },
    actionType: 'deleteMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.Delete, {
      fill: accent_red,
      size: 24
    }),
    title: t('Delete Message'),
    titleStyle: {
      color: accent_red
    }
  };
  var deleteForMeMessage = {
    action: function action() {
      dismissOverlay();
      if (handleDeleteForMe) {
        handleDeleteForMe(message);
      }
      handleDeleteForMeMessage();
    },
    actionType: 'deleteForMeMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.Delete, {
      fill: accent_red,
      size: 24
    }),
    title: t('Delete for me')
  };
  var editMessage = {
    action: function action() {
      dismissOverlay();
      if (handleEdit) {
        handleEdit(message);
      }
      handleEditMessage();
    },
    actionType: 'editMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.Edit, {
      pathFill: grey
    }),
    title: t('Edit Message')
  };
  var flagMessage = {
    action: function action() {
      dismissOverlay();
      if (handleFlag) {
        handleFlag(message);
      }
      handleFlagMessage();
    },
    actionType: 'flagMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.MessageFlag, {
      pathFill: grey
    }),
    title: t('Flag Message')
  };
  var markUnread = {
    action: function action() {
      dismissOverlay();
      if (handleMarkUnread) {
        handleMarkUnread(message);
      }
      handleMarkUnreadMessage();
    },
    actionType: 'markUnread',
    icon: (0, _jsxRuntime.jsx)(_icons.UnreadIndicator, {
      fill: grey,
      size: 24
    }),
    title: t('Mark as Unread')
  };
  var pinMessage = {
    action: function action() {
      dismissOverlay();
      if (handlePinMessage) {
        handlePinMessage(message);
      }
      handleTogglePinMessage();
    },
    actionType: 'pinMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.Pin, {
      pathFill: grey,
      size: 24
    }),
    title: t('Pin to Conversation')
  };
  var unpinMessage = {
    action: function action() {
      dismissOverlay();
      if (handlePinMessage) {
        handlePinMessage(message);
      }
      handleTogglePinMessage();
    },
    actionType: 'unpinMessage',
    icon: (0, _jsxRuntime.jsx)(_icons.Unpin, {
      pathFill: grey
    }),
    title: t('Unpin from Conversation')
  };
  var handleReaction = !error ? selectReaction ? selectReaction(message) : (function () {
    var _ref2 = (0, _asyncToGenerator2.default)(function* (reactionType) {
      if (handleReactionProp) {
        handleReactionProp(message, reactionType);
      }
      yield handleToggleReaction(reactionType);
    });
    return function (_x) {
      return _ref2.apply(this, arguments);
    };
  }()) : undefined;
  var muteUser = {
    action: function () {
      var _action2 = (0, _asyncToGenerator2.default)(function* () {
        var _message$user4;
        dismissOverlay();
        if ((_message$user4 = message.user) != null && _message$user4.id) {
          if (handleMute) {
            handleMute(message);
          }
          yield handleToggleMuteUser();
        }
      });
      function action() {
        return _action2.apply(this, arguments);
      }
      return action;
    }(),
    actionType: 'muteUser',
    icon: (0, _jsxRuntime.jsx)(_icons.Mute, {
      pathFill: grey
    }),
    title: isMuted ? t('Unmute User') : t('Mute User')
  };
  var quotedReply = {
    action: function action() {
      dismissOverlay();
      if (handleQuotedReply) {
        handleQuotedReply(message);
      }
      handleQuotedReplyMessage();
    },
    actionType: 'quotedReply',
    icon: (0, _jsxRuntime.jsx)(_icons.CurveLineLeftUp, {
      pathFill: grey
    }),
    title: t('Reply')
  };
  var retry = {
    action: function () {
      var _action3 = (0, _asyncToGenerator2.default)(function* () {
        dismissOverlay();
        var messageWithoutReservedFields = (0, _removeReservedFields.removeReservedFields)(message);
        if (handleRetry) {
          handleRetry(messageWithoutReservedFields);
        }
        yield handleResendMessage();
      });
      function action() {
        return _action3.apply(this, arguments);
      }
      return action;
    }(),
    actionType: 'retry',
    icon: (0, _jsxRuntime.jsx)(_icons.Resend, {
      pathFill: grey
    }),
    title: t('Resend')
  };
  var threadReply = {
    action: function action() {
      dismissOverlay();
      if (handleThreadReply) {
        handleThreadReply(message);
      }
      onOpenThread();
    },
    actionType: 'threadReply',
    icon: (0, _jsxRuntime.jsx)(_icons.ThreadReply, {
      pathFill: grey
    }),
    title: t('Thread Reply')
  };
  return {
    banUser: banUser,
    copyMessage: copyMessage,
    deleteForMeMessage: deleteForMeMessage,
    deleteMessage: deleteMessage,
    editMessage: editMessage,
    flagMessage: flagMessage,
    handleReaction: handleReaction,
    markUnread: markUnread,
    muteUser: muteUser,
    pinMessage: pinMessage,
    quotedReply: quotedReply,
    retry: retry,
    threadReply: threadReply,
    unpinMessage: unpinMessage
  };
};
