// stream-chat-react-native-core/lib/module/components/Channel/hooks/useCreateChannelContext.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCreateChannelContext = void 0;
var _react = require("react");
var useCreateChannelContext = exports.useCreateChannelContext = function useCreateChannelContext(_ref) {
  var channel = _ref.channel,
    channelUnreadState = _ref.channelUnreadState,
    channelUnreadStateStore = _ref.channelUnreadStateStore,
    disabled = _ref.disabled,
    EmptyStateIndicator = _ref.EmptyStateIndicator,
    enableMessageGroupingByUser = _ref.enableMessageGroupingByUser,
    enforceUniqueReaction = _ref.enforceUniqueReaction,
    error = _ref.error,
    hideDateSeparators = _ref.hideDateSeparators,
    hideStickyDateHeader = _ref.hideStickyDateHeader,
    highlightedMessageId = _ref.highlightedMessageId,
    isChannelActive = _ref.isChannelActive,
    lastRead = _ref.lastRead,
    loadChannelAroundMessage = _ref.loadChannelAroundMessage,
    loadChannelAtFirstUnreadMessage = _ref.loadChannelAtFirstUnreadMessage,
    loading = _ref.loading,
    LoadingIndicator = _ref.LoadingIndicator,
    markRead = _ref.markRead,
    maxTimeBetweenGroupedMessages = _ref.maxTimeBetweenGroupedMessages,
    maximumMessageLimit = _ref.maximumMessageLimit,
    members = _ref.members,
    NetworkDownIndicator = _ref.NetworkDownIndicator,
    read = _ref.read,
    reloadChannel = _ref.reloadChannel,
    scrollToFirstUnreadThreshold = _ref.scrollToFirstUnreadThreshold,
    setChannelUnreadState = _ref.setChannelUnreadState,
    setLastRead = _ref.setLastRead,
    setTargetedMessage = _ref.setTargetedMessage,
    hasPendingInitialTargetLoad = _ref.hasPendingInitialTargetLoad,
    StickyHeader = _ref.StickyHeader,
    targetedMessage = _ref.targetedMessage,
    threadList = _ref.threadList,
    uploadAbortControllerRef = _ref.uploadAbortControllerRef,
    watcherCount = _ref.watcherCount,
    watchers = _ref.watchers;
  var channelId = channel == null ? void 0 : channel.id;
  var lastReadTime = lastRead == null ? void 0 : lastRead.getTime();
  var membersLength = Object.keys(members).length;
  var readUsers = Object.values(read);
  var readUsersLength = readUsers.length;
  var readUsersLastReads = readUsers.map(function (_ref2) {
    var _last_read$toISOStrin;
    var last_read = _ref2.last_read;
    return (_last_read$toISOStrin = last_read == null ? void 0 : last_read.toISOString()) != null ? _last_read$toISOStrin : '';
  }).join();
  var channelContext = (0, _react.useMemo)(function () {
    return {
      channel: channel,
      channelUnreadState: channelUnreadState,
      channelUnreadStateStore: channelUnreadStateStore,
      disabled: disabled,
      EmptyStateIndicator: EmptyStateIndicator,
      enableMessageGroupingByUser: enableMessageGroupingByUser,
      enforceUniqueReaction: enforceUniqueReaction,
      error: error,
      hasPendingInitialTargetLoad: hasPendingInitialTargetLoad,
      hideDateSeparators: hideDateSeparators,
      hideStickyDateHeader: hideStickyDateHeader,
      highlightedMessageId: highlightedMessageId,
      isChannelActive: isChannelActive,
      lastRead: lastRead,
      loadChannelAroundMessage: loadChannelAroundMessage,
      loadChannelAtFirstUnreadMessage: loadChannelAtFirstUnreadMessage,
      loading: loading,
      LoadingIndicator: LoadingIndicator,
      markRead: markRead,
      maximumMessageLimit: maximumMessageLimit,
      maxTimeBetweenGroupedMessages: maxTimeBetweenGroupedMessages,
      members: members,
      NetworkDownIndicator: NetworkDownIndicator,
      read: read,
      reloadChannel: reloadChannel,
      scrollToFirstUnreadThreshold: scrollToFirstUnreadThreshold,
      setChannelUnreadState: setChannelUnreadState,
      setLastRead: setLastRead,
      setTargetedMessage: setTargetedMessage,
      StickyHeader: StickyHeader,
      targetedMessage: targetedMessage,
      threadList: threadList,
      uploadAbortControllerRef: uploadAbortControllerRef,
      watcherCount: watcherCount,
      watchers: watchers
    };
  }, [channelId, disabled, error, isChannelActive, highlightedMessageId, lastReadTime, loading, membersLength, readUsersLength, readUsersLastReads, targetedMessage, threadList, watcherCount, maximumMessageLimit]);
  return channelContext;
};
