// stream-chat-react-native-core/lib/module/components/Channel/hooks/useCreateMessagesContext.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCreateMessagesContext = void 0;
var _react = require("react");
var useCreateMessagesContext = exports.useCreateMessagesContext = function useCreateMessagesContext(_ref) {
  var additionalPressableProps = _ref.additionalPressableProps,
    Attachment = _ref.Attachment,
    AttachmentActions = _ref.AttachmentActions,
    AudioAttachment = _ref.AudioAttachment,
    Card = _ref.Card,
    CardCover = _ref.CardCover,
    CardFooter = _ref.CardFooter,
    CardHeader = _ref.CardHeader,
    channelId = _ref.channelId,
    customMessageSwipeAction = _ref.customMessageSwipeAction,
    DateHeader = _ref.DateHeader,
    deletedMessagesVisibilityType = _ref.deletedMessagesVisibilityType,
    deleteMessage = _ref.deleteMessage,
    deleteReaction = _ref.deleteReaction,
    disableTypingIndicator = _ref.disableTypingIndicator,
    dismissKeyboardOnMessageTouch = _ref.dismissKeyboardOnMessageTouch,
    enableMessageGroupingByUser = _ref.enableMessageGroupingByUser,
    enableSwipeToReply = _ref.enableSwipeToReply,
    FileAttachment = _ref.FileAttachment,
    FileAttachmentGroup = _ref.FileAttachmentGroup,
    FileAttachmentIcon = _ref.FileAttachmentIcon,
    FlatList = _ref.FlatList,
    forceAlignMessages = _ref.forceAlignMessages,
    Gallery = _ref.Gallery,
    getMessageGroupStyle = _ref.getMessageGroupStyle,
    getMessagesGroupStyles = _ref.getMessagesGroupStyles,
    Giphy = _ref.Giphy,
    giphyVersion = _ref.giphyVersion,
    handleBan = _ref.handleBan,
    handleCopy = _ref.handleCopy,
    handleDelete = _ref.handleDelete,
    handleDeleteForMe = _ref.handleDeleteForMe,
    handleEdit = _ref.handleEdit,
    handleFlag = _ref.handleFlag,
    handleMarkUnread = _ref.handleMarkUnread,
    handleMute = _ref.handleMute,
    handlePinMessage = _ref.handlePinMessage,
    handleQuotedReply = _ref.handleQuotedReply,
    handleReaction = _ref.handleReaction,
    handleRetry = _ref.handleRetry,
    handleThreadReply = _ref.handleThreadReply,
    hasCreatePoll = _ref.hasCreatePoll,
    ImageLoadingFailedIndicator = _ref.ImageLoadingFailedIndicator,
    ImageLoadingIndicator = _ref.ImageLoadingIndicator,
    ImageReloadIndicator = _ref.ImageReloadIndicator,
    initialScrollToFirstUnreadMessage = _ref.initialScrollToFirstUnreadMessage,
    InlineDateSeparator = _ref.InlineDateSeparator,
    InlineUnreadIndicator = _ref.InlineUnreadIndicator,
    isAttachmentEqual = _ref.isAttachmentEqual,
    isMessageAIGenerated = _ref.isMessageAIGenerated,
    legacyImageViewerSwipeBehaviour = _ref.legacyImageViewerSwipeBehaviour,
    markdownRules = _ref.markdownRules,
    Message = _ref.Message,
    MessageActionList = _ref.MessageActionList,
    MessageActionListItem = _ref.MessageActionListItem,
    messageActions = _ref.messageActions,
    MessageAvatar = _ref.MessageAvatar,
    MessageBlocked = _ref.MessageBlocked,
    MessageBounce = _ref.MessageBounce,
    MessageContent = _ref.MessageContent,
    messageContentOrder = _ref.messageContentOrder,
    MessageDeleted = _ref.MessageDeleted,
    MessageEditedTimestamp = _ref.MessageEditedTimestamp,
    MessageError = _ref.MessageError,
    MessageFooter = _ref.MessageFooter,
    MessageHeader = _ref.MessageHeader,
    MessageList = _ref.MessageList,
    MessageLocation = _ref.MessageLocation,
    MessageMenu = _ref.MessageMenu,
    MessagePinnedHeader = _ref.MessagePinnedHeader,
    MessageReactionPicker = _ref.MessageReactionPicker,
    MessageReplies = _ref.MessageReplies,
    MessageRepliesAvatars = _ref.MessageRepliesAvatars,
    MessageSimple = _ref.MessageSimple,
    MessageStatus = _ref.MessageStatus,
    MessageSwipeContent = _ref.MessageSwipeContent,
    messageSwipeToReplyHitSlop = _ref.messageSwipeToReplyHitSlop,
    MessageSystem = _ref.MessageSystem,
    MessageText = _ref.MessageText,
    messageTextNumberOfLines = _ref.messageTextNumberOfLines,
    MessageTimestamp = _ref.MessageTimestamp,
    MessageUserReactions = _ref.MessageUserReactions,
    MessageUserReactionsAvatar = _ref.MessageUserReactionsAvatar,
    MessageUserReactionsItem = _ref.MessageUserReactionsItem,
    myMessageTheme = _ref.myMessageTheme,
    onLongPressMessage = _ref.onLongPressMessage,
    onPressInMessage = _ref.onPressInMessage,
    onPressMessage = _ref.onPressMessage,
    PollContent = _ref.PollContent,
    ReactionListBottom = _ref.ReactionListBottom,
    reactionListPosition = _ref.reactionListPosition,
    ReactionListTop = _ref.ReactionListTop,
    removeMessage = _ref.removeMessage,
    Reply = _ref.Reply,
    retrySendMessage = _ref.retrySendMessage,
    ScrollToBottomButton = _ref.ScrollToBottomButton,
    selectReaction = _ref.selectReaction,
    sendReaction = _ref.sendReaction,
    shouldShowUnreadUnderlay = _ref.shouldShowUnreadUnderlay,
    StreamingMessageView = _ref.StreamingMessageView,
    supportedReactions = _ref.supportedReactions,
    targetedMessage = _ref.targetedMessage,
    TypingIndicator = _ref.TypingIndicator,
    TypingIndicatorContainer = _ref.TypingIndicatorContainer,
    UnreadMessagesNotification = _ref.UnreadMessagesNotification,
    updateMessage = _ref.updateMessage,
    UrlPreview = _ref.UrlPreview,
    VideoThumbnail = _ref.VideoThumbnail;
  var additionalTouchablePropsLength = Object.keys(additionalPressableProps || {}).length;
  var markdownRulesLength = Object.keys(markdownRules || {}).length;
  var messageContentOrderValue = messageContentOrder.join();
  var supportedReactionsLength = supportedReactions == null ? void 0 : supportedReactions.length;
  var messagesContext = (0, _react.useMemo)(function () {
    return {
      additionalPressableProps: additionalPressableProps,
      Attachment: Attachment,
      AttachmentActions: AttachmentActions,
      AudioAttachment: AudioAttachment,
      Card: Card,
      CardCover: CardCover,
      CardFooter: CardFooter,
      CardHeader: CardHeader,
      customMessageSwipeAction: customMessageSwipeAction,
      DateHeader: DateHeader,
      deletedMessagesVisibilityType: deletedMessagesVisibilityType,
      deleteMessage: deleteMessage,
      deleteReaction: deleteReaction,
      disableTypingIndicator: disableTypingIndicator,
      dismissKeyboardOnMessageTouch: dismissKeyboardOnMessageTouch,
      enableMessageGroupingByUser: enableMessageGroupingByUser,
      enableSwipeToReply: enableSwipeToReply,
      FileAttachment: FileAttachment,
      FileAttachmentGroup: FileAttachmentGroup,
      FileAttachmentIcon: FileAttachmentIcon,
      FlatList: FlatList,
      forceAlignMessages: forceAlignMessages,
      Gallery: Gallery,
      getMessageGroupStyle: getMessageGroupStyle,
      getMessagesGroupStyles: getMessagesGroupStyles,
      Giphy: Giphy,
      giphyVersion: giphyVersion,
      handleBan: handleBan,
      handleCopy: handleCopy,
      handleDelete: handleDelete,
      handleDeleteForMe: handleDeleteForMe,
      handleEdit: handleEdit,
      handleFlag: handleFlag,
      handleMarkUnread: handleMarkUnread,
      handleMute: handleMute,
      handlePinMessage: handlePinMessage,
      handleQuotedReply: handleQuotedReply,
      handleReaction: handleReaction,
      handleRetry: handleRetry,
      handleThreadReply: handleThreadReply,
      hasCreatePoll: hasCreatePoll,
      ImageLoadingFailedIndicator: ImageLoadingFailedIndicator,
      ImageLoadingIndicator: ImageLoadingIndicator,
      ImageReloadIndicator: ImageReloadIndicator,
      initialScrollToFirstUnreadMessage: initialScrollToFirstUnreadMessage,
      InlineDateSeparator: InlineDateSeparator,
      InlineUnreadIndicator: InlineUnreadIndicator,
      isAttachmentEqual: isAttachmentEqual,
      isMessageAIGenerated: isMessageAIGenerated,
      legacyImageViewerSwipeBehaviour: legacyImageViewerSwipeBehaviour,
      markdownRules: markdownRules,
      Message: Message,
      MessageActionList: MessageActionList,
      MessageActionListItem: MessageActionListItem,
      messageActions: messageActions,
      MessageAvatar: MessageAvatar,
      MessageBlocked: MessageBlocked,
      MessageBounce: MessageBounce,
      MessageContent: MessageContent,
      messageContentOrder: messageContentOrder,
      MessageDeleted: MessageDeleted,
      MessageEditedTimestamp: MessageEditedTimestamp,
      MessageError: MessageError,
      MessageFooter: MessageFooter,
      MessageHeader: MessageHeader,
      MessageList: MessageList,
      MessageLocation: MessageLocation,
      MessageMenu: MessageMenu,
      MessagePinnedHeader: MessagePinnedHeader,
      MessageReactionPicker: MessageReactionPicker,
      MessageReplies: MessageReplies,
      MessageRepliesAvatars: MessageRepliesAvatars,
      MessageSimple: MessageSimple,
      MessageStatus: MessageStatus,
      MessageSwipeContent: MessageSwipeContent,
      messageSwipeToReplyHitSlop: messageSwipeToReplyHitSlop,
      MessageSystem: MessageSystem,
      MessageText: MessageText,
      messageTextNumberOfLines: messageTextNumberOfLines,
      MessageTimestamp: MessageTimestamp,
      MessageUserReactions: MessageUserReactions,
      MessageUserReactionsAvatar: MessageUserReactionsAvatar,
      MessageUserReactionsItem: MessageUserReactionsItem,
      myMessageTheme: myMessageTheme,
      onLongPressMessage: onLongPressMessage,
      onPressInMessage: onPressInMessage,
      onPressMessage: onPressMessage,
      PollContent: PollContent,
      ReactionListBottom: ReactionListBottom,
      reactionListPosition: reactionListPosition,
      ReactionListTop: ReactionListTop,
      removeMessage: removeMessage,
      Reply: Reply,
      retrySendMessage: retrySendMessage,
      ScrollToBottomButton: ScrollToBottomButton,
      selectReaction: selectReaction,
      sendReaction: sendReaction,
      shouldShowUnreadUnderlay: shouldShowUnreadUnderlay,
      StreamingMessageView: StreamingMessageView,
      supportedReactions: supportedReactions,
      targetedMessage: targetedMessage,
      TypingIndicator: TypingIndicator,
      TypingIndicatorContainer: TypingIndicatorContainer,
      UnreadMessagesNotification: UnreadMessagesNotification,
      updateMessage: updateMessage,
      UrlPreview: UrlPreview,
      VideoThumbnail: VideoThumbnail
    };
  }, [additionalTouchablePropsLength, channelId, disableTypingIndicator, dismissKeyboardOnMessageTouch, initialScrollToFirstUnreadMessage, markdownRulesLength, messageContentOrderValue, supportedReactionsLength, myMessageTheme, targetedMessage, hasCreatePoll]);
  return messagesContext;
};
