// stream-chat-react-native-core/lib/commonjs/contexts/chatConfigContext/ChatConfigContext.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useChatConfigContext = exports.chatConfigContextDefaultvalue = exports.ChatConfigProvider = exports.ChatConfigContext = void 0;
var _react = _interopRequireWildcard(require("react"));
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/contexts/chatConfigContext/ChatConfigContext.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var chatConfigContextDefaultvalue = exports.chatConfigContextDefaultvalue = {
  resizableCDNHosts: ['.stream-io-cdn.com']
};
var ChatConfigContext = exports.ChatConfigContext = _react.default.createContext(chatConfigContextDefaultvalue);
var ChatConfigProvider = exports.ChatConfigProvider = function ChatConfigProvider(_ref) {
  var children = _ref.children,
    _ref$value = _ref.value,
    value = _ref$value === void 0 ? chatConfigContextDefaultvalue : _ref$value;
  return (0, _jsxRuntime.jsx)(ChatConfigContext.Provider, {
    value: value,
    children: children
  });
};
var useChatConfigContext = exports.useChatConfigContext = function useChatConfigContext() {
  var contextValue = (0, _react.useContext)(ChatConfigContext);
  return contextValue;
};
