// stream-chat-react-native-core/lib/module/components/MessagePreview/MessagePreview.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessagePreview = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _contexts = require("../../contexts");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessagePreview/MessagePreview.tsx";
var MessagePreview = exports.MessagePreview = function MessagePreview(_ref) {
  var previews = _ref.previews;
  var _useTheme = (0, _contexts.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    message = _useTheme$theme.messagePreview.message,
    _useTheme$theme$color = _useTheme$theme.colors,
    accent_blue = _useTheme$theme$color.accent_blue,
    grey = _useTheme$theme$color.grey;
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: styles.container,
    children: previews == null ? void 0 : previews.map(function (preview, index) {
      return preview.text ? (0, _jsxRuntime.jsx)(_reactNative.Text, {
        numberOfLines: 1,
        style: [styles.message, {
          color: preview != null && preview.draft ? accent_blue : grey
        }, preview.bold ? styles.bold : {}, message],
        children: preview.text
      }, `${preview.text}_${index}`) : null;
    })
  });
};
var styles = _reactNative.StyleSheet.create({
  bold: {
    fontWeight: '500'
  },
  container: {
    flexDirection: 'row',
    flexShrink: 1
  },
  message: {
    flexShrink: 1,
    marginRight: 2
  }
});
