// stream-chat-react-native-core/lib/module/components/MessageMenu/MessageActionListItem.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageActionListItem = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageMenu/MessageActionListItem.tsx";
var MessageActionListItem = exports.MessageActionListItem = function MessageActionListItem(props) {
  var action = props.action,
    actionType = props.actionType,
    icon = props.icon,
    title = props.title,
    titleStyle = props.titleStyle;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    black = _useTheme$theme.colors.black,
    _useTheme$theme$messa = _useTheme$theme.messageMenu.actionListItem,
    container = _useTheme$theme$messa.container,
    iconTheme = _useTheme$theme$messa.icon,
    titleTheme = _useTheme$theme$messa.title;
  return (0, _jsxRuntime.jsx)(_reactNative.Pressable, {
    onPress: action,
    style: function style(_ref) {
      var pressed = _ref.pressed;
      return [{
        opacity: pressed ? 0.5 : 1
      }];
    },
    children: (0, _jsxRuntime.jsxs)(_reactNative.View, {
      accessibilityLabel: `${actionType} action list item`,
      style: [styles.container, container],
      children: [(0, _jsxRuntime.jsx)(_reactNative.View, {
        style: iconTheme,
        children: icon
      }), (0, _jsxRuntime.jsx)(_reactNative.Text, {
        style: [styles.titleStyle, {
          color: black
        }, titleStyle, titleTheme],
        children: title
      })]
    })
  });
};
var styles = _reactNative.StyleSheet.create({
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    paddingVertical: 8
  },
  titleStyle: {
    paddingLeft: 16
  }
});
