// stream-chat-react-native-core/lib/module/components/MessageList/hooks/useMessageGroupStyles.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMessageGroupStyles = void 0;
var _react = require("react");
var _useMessageDateSeparator = require("./useMessageDateSeparator");
var _getGroupStyles = require("../utils/getGroupStyles");
var useMessageGroupStyles = exports.useMessageGroupStyles = function useMessageGroupStyles(_ref) {
  var noGroupByUser = _ref.noGroupByUser,
    dateSeparatorDate = _ref.dateSeparatorDate,
    maxTimeBetweenGroupedMessages = _ref.maxTimeBetweenGroupedMessages,
    message = _ref.message,
    previousMessage = _ref.previousMessage,
    nextMessage = _ref.nextMessage,
    _ref$getMessageGroupS = _ref.getMessageGroupStyle,
    getMessageGroupStyle = _ref$getMessageGroupS === void 0 ? _getGroupStyles.getGroupStyle : _ref$getMessageGroupS;
  var nextMessageDateSeparatorDate = (0, _useMessageDateSeparator.useMessageDateSeparator)({
    message: nextMessage,
    previousMessage: message
  });
  var groupStyles = (0, _react.useMemo)(function () {
    if (noGroupByUser) {
      return [];
    }
    return getMessageGroupStyle({
      dateSeparatorDate: dateSeparatorDate,
      maxTimeBetweenGroupedMessages: maxTimeBetweenGroupedMessages,
      message: message,
      nextMessage: nextMessage,
      nextMessageDateSeparatorDate: nextMessageDateSeparatorDate,
      previousMessage: previousMessage
    });
  }, [noGroupByUser, getMessageGroupStyle, dateSeparatorDate, maxTimeBetweenGroupedMessages, message, nextMessage, nextMessageDateSeparatorDate, previousMessage]);
  return groupStyles;
};
