// stream-chat-react-native-core/lib/module/contexts/messageComposerContext/MessageComposerContext.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMessageComposerContext = exports.MessageComposerProvider = exports.MessageComposerContext = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _MessageComposerAPIContext = require("./MessageComposerAPIContext");
var _hooks = require("../../hooks");
var _useCreateMessageComposer = require("../messageInputContext/hooks/useCreateMessageComposer");
var _defaultBaseContextValue = require("../utils/defaultBaseContextValue");
var _isTestEnvironment = require("../utils/isTestEnvironment");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/contexts/messageComposerContext/MessageComposerContext.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MessageComposerContext = exports.MessageComposerContext = _react.default.createContext(_defaultBaseContextValue.DEFAULT_BASE_CONTEXT_VALUE);
var MessageComposerProvider = exports.MessageComposerProvider = function MessageComposerProvider(_ref) {
  var children = _ref.children,
    value = _ref.value;
  var _useState = (0, _react.useState)(undefined),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    editing = _useState2[0],
    setEditing = _useState2[1];
  var setEditingState = (0, _hooks.useStableCallback)(function (message) {
    setEditing(message);
  });
  var clearEditingState = (0, _hooks.useStableCallback)(function () {
    return setEditing(undefined);
  });
  var messageComposerContextValue = (0, _react.useMemo)(function () {
    return Object.assign({
      editing: editing
    }, value);
  }, [editing, value]);
  var messageComposer = (0, _useCreateMessageComposer.useCreateMessageComposer)(messageComposerContextValue);
  var setQuotedMessage = (0, _hooks.useStableCallback)(function (message) {
    return messageComposer.setQuotedMessage(message);
  });
  var messageComposerAPIContextValue = (0, _react.useMemo)(function () {
    return {
      clearEditingState: clearEditingState,
      setEditingState: setEditingState,
      setQuotedMessage: setQuotedMessage
    };
  }, [clearEditingState, setEditingState, setQuotedMessage]);
  return (0, _jsxRuntime.jsx)(MessageComposerContext.Provider, {
    value: messageComposerContextValue,
    children: (0, _jsxRuntime.jsx)(_MessageComposerAPIContext.MessageComposerAPIProvider, {
      value: messageComposerAPIContextValue,
      children: children
    })
  });
};
var useMessageComposerContext = exports.useMessageComposerContext = function useMessageComposerContext() {
  var contextValue = (0, _react.useContext)(MessageComposerContext);
  if (contextValue === _defaultBaseContextValue.DEFAULT_BASE_CONTEXT_VALUE && !(0, _isTestEnvironment.isTestEnvironment)()) {
    throw new Error('The useMessageComposerContext hook was called outside of the MessageComposerContext provider.');
  }
  return contextValue;
};
