// stream-chat-react-native-core/lib/module/components/MessageList/InlineUnreadIndicator.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlineUnreadIndicator = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _TranslationContext = require("../../contexts/translationContext/TranslationContext");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageList/InlineUnreadIndicator.tsx";
var InlineUnreadIndicator = exports.InlineUnreadIndicator = function InlineUnreadIndicator() {
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    _useTheme$theme$color = _useTheme$theme.colors,
    grey = _useTheme$theme$color.grey,
    light_gray = _useTheme$theme$color.light_gray,
    _useTheme$theme$messa = _useTheme$theme.messageList.inlineUnreadIndicator,
    container = _useTheme$theme$messa.container,
    text = _useTheme$theme$messa.text;
  var _useTranslationContex = (0, _TranslationContext.useTranslationContext)(),
    t = _useTranslationContex.t;
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    accessibilityLabel: "Inline unread indicator",
    style: [styles.container, {
      backgroundColor: light_gray
    }, container],
    children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: [styles.text, {
        color: grey
      }, text],
      children: t('Unread Messages')
    })
  });
};
var styles = _reactNative.StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 2,
    padding: 10
  },
  text: {
    fontSize: 12
  }
});
