// stream-chat-react-native-core/lib/module/components/MessageList/NetworkDownIndicator.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NetworkDownIndicator = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _ChannelContext = require("../../contexts/channelContext/ChannelContext");
var _ChatContext = require("../../contexts/chatContext/ChatContext");
var _ThemeContext = require("../../contexts/themeContext/ThemeContext");
var _TranslationContext = require("../../contexts/translationContext/TranslationContext");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/MessageList/NetworkDownIndicator.tsx";
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = _reactNative.StyleSheet.create({
  errorNotification: {
    alignItems: 'center',
    left: 0,
    paddingVertical: 4,
    position: 'absolute',
    right: 0,
    top: 0
  },
  errorNotificationText: {
    color: '#FFFFFF',
    fontSize: 14
  }
});
var NetworkDownIndicator = exports.NetworkDownIndicator = function NetworkDownIndicator() {
  var _useChannelContext = (0, _ChannelContext.useChannelContext)(),
    error = _useChannelContext.error;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    isOnline = _useChatContext.isOnline;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    grey = _useTheme$theme.colors.grey,
    _useTheme$theme$messa = _useTheme$theme.messageList,
    errorNotification = _useTheme$theme$messa.errorNotification,
    errorNotificationText = _useTheme$theme$messa.errorNotificationText;
  var _useTranslationContex = (0, _TranslationContext.useTranslationContext)(),
    t = _useTranslationContex.t;
  var indicatorText = (0, _react.useMemo)(function () {
    if (!isOnline) {
      return t('Reconnecting...');
    } else if (error) {
      return t('Error loading messages for this channel...');
    }
    return '';
  }, [error, isOnline, t]);
  if (!indicatorText) {
    return null;
  }
  return (0, _jsxRuntime.jsx)(_reactNative.View, {
    style: [styles.errorNotification, {
      backgroundColor: `${grey}E6`
    }, errorNotification],
    testID: "error-notification",
    children: (0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: [styles.errorNotificationText, errorNotificationText],
      children: indicatorText
    })
  });
};
