// stream-chat-react-native-core/lib/module/contexts/themeContext/utils/theme.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultTheme = exports.DEFAULT_STATUS_ICON_SIZE = exports.Colors = exports.BASE_AVATAR_SIZE = void 0;
var DEFAULT_STATUS_ICON_SIZE = exports.DEFAULT_STATUS_ICON_SIZE = 16;
var BASE_AVATAR_SIZE = exports.BASE_AVATAR_SIZE = 32;
var Colors = exports.Colors = {
  accent_blue: '#005FFF',
  accent_dark_blue: '#005DFF',
  accent_error: '#FF3842',
  accent_green: '#20E070',
  accent_info: '#1FE06F',
  accent_red: '#FF3742',
  bg_gradient_end: '#F7F7F7',
  bg_gradient_start: '#FCFCFC',
  bg_user: '#F7F7F8',
  black: '#000000',
  blue_alice: '#E9F2FF',
  border: '#00000014',
  code_block: '#DDDDDD',
  disabled: '#B4BBBA',
  grey: '#7A7A7A',
  grey_dark: '#72767E',
  grey_gainsboro: '#DBDBDB',
  grey_whisper: '#ECEBEB',
  icon_background: '#FFFFFF',
  label_bg_transparent: '#00000033',
  light_blue: '#E0F0FF',
  light_gray: '#E9EAED',
  modal_shadow: '#00000099',
  overlay: '#000000CC',
  shadow_icon: '#00000040',
  static_black: '#000000',
  static_white: '#ffffff',
  targetedMessageBackground: '#FBF4DD',
  text_high_emphasis: '#080707',
  text_low_emphasis: '#7E828B',
  transparent: 'transparent',
  white: '#FFFFFF',
  white_smoke: '#F2F2F2',
  white_snow: '#FCFCFC'
};
var defaultTheme = exports.defaultTheme = {
  aiTypingIndicatorView: {
    container: {},
    text: {}
  },
  attachmentPicker: {
    bottomSheetContentContainer: {},
    durationText: {},
    errorButtonText: {},
    errorContainer: {},
    errorText: {},
    handle: {
      container: {},
      indicator: {}
    },
    image: {},
    imageOverlay: {},
    imageOverlaySelectedComponent: {
      check: {}
    }
  },
  attachmentSelectionBar: {
    container: {},
    icon: {}
  },
  audioAttachment: {
    container: {},
    leftContainer: {},
    playPauseButton: {},
    progressControlContainer: {},
    progressDurationText: {},
    rightContainer: {},
    speedChangeButton: {},
    speedChangeButtonText: {}
  },
  avatar: {
    BASE_AVATAR_SIZE: BASE_AVATAR_SIZE,
    container: {},
    image: {
      borderRadius: 16,
      height: 32,
      width: 32
    },
    presenceIndicator: {
      cx: 6,
      cy: 6,
      r: 5,
      strokeWidth: 2
    },
    presenceIndicatorContainer: {}
  },
  bottomSheetModal: {
    container: {},
    contentContainer: {},
    handle: {},
    overlay: {},
    wrapper: {}
  },
  channel: {
    selectChannel: {}
  },
  channelListFooterLoadingIndicator: {
    container: {}
  },
  channelListHeaderErrorIndicator: {
    container: {},
    errorText: {}
  },
  channelListLoadingIndicator: {
    container: {}
  },
  channelListMessenger: {
    flatList: {},
    flatListContent: {}
  },
  channelListSkeleton: {
    animationTime: 1800,
    background: {},
    container: {},
    gradientStart: {
      stopOpacity: 0
    },
    gradientStop: {
      stopOpacity: 0.5
    },
    height: 64
  },
  channelPreview: {
    avatar: {
      size: 40
    },
    checkAllIcon: {
      height: DEFAULT_STATUS_ICON_SIZE,
      width: DEFAULT_STATUS_ICON_SIZE
    },
    checkIcon: {
      height: DEFAULT_STATUS_ICON_SIZE,
      width: DEFAULT_STATUS_ICON_SIZE
    },
    container: {},
    contentContainer: {},
    date: {},
    message: {
      container: {}
    },
    mutedStatus: {
      height: 20,
      iconStyle: {},
      width: 20
    },
    row: {},
    title: {},
    unreadContainer: {},
    unreadText: {}
  },
  colors: Object.assign({}, Colors),
  dateHeader: {
    container: {},
    text: {}
  },
  emptyStateIndicator: {
    channelContainer: {},
    channelDetails: {},
    channelTitle: {},
    messageContainer: {},
    messageTitle: {}
  },
  groupAvatar: {
    container: {},
    image: {
      resizeMode: 'cover'
    }
  },
  iconBadge: {
    icon: {},
    iconInner: {},
    unreadCount: {}
  },
  iconSquare: {
    container: {},
    image: {}
  },
  imageGallery: {
    footer: {
      centerContainer: {},
      container: {},
      imageCountText: {},
      innerContainer: {},
      leftContainer: {},
      rightContainer: {}
    },
    grid: {
      container: {},
      contentContainer: {},
      gridAvatar: {},
      gridAvatarWrapper: {},
      gridImage: {},
      handle: {},
      handleText: {},
      overlay: {}
    },
    header: {
      centerContainer: {},
      container: {},
      dateText: {},
      innerContainer: {},
      leftContainer: {},
      rightContainer: {},
      usernameText: {}
    },
    videoControl: {
      durationTextStyle: {},
      progressDurationText: {},
      roundedView: {},
      videoContainer: {}
    }
  },
  inlineDateSeparator: {
    container: {},
    text: {}
  },
  loadingDots: {
    container: {},
    loadingDot: {},
    spacing: 4
  },
  loadingErrorIndicator: {
    container: {},
    errorText: {},
    retryText: {}
  },
  loadingIndicator: {
    container: {},
    loadingText: {}
  },
  messageInput: {
    attachButton: {},
    attachButtonContainer: {},
    attachmentSelectionBar: {},
    attachmentSeparator: {},
    attachmentUnsupportedIndicator: {
      container: {},
      text: {},
      warningIcon: {}
    },
    attachmentUploadPreviewList: {
      filesFlatList: {},
      imagesFlatList: {},
      wrapper: {}
    },
    audioRecorder: {
      arrowLeftIcon: {},
      checkContainer: {},
      circleStopIcon: {},
      deleteContainer: {},
      deleteIcon: {},
      micContainer: {},
      micIcon: {},
      pausedContainer: {},
      sendCheckIcon: {},
      slideToCancelContainer: {}
    },
    audioRecordingButton: {
      container: {},
      micIcon: {}
    },
    audioRecordingInProgress: {
      container: {},
      durationText: {}
    },
    audioRecordingLockIndicator: {
      arrowUpIcon: {},
      container: {},
      lockIcon: {}
    },
    audioRecordingPreview: {
      container: {},
      currentTime: {},
      infoContainer: {},
      pauseIcon: {},
      playIcon: {},
      progressBar: {}
    },
    audioRecordingWaveform: {
      container: {},
      waveform: {}
    },
    autoCompleteInputContainer: {},
    commandInput: {
      closeButton: {},
      container: {},
      text: {}
    },
    commandsButton: {},
    composerContainer: {},
    container: {},
    cooldownTimer: {
      container: {},
      text: {}
    },
    dismissAttachmentUpload: {
      dismiss: {},
      dismissIcon: {},
      dismissIconColor: ''
    },
    editingBoxContainer: {},
    editingBoxHeader: {},
    editingBoxHeaderTitle: {},
    editingStateHeader: {
      editingBoxHeader: {},
      editingBoxHeaderTitle: {}
    },
    fileAttachmentUploadPreview: {
      fileContainer: {},
      filenameText: {},
      fileSizeText: {},
      fileTextContainer: {},
      uploadProgressOverlay: {},
      wrapper: {}
    },
    fileUploadPreview: {
      flatList: {}
    },
    focusedInputBoxContainer: {},
    imageAttachmentUploadPreview: {
      itemContainer: {},
      upload: {}
    },
    imageUploadPreview: {
      flatList: {}
    },
    inputBox: {},
    inputBoxContainer: {},
    micButtonContainer: {},
    moreOptionsButton: {},
    nativeAttachmentPicker: {
      buttonContainer: {},
      buttonDimmerStyle: {},
      container: {}
    },
    optionsContainer: {},
    replyContainer: {},
    searchIcon: {},
    sendButton: {},
    sendButtonContainer: {},
    sendMessageDisallowedIndicator: {
      container: {},
      text: {}
    },
    sendRightIcon: {},
    sendUpIcon: {},
    showThreadMessageInChannelButton: {
      check: {},
      checkBoxActive: {},
      checkBoxInactive: {},
      container: {},
      innerContainer: {},
      text: {}
    },
    stopMessageStreamingButton: {},
    stopMessageStreamingIcon: {},
    suggestions: {
      command: {
        args: {},
        container: {},
        iconContainer: {},
        title: {}
      },
      emoji: {
        container: {},
        text: {}
      },
      header: {
        container: {},
        title: {}
      },
      item: {},
      mention: {
        avatarSize: 40,
        column: {},
        container: {},
        name: {},
        tag: {}
      }
    },
    suggestionsListContainer: {
      container: {},
      flatlist: {}
    },
    uploadProgressIndicator: {
      container: {},
      indicatorColor: '',
      overlay: {}
    },
    videoAttachmentUploadPreview: {
      itemContainer: {},
      recorderIcon: {},
      recorderIconContainer: {},
      upload: {}
    }
  },
  messageList: {
    container: {},
    contentContainer: {},
    errorNotification: {},
    errorNotificationText: {},
    inlineUnreadIndicator: {
      container: {},
      text: {}
    },
    listContainer: {},
    messageContainer: {},
    messageSystem: {
      container: {},
      dateText: {},
      line: {},
      text: {},
      textContainer: {}
    },
    scrollToBottomButton: {
      container: {},
      touchable: {},
      unreadCountNotificationContainer: {},
      unreadCountNotificationText: {},
      wrapper: {}
    },
    typingIndicatorContainer: {},
    unreadMessagesNotification: {
      closeButtonContainer: {},
      closeIcon: {},
      container: {},
      text: {}
    }
  },
  messageMenu: {
    actionList: {
      container: {},
      contentContainer: {}
    },
    actionListItem: {
      container: {},
      icon: {},
      title: {}
    },
    bottomSheet: {},
    reactionButton: {},
    reactionPicker: {
      buttonContainer: {},
      container: {},
      contentContainer: {},
      reactionIconSize: 24
    },
    userReactions: {
      avatarContainer: {},
      avatarInnerContainer: {},
      avatarName: {},
      avatarNameContainer: {},
      avatarSize: 64,
      container: {},
      contentContainer: {},
      flatlistColumnContainer: {},
      flatlistContainer: {},
      radius: 2,
      reactionBubble: {},
      reactionBubbleBackground: {},
      reactionBubbleBorderRadius: 24,
      reactionSelectorContainer: {},
      reactionsText: {},
      title: {}
    }
  },
  messagePreview: {
    message: {}
  },
  messageSimple: {
    actions: {
      button: {},
      buttonText: {},
      container: {}
    },
    avatarWrapper: {
      container: {},
      leftAlign: {
        marginRight: 8
      },
      rightAlign: {
        marginLeft: 8
      },
      spacer: {
        height: BASE_AVATAR_SIZE,
        width: BASE_AVATAR_SIZE
      }
    },
    card: {
      authorName: {},
      authorNameContainer: {},
      authorNameFooter: {},
      authorNameFooterContainer: {},
      authorNameMask: {},
      container: {},
      cover: {},
      footer: {
        description: {},
        title: {
          fontWeight: '700'
        }
      },
      noURI: {
        borderLeftWidth: 2,
        paddingLeft: 8
      },
      playButtonStyle: {
        durationTextStyle: {},
        roundedView: {},
        videoContainer: {}
      },
      playIcon: {
        height: 32,
        width: 32
      }
    },
    container: {},
    content: {
      container: {
        borderRadiusL: 16,
        borderRadiusS: 0
      },
      containerInner: {},
      deletedContainer: {},
      deletedContainerInner: {},
      deletedMetaText: {
        paddingHorizontal: 5
      },
      deletedText: {
        em: {
          fontSize: 15,
          fontStyle: 'italic',
          fontWeight: '400'
        }
      },
      editedLabel: {},
      editedTimestampContainer: {},
      errorContainer: {
        paddingRight: 12,
        paddingTop: 0
      },
      errorIcon: {
        height: 20,
        width: 20
      },
      errorIconContainer: {
        bottom: -2,
        position: 'absolute',
        right: -12
      },
      eyeIcon: {
        height: 16,
        width: 16
      },
      lastMessageContainer: {},
      markdown: {},
      messageGroupedSingleOrBottomContainer: {},
      messageGroupedTopContainer: {},
      messageUser: {
        fontSize: 12,
        fontWeight: '700',
        paddingRight: 6
      },
      metaContainer: {},
      metaText: {
        fontSize: 12
      },
      replyBorder: {},
      replyContainer: {},
      textContainer: {
        onlyEmojiMarkdown: {
          text: {
            fontSize: 50
          }
        }
      },
      timestampText: {},
      wrapper: {}
    },
    contentWrapper: {},
    file: {
      container: {},
      details: {},
      fileSize: {},
      icon: {},
      title: {}
    },
    fileAttachmentGroup: {
      attachmentContainer: {},
      container: {}
    },
    gallery: {
      galleryContainer: {},
      galleryItemColumn: {},
      gridHeight: 195,
      gridWidth: 256,
      image: {},
      imageBorderRadius: undefined,
      imageContainer: {},
      imageContainerStyle: {},
      maxHeight: 300,
      maxWidth: 256,
      minHeight: 100,
      minWidth: 170,
      moreImagesContainer: {},
      moreImagesText: {},
      thumbnail: {}
    },
    giphy: {
      buttonContainer: {},
      cancel: {},
      container: {},
      giphy: {},
      giphyContainer: {},
      giphyHeaderText: {},
      giphyHeaderTitle: {},
      giphyMask: {},
      giphyMaskText: {},
      header: {},
      selectionContainer: {},
      send: {},
      shuffle: {},
      title: {}
    },
    headerWrapper: {},
    lastMessageContainer: {},
    loadingIndicator: {
      container: {},
      roundedView: {}
    },
    messageBlocked: {
      container: {},
      line: {},
      text: {},
      textContainer: {}
    },
    messageGroupedSingleOrBottomContainer: {},
    messageGroupedTopContainer: {},
    pinnedHeader: {
      container: {},
      label: {}
    },
    reactionListBottom: {
      contentContainer: {},
      item: {
        container: {},
        countText: {},
        icon: {},
        iconSize: 16
      }
    },
    reactionListTop: {
      container: {},
      item: {
        container: {},
        icon: {},
        iconFillColor: Colors.accent_blue,
        iconSize: 24,
        iconUnFillColor: Colors.grey,
        reactionSize: 24
      },
      position: 16
    },
    replies: {
      avatar: {},
      avatarContainerMultiple: {},
      avatarContainerSingle: {},
      container: {},
      leftAvatarsContainer: {},
      leftCurve: {},
      messageRepliesText: {},
      rightAvatarsContainer: {},
      rightCurve: {}
    },
    status: {
      checkAllIcon: {
        height: DEFAULT_STATUS_ICON_SIZE,
        width: DEFAULT_STATUS_ICON_SIZE
      },
      checkIcon: {
        height: DEFAULT_STATUS_ICON_SIZE,
        width: DEFAULT_STATUS_ICON_SIZE
      },
      readByCount: {},
      statusContainer: {},
      timeIcon: {
        height: DEFAULT_STATUS_ICON_SIZE,
        width: DEFAULT_STATUS_ICON_SIZE
      }
    },
    swipeContentContainer: {},
    swipeLeftContent: {
      container: {}
    },
    targetedMessageContainer: {},
    videoThumbnail: {
      container: {},
      roundedView: {}
    },
    wrapper: {}
  },
  poll: {
    allOptions: {
      listContainer: {},
      titleContainer: {},
      titleText: {},
      wrapper: {}
    },
    answersList: {
      buttonContainer: {},
      container: {},
      item: {
        answerText: {},
        container: {},
        infoContainer: {},
        userInfoContainer: {}
      }
    },
    button: {
      container: {},
      text: {}
    },
    createContent: {
      addComment: {
        title: {},
        wrapper: {}
      },
      anonymousPoll: {
        title: {},
        wrapper: {}
      },
      headerContainer: {},
      maxVotes: {
        input: {},
        validationText: {},
        wrapper: {}
      },
      multipleAnswers: {
        row: {},
        title: {},
        wrapper: {}
      },
      name: {
        input: {},
        title: {}
      },
      pollOptions: {
        addOption: {
          text: {},
          wrapper: {}
        },
        container: {},
        optionStyle: {
          input: {},
          validationErrorText: {},
          wrapper: {}
        },
        title: {}
      },
      scrollView: {},
      sendButton: {},
      suggestOption: {
        title: {},
        wrapper: {}
      }
    },
    fullResults: {
      container: {},
      contentContainer: {},
      headerContainer: {},
      headerText: {}
    },
    inputDialog: {
      button: {},
      buttonContainer: {},
      container: {},
      input: {},
      title: {},
      transparentContainer: {}
    },
    message: {
      container: {},
      header: {
        subtitle: {},
        title: {}
      },
      option: {
        container: {},
        progressBar: {},
        progressBarEmptyFill: '',
        progressBarVotedFill: '',
        progressBarWinnerFill: '',
        text: {},
        voteButtonActive: '',
        voteButtonContainer: {},
        voteButtonInactive: '',
        votesContainer: {},
        wrapper: {}
      },
      optionsWrapper: {}
    },
    modalHeader: {
      container: {},
      title: {}
    },
    results: {
      container: {},
      item: {
        container: {},
        headerContainer: {},
        title: {},
        voteCount: {}
      },
      scrollView: {},
      title: {},
      vote: {
        container: {},
        dateText: {},
        userName: {}
      }
    }
  },
  progressControl: {
    container: {},
    filledColor: '',
    filledStyles: {},
    thumb: {}
  },
  reply: {
    container: {},
    fileAttachmentContainer: {},
    imageAttachment: {},
    markdownStyles: {},
    messageContainer: {},
    secondaryText: {},
    textContainer: {},
    videoThumbnail: {
      container: {},
      image: {}
    }
  },
  screenPadding: 8,
  spinner: {},
  thread: {
    newThread: {
      text: {}
    }
  },
  threadListItem: {
    boldText: {},
    contentRow: {},
    contentTextWrapper: {},
    dateText: {},
    headerRow: {},
    infoRow: {},
    parentMessagePreviewContainer: {},
    parentMessageText: {},
    previewMessageContainer: {},
    touchableWrapper: {},
    unreadBubbleText: {},
    unreadBubbleWrapper: {}
  },
  threadListUnreadBanner: {
    text: {},
    touchableWrapper: {}
  },
  typingIndicator: {
    container: {},
    text: {
      fontSize: 14
    }
  },
  waveProgressBar: {
    container: {},
    thumb: {},
    waveform: {}
  }
};
