// react-native-chat-sdk/lib/commonjs/common/ChatRoom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChatRoomPermissionType = exports.ChatRoom = void 0;
exports.ChatRoomPermissionTypeFromNumber = ChatRoomPermissionTypeFromNumber;
exports.ChatRoomPermissionTypeToString = ChatRoomPermissionTypeToString;
/**
 * The chat room role types.
 */
let ChatRoomPermissionType = exports.ChatRoomPermissionType = /*#__PURE__*/function (ChatRoomPermissionType) {
  ChatRoomPermissionType[ChatRoomPermissionType["None"] = -1] = "None";
  ChatRoomPermissionType[ChatRoomPermissionType["Member"] = 0] = "Member";
  ChatRoomPermissionType[ChatRoomPermissionType["Admin"] = 1] = "Admin";
  ChatRoomPermissionType[ChatRoomPermissionType["Owner"] = 2] = "Owner";
  return ChatRoomPermissionType;
}({});
/**
 * Converts the chat room role type from Int to enum.
 *
 * @param params The chat room role type of the Int type.
 * @returns The chat room role type of the enum type.
 */
function ChatRoomPermissionTypeFromNumber(params) {
  switch (params) {
    case -1:
      return ChatRoomPermissionType.None;
    case 0:
      return ChatRoomPermissionType.Member;
    case 1:
      return ChatRoomPermissionType.Admin;
    case 2:
      return ChatRoomPermissionType.Owner;
    default:
      return params;
  }
}

/**
 * Converts the chat room role type from enum to string.
 *
 * @param params The chat room role type of the enum type.
 * @returns The chat room role type of the string type.
 */
function ChatRoomPermissionTypeToString(params) {
  return ChatRoomPermissionType[params];
}

/**
 * The chat room instance class.
 *
 * To get the correct value, ensure that you call {@link ChatRoomManager.fetchChatRoomInfoFromServer} to get chat room details before calling this method.
 */
class ChatRoom {
  /**
   * The chat room ID.
   */

  /**
   * The chat room name.
   */

  /**
   * The chat room description.
   */

  /**
   * The user ID of the chat room owner.
   */

  /**
   * The chat room announcement.
   */

  /**
   * The number of members in the chat room.
   */

  /**
   * The maximum number of users allowed to join a chat room. This field is specified during the creation of a chat room.
   */

  /**
   * The admin list of the chat room.
   */

  /**
   * The member list of the chat room.
   */

  /**
   * The block list of the chat room.
   */

  /**
   * The mute list of the chat room.
   *
   * @deprecated 2024-12-03 Please use `muteKVList` instead.
   */

  /**
   * The mute list of the chat room.
   *
   * key: userId
   * value: mute expire time
   */

  /**
   * Whether all members are muted in the chat room.
   * - `true`: Yes.
   * - `false`: No.
   */

  /**
   * Current user is in allow list or not.
   * This property is available once join the chat room.
   * This property will be updated when current user is added or removed from the white list.
   * - `true`: In white list.
   * - `false`: Not in white list.
   */

  /**
   * Gets the timestamp(ms) when the chat room was created.
   * This property is available once join the chat room.
   */

  /**
   * Gets the timestamp(ms) when Current user will be unmuted.
   *
   * This property is available once join the chat room.
   * This property will be updated when current use is muted or unmuted.
   *
   * - Current use is not muted if it is zero.
   * - Means cannot get MuteUntilTimeStamp correctly if it is be set with -1;
   */

  /**
   * The role of the current user in the chat room. For role types, see {@link ChatRoomPermissionType}.
   */

  constructor(params) {
    this.roomId = params.roomId;
    this.roomName = params.roomName;
    this.description = params.description;
    this.owner = params.owner;
    this.announcement = params.announcement;
    this.memberCount = params.memberCount;
    this.maxUsers = params.maxUsers;
    this.adminList = params.adminList;
    this.memberList = params.memberList;
    this.blockList = params.blockList;
    this.muteList = params.muteList;
    this.muteKVList = params.muteKVList;
    this.isAllMemberMuted = params.isAllMemberMuted;
    this.permissionType = ChatRoomPermissionTypeFromNumber(params.permissionType);
    this.isInWhitelist = params.isInWhitelist;
    this.createTimestamp = params.createTimestamp;
    this.muteExpireTimestamp = params.muteExpireTimestamp;
  }
}
exports.ChatRoom = ChatRoom;
