// stream-chat-react-native-core/lib/module/components/Message/MessageSimple/MessageWrapper.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageWrapper = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _useMessageDateSeparator = require("../../../components/MessageList/hooks/useMessageDateSeparator");
var _useMessageGroupStyles = require("../../../components/MessageList/hooks/useMessageGroupStyles");
var _ChannelContext = require("../../../contexts/channelContext/ChannelContext");
var _ChatContext = require("../../../contexts/chatContext/ChatContext");
var _MessageListItemContext = require("../../../contexts/messageListItemContext/MessageListItemContext");
var _MessagesContext = require("../../../contexts/messagesContext/MessagesContext");
var _ThemeContext = require("../../../contexts/themeContext/ThemeContext");
var _useStateStore2 = require("../../../hooks/useStateStore");
var _jsxRuntime = require("react/jsx-runtime");
var _this = this,
  _jsxFileName = "/home/runner/work/stream-chat-react-native/stream-chat-react-native/package/src/components/Message/MessageSimple/MessageWrapper.tsx";
var channelUnreadStateSelector = function channelUnreadStateSelector(state) {
  var _state$channelUnreadS, _state$channelUnreadS2, _state$channelUnreadS3, _state$channelUnreadS4;
  return {
    first_unread_message_id: (_state$channelUnreadS = state.channelUnreadState) == null ? void 0 : _state$channelUnreadS.first_unread_message_id,
    last_read_message_id: (_state$channelUnreadS2 = state.channelUnreadState) == null ? void 0 : _state$channelUnreadS2.last_read_message_id,
    last_read_timestamp: (_state$channelUnreadS3 = state.channelUnreadState) == null || (_state$channelUnreadS3 = _state$channelUnreadS3.last_read) == null ? void 0 : _state$channelUnreadS3.getTime(),
    unread_messages: (_state$channelUnreadS4 = state.channelUnreadState) == null ? void 0 : _state$channelUnreadS4.unread_messages
  };
};
var MessageWrapper = exports.MessageWrapper = _react.default.memo(function (props) {
  var _message$user;
  var message = props.message,
    previousMessage = props.previousMessage,
    nextMessage = props.nextMessage;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  var _useChannelContext = (0, _ChannelContext.useChannelContext)(),
    channelUnreadStateStore = _useChannelContext.channelUnreadStateStore,
    channel = _useChannelContext.channel,
    hideDateSeparators = _useChannelContext.hideDateSeparators,
    highlightedMessageId = _useChannelContext.highlightedMessageId,
    maxTimeBetweenGroupedMessages = _useChannelContext.maxTimeBetweenGroupedMessages,
    threadList = _useChannelContext.threadList;
  var _useMessagesContext = (0, _MessagesContext.useMessagesContext)(),
    getMessageGroupStyle = _useMessagesContext.getMessageGroupStyle,
    InlineDateSeparator = _useMessagesContext.InlineDateSeparator,
    InlineUnreadIndicator = _useMessagesContext.InlineUnreadIndicator,
    Message = _useMessagesContext.Message,
    MessageSystem = _useMessagesContext.MessageSystem,
    myMessageTheme = _useMessagesContext.myMessageTheme,
    shouldShowUnreadUnderlay = _useMessagesContext.shouldShowUnreadUnderlay;
  var _useMessageListItemCo = (0, _MessageListItemContext.useMessageListItemContext)(),
    goToMessage = _useMessageListItemCo.goToMessage,
    onThreadSelect = _useMessageListItemCo.onThreadSelect,
    noGroupByUser = _useMessageListItemCo.noGroupByUser,
    modifiedTheme = _useMessageListItemCo.modifiedTheme;
  var dateSeparatorDate = (0, _useMessageDateSeparator.useMessageDateSeparator)({
    hideDateSeparators: hideDateSeparators,
    message: message,
    previousMessage: previousMessage
  });
  var isNewestMessage = nextMessage === undefined;
  var groupStyles = (0, _useMessageGroupStyles.useMessageGroupStyles)({
    dateSeparatorDate: dateSeparatorDate,
    getMessageGroupStyle: getMessageGroupStyle,
    maxTimeBetweenGroupedMessages: maxTimeBetweenGroupedMessages,
    message: message,
    nextMessage: nextMessage,
    noGroupByUser: noGroupByUser,
    previousMessage: previousMessage
  });
  var _useStateStore = (0, _useStateStore2.useStateStore)(channelUnreadStateStore.state, channelUnreadStateSelector),
    first_unread_message_id = _useStateStore.first_unread_message_id,
    last_read_timestamp = _useStateStore.last_read_timestamp,
    last_read_message_id = _useStateStore.last_read_message_id,
    unread_messages = _useStateStore.unread_messages;
  var _useTheme = (0, _ThemeContext.useTheme)(),
    _useTheme$theme = _useTheme.theme,
    messageContainer = _useTheme$theme.messageList.messageContainer,
    screenPadding = _useTheme$theme.screenPadding;
  if (!channel || channel.disconnected) {
    return null;
  }
  var createdAtTimestamp = message.created_at && new Date(message.created_at).getTime();
  var isLastReadMessage = last_read_message_id === message.id || !unread_messages && createdAtTimestamp === last_read_timestamp;
  var showUnreadSeparator = isLastReadMessage && !isNewestMessage && (!!first_unread_message_id || !!unread_messages);
  var showUnreadUnderlay = !!shouldShowUnreadUnderlay && showUnreadSeparator;
  var wrapMessageInTheme = client.userID === ((_message$user = message.user) == null ? void 0 : _message$user.id) && !!myMessageTheme;
  var renderDateSeperator = dateSeparatorDate ? (0, _jsxRuntime.jsx)(InlineDateSeparator, {
    date: dateSeparatorDate
  }) : null;
  var renderMessage = (0, _jsxRuntime.jsx)(Message, {
    goToMessage: goToMessage,
    groupStyles: groupStyles,
    isTargetedMessage: highlightedMessageId === message.id,
    message: message,
    onThreadSelect: onThreadSelect,
    showUnreadUnderlay: showUnreadUnderlay,
    style: [messageContainer],
    threadList: threadList
  });
  return (0, _jsxRuntime.jsxs)(_reactNative.View, {
    testID: `message-list-item-${message.id}`,
    children: [message.type === 'system' ? (0, _jsxRuntime.jsx)(MessageSystem, {
      message: message,
      style: [{
        paddingHorizontal: screenPadding
      }, messageContainer]
    }) : wrapMessageInTheme ? (0, _jsxRuntime.jsx)(_ThemeContext.ThemeProvider, {
      mergedStyle: modifiedTheme,
      children: (0, _jsxRuntime.jsxs)(_reactNative.View, {
        testID: `message-list-item-${message.id}`,
        children: [renderDateSeperator, renderMessage]
      })
    }) : (0, _jsxRuntime.jsxs)(_reactNative.View, {
      testID: `message-list-item-${message.id}`,
      children: [renderDateSeperator, renderMessage]
    }), showUnreadUnderlay && (0, _jsxRuntime.jsx)(InlineUnreadIndicator, {})]
  });
});
